
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>matminer.featurizers package &#8212; matminer 0.6.5 documentation</title>
    <link rel="stylesheet" href="_static/nature.css" type="text/css" />
    <link rel="stylesheet" href="_static/pygments.css" type="text/css" />
    <script id="documentation_options" data-url_root="./" src="_static/documentation_options.js"></script>
    <script src="_static/jquery.js"></script>
    <script src="_static/underscore.js"></script>
    <script src="_static/doctools.js"></script>
    <script src="_static/language_data.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
 
<link href='https://fonts.googleapis.com/css?family=Lato:400,700' rel='stylesheet' type='text/css'>

  </head><body>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             accesskey="I">index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">matminer 0.6.5 documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">matminer.featurizers package</a></li> 
      </ul>
    </div>  

    <div class="document">
      <div class="documentwrapper">
        <div class="bodywrapper">
          <div class="body" role="main">
            
  <div class="section" id="matminer-featurizers-package">
<h1>matminer.featurizers package<a class="headerlink" href="#matminer-featurizers-package" title="Permalink to this headline">¶</a></h1>
<div class="section" id="subpackages">
<h2>Subpackages<a class="headerlink" href="#subpackages" title="Permalink to this headline">¶</a></h2>
<div class="toctree-wrapper compound">
<ul>
<li class="toctree-l1"><a class="reference internal" href="matminer.featurizers.tests.html">matminer.featurizers.tests package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="matminer.featurizers.tests.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="matminer.featurizers.tests.html#module-matminer.featurizers.tests.test_bandstructure">matminer.featurizers.tests.test_bandstructure module</a></li>
<li class="toctree-l2"><a class="reference internal" href="matminer.featurizers.tests.html#module-matminer.featurizers.tests.test_base">matminer.featurizers.tests.test_base module</a></li>
<li class="toctree-l2"><a class="reference internal" href="matminer.featurizers.tests.html#module-matminer.featurizers.tests.test_composition">matminer.featurizers.tests.test_composition module</a></li>
<li class="toctree-l2"><a class="reference internal" href="matminer.featurizers.tests.html#module-matminer.featurizers.tests.test_conversions">matminer.featurizers.tests.test_conversions module</a></li>
<li class="toctree-l2"><a class="reference internal" href="matminer.featurizers.tests.html#module-matminer.featurizers.tests.test_dos">matminer.featurizers.tests.test_dos module</a></li>
<li class="toctree-l2"><a class="reference internal" href="matminer.featurizers.tests.html#module-matminer.featurizers.tests.test_function">matminer.featurizers.tests.test_function module</a></li>
<li class="toctree-l2"><a class="reference internal" href="matminer.featurizers.tests.html#module-matminer.featurizers.tests.test_site">matminer.featurizers.tests.test_site module</a></li>
<li class="toctree-l2"><a class="reference internal" href="matminer.featurizers.tests.html#module-matminer.featurizers.tests.test_structure">matminer.featurizers.tests.test_structure module</a></li>
<li class="toctree-l2"><a class="reference internal" href="matminer.featurizers.tests.html#module-matminer.featurizers.tests">Module contents</a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="matminer.featurizers.utils.html">matminer.featurizers.utils package</a><ul>
<li class="toctree-l2"><a class="reference internal" href="matminer.featurizers.utils.html#subpackages">Subpackages</a><ul>
<li class="toctree-l3"><a class="reference internal" href="matminer.featurizers.utils.tests.html">matminer.featurizers.utils.tests package</a><ul>
<li class="toctree-l4"><a class="reference internal" href="matminer.featurizers.utils.tests.html#submodules">Submodules</a></li>
<li class="toctree-l4"><a class="reference internal" href="matminer.featurizers.utils.tests.html#module-matminer.featurizers.utils.tests.test_cgcnn">matminer.featurizers.utils.tests.test_cgcnn module</a></li>
<li class="toctree-l4"><a class="reference internal" href="matminer.featurizers.utils.tests.html#module-matminer.featurizers.utils.tests.test_grdf">matminer.featurizers.utils.tests.test_grdf module</a></li>
<li class="toctree-l4"><a class="reference internal" href="matminer.featurizers.utils.tests.html#module-matminer.featurizers.utils.tests.test_stats">matminer.featurizers.utils.tests.test_stats module</a></li>
<li class="toctree-l4"><a class="reference internal" href="matminer.featurizers.utils.tests.html#module-matminer.featurizers.utils.tests">Module contents</a></li>
</ul>
</li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="matminer.featurizers.utils.html#submodules">Submodules</a></li>
<li class="toctree-l2"><a class="reference internal" href="matminer.featurizers.utils.html#module-matminer.featurizers.utils.cgcnn">matminer.featurizers.utils.cgcnn module</a></li>
<li class="toctree-l2"><a class="reference internal" href="matminer.featurizers.utils.html#module-matminer.featurizers.utils.grdf">matminer.featurizers.utils.grdf module</a></li>
<li class="toctree-l2"><a class="reference internal" href="matminer.featurizers.utils.html#module-matminer.featurizers.utils.stats">matminer.featurizers.utils.stats module</a></li>
<li class="toctree-l2"><a class="reference internal" href="matminer.featurizers.utils.html#module-matminer.featurizers.utils">Module contents</a></li>
</ul>
</li>
</ul>
</div>
</div>
<div class="section" id="submodules">
<h2>Submodules<a class="headerlink" href="#submodules" title="Permalink to this headline">¶</a></h2>
</div>
<div class="section" id="module-matminer.featurizers.bandstructure">
<span id="matminer-featurizers-bandstructure-module"></span><h2>matminer.featurizers.bandstructure module<a class="headerlink" href="#module-matminer.featurizers.bandstructure" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt id="matminer.featurizers.bandstructure.BandFeaturizer">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.bandstructure.</code><code class="sig-name descname">BandFeaturizer</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">kpoints</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">find_method</span><span class="o">=</span><span class="default_value">'nearest'</span></em>, <em class="sig-param"><span class="n">nbands</span><span class="o">=</span><span class="default_value">2</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.bandstructure.BandFeaturizer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Featurizes a pymatgen band structure object.</p>
<dl>
<dt>Args:</dt><dd><dl>
<dt>kpoints ([1x3 numpy array]): list of fractional coordinates of</dt><dd><p>k-points at which energy is extracted.</p>
</dd>
<dt>find_method (str): the method for finding or interpolating for energy</dt><dd><p>at given kpoints. It does nothing if kpoints is None.
options are:</p>
<blockquote>
<div><dl class="simple">
<dt>‘nearest’: the energy of the nearest available k-point to</dt><dd><p>the input k-point is returned.</p>
</dd>
</dl>
<p>‘linear’: the result of linear interpolation is returned
see the documentation for scipy.interpolate.griddata</p>
</div></blockquote>
</dd>
</dl>
<p>nbands (int): the number of valence/conduction bands to be featurized</p>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.bandstructure.BandFeaturizer.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">kpoints</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">find_method</span><span class="o">=</span><span class="default_value">'nearest'</span></em>, <em class="sig-param"><span class="n">nbands</span><span class="o">=</span><span class="default_value">2</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.bandstructure.BandFeaturizer.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.bandstructure.BandFeaturizer.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.bandstructure.BandFeaturizer.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.bandstructure.BandFeaturizer.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.bandstructure.BandFeaturizer.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.bandstructure.BandFeaturizer.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">bs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.bandstructure.BandFeaturizer.featurize" title="Permalink to this definition">¶</a></dt>
<dd><dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>bs (pymatgen BandStructure or BandStructureSymmLine or their dict):</dt><dd><p>The band structure to featurize. To obtain all features, bs
should include the structure attribute.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><blockquote>
<div><dl class="simple">
<dt>([float]): a list of band structure features. If not bs.structure,</dt><dd><p>features that require the structure will be returned as NaN.</p>
</dd>
</dl>
</div></blockquote>
<dl>
<dt>List of currently supported features:</dt><dd><p>band_gap (eV): the difference between the CBM and VBM energy
is_gap_direct (0.0|1.0): whether the band gap is direct or not
direct_gap (eV): the minimum direct distance of the last</p>
<blockquote>
<div><p>valence band and the first conduction band</p>
</div></blockquote>
<dl class="simple">
<dt>p_ex1_norm (float): k-space distance between Gamma point</dt><dd><p>and k-point of VBM</p>
</dd>
<dt>n_ex1_norm (float): k-space distance between Gamma point</dt><dd><p>and k-point of CBM</p>
</dd>
</dl>
<p>p_ex1_degen: degeneracy of VBM
n_ex1_degen: degeneracy of CBM
if kpoints is provided (e.g. for kpoints == [[0.0, 0.0, 0.0]]):</p>
<blockquote>
<div><dl class="simple">
<dt>n_0.0;0.0;0.0_en: (energy of the first conduction band at</dt><dd><p>[0.0, 0.0, 0.0] - CBM energy)</p>
</dd>
<dt>p_0.0;0.0;0.0_en: (energy of the last valence band at</dt><dd><p>[0.0, 0.0, 0.0] - VBM energy)</p>
</dd>
</dl>
</div></blockquote>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.bandstructure.BandFeaturizer.get_bindex_bspin">
<em class="property">static </em><code class="sig-name descname">get_bindex_bspin</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">extremum</span></em>, <em class="sig-param"><span class="n">is_cbm</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.bandstructure.BandFeaturizer.get_bindex_bspin" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the band index and spin of band extremum</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>extremum (dict): dictionary containing the CBM/VBM, i.e. output of</dt><dd><p>Bandstructure.get_cbm()</p>
</dd>
</dl>
<p>is_cbm (bool): whether the extremum is the CBM or not</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.bandstructure.BandFeaturizer.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.bandstructure.BandFeaturizer.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.bandstructure.BranchPointEnergy">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.bandstructure.</code><code class="sig-name descname">BranchPointEnergy</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">n_vb</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">n_cb</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">calculate_band_edges</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">atol</span><span class="o">=</span><span class="default_value">1e-05</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.bandstructure.BranchPointEnergy" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Branch point energy and absolute band edge position.</p>
<p>Calculates the branch point energy and (optionally) an absolute band
edge position assuming the branch point energy is the center of the gap</p>
<dl>
<dt>Args:</dt><dd><p>n_vb (int): number of valence bands to include in BPE calc
n_cb (int): number of conduction bands to include in BPE calc
calculate_band_edges: (bool) whether to also return band edge</p>
<blockquote>
<div><p>positions</p>
</div></blockquote>
<dl class="simple">
<dt>atol (float): absolute tolerance when finding equivalent fractional</dt><dd><p>k-points in irreducible brillouin zone (IBZ) when weights is None</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.bandstructure.BranchPointEnergy.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">n_vb</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">n_cb</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">calculate_band_edges</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">atol</span><span class="o">=</span><span class="default_value">1e-05</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.bandstructure.BranchPointEnergy.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.bandstructure.BranchPointEnergy.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.bandstructure.BranchPointEnergy.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.bandstructure.BranchPointEnergy.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.bandstructure.BranchPointEnergy.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Returns ([str]): absolute energy levels as provided in the input</dt><dd><p>BandStructure. “absolute” means no reference energy is subtracted
from branch_point_energy, vbm or cbm.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.bandstructure.BranchPointEnergy.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">bs</span></em>, <em class="sig-param"><span class="n">target_gap</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">weights</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.bandstructure.BranchPointEnergy.featurize" title="Permalink to this definition">¶</a></dt>
<dd><dl>
<dt>Args:</dt><dd><p>bs (BandStructure): Uniform (not symm line) band structure
target_gap (float): if set the band gap is scissored to match this</p>
<blockquote>
<div><p>number</p>
</div></blockquote>
<dl class="simple">
<dt>weights ([float]): if set, its length has to be equal to bs.kpoints</dt><dd><p>to explicitly determine the k-point weights when averaging</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>(int) branch point energy on same energy scale as BS eigenvalues</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.bandstructure.BranchPointEnergy.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.bandstructure.BranchPointEnergy.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-matminer.featurizers.base">
<span id="matminer-featurizers-base-module"></span><h2>matminer.featurizers.base module<a class="headerlink" href="#module-matminer.featurizers.base" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt id="matminer.featurizers.base.BaseFeaturizer">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.base.</code><code class="sig-name descname">BaseFeaturizer</code><a class="headerlink" href="#matminer.featurizers.base.BaseFeaturizer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">sklearn.base.BaseEstimator</span></code>, <code class="xref py py-class docutils literal notranslate"><span class="pre">sklearn.base.TransformerMixin</span></code>, <a class="reference external" href="https://docs.python.org/3/library/abc.html#abc.ABC" title="(in Python v3.9)"><code class="xref py py-class docutils literal notranslate"><span class="pre">abc.ABC</span></code></a></p>
<p>Abstract class to calculate features from raw materials input data
such a compound formula or a pymatgen crystal structure or
bandstructure object.</p>
<p>## Using a BaseFeaturizer Class</p>
<p>There are multiple ways for running the featurize routines:</p>
<blockquote>
<div><p><cite>featurize</cite>: Featurize a single entry
<cite>featurize_many</cite>: Featurize a list of entries
<cite>featurize_dataframe</cite>: Compute features for many entries, store results</p>
<blockquote>
<div><p>as columns in a dataframe</p>
</div></blockquote>
</div></blockquote>
<p>Some featurizers require first calling the <cite>fit</cite> method before the
featurization methods can function. Generally, you pass the dataset to
fit to determine which features a featurizer should compute. For example,
a featurizer that returns the partial radial distribution function
may need to know which elements are present in a dataset.</p>
<p>You can can also use the <cite>precheck</cite> and <cite>precheck_dataframe</cite> methods to
ensure a featurizer is in scope for a given sample (or dataset) before
featurizing.</p>
<p>You can also employ the featurizer as part of a ScikitLearn Pipeline object.
For these cases, ScikitLearn calls the <cite>transform</cite> function of the
<cite>BaseFeaturizer</cite> which is a less-featured wrapper of <cite>featurize_many</cite>. You
would then provide your input data as an array to the Pipeline, which would
output the features as an array.</p>
<p>Beyond the featurizing capability, BaseFeaturizer also includes methods
for retrieving proper references for a featurizer. The <cite>citations</cite> function
returns a list of papers that should be cited. The <cite>implementors</cite> function
returns a list of people who wrote the featurizer, so that you know
who to contact with questions.</p>
<p>## Implementing a New BaseFeaturizer Class</p>
<dl>
<dt>These operations must be implemented for each new featurizer:</dt><dd><dl class="simple">
<dt><cite>featurize</cite> - Takes a single material as input, returns the features of</dt><dd><p>that material.</p>
</dd>
<dt><cite>feature_labels</cite> - Generates a human-meaningful name for each of the</dt><dd><p>features.</p>
</dd>
</dl>
<p><cite>citations</cite> - Returns a list of citations in BibTeX format
<cite>implementors</cite> - Returns a list of people who contributed to writing a</p>
<blockquote>
<div><p>paper.</p>
</div></blockquote>
</dd>
</dl>
<p>None of these operations should change the state of the featurizer. I.e.,
running each method twice should not produce different results, no class
attributes should be changed, and running one operation should not affect
the output of another.</p>
<p>All options of the featurizer must be set by the <cite>__init__</cite> function. All
options must be listed as keyword arguments with default values, and the
value must be saved as a class attribute with the same name (e.g., argument
<cite>n</cite> should be stored in <cite>self.n</cite>). These requirements are necessary for
compatibility with the <cite>get_params</cite> and <cite>set_params</cite> methods of
<cite>BaseEstimator</cite>, which enable easy interoperability with ScikitLearn</p>
<p>Depending on the complexity of your featurizer, it may be worthwhile to
implement a <cite>from_preset</cite> class method. The <cite>from_preset</cite> method takes the
name of a preset and returns an instance of the featurizer with some
hard-coded set of inputs. The <cite>from_preset</cite> option is particularly useful
for defining the settings used by papers in the literature.</p>
<p>Optionally, you can implement the <cite>fit</cite> operation if there are attributes of
your featurizer that must be set for the featurizer to work. Any variables
that are set by fitting should be stored as class attributes that end with
an underscore. (This follows the pattern used by ScikitLearn).</p>
<p>Another option to consider is whether it is worth making any utility
operations for your featurizer. <cite>featurize</cite> must return a list of features,
but this may not be the most natural representation for your features (e.g.,
a <cite>dict</cite> could be better). Making a separate function for computing features
in this natural representation and having the <cite>featurize</cite> function call this
method and then convert the data into a list is a recommended approach.
Users who want to compute the representation in the natural form can use the
utility function and users who want the data in a ML-ready format (list) can
call <cite>featurize</cite>. See <cite>PartialRadialDistributionFunction</cite> for an example of
this concept.</p>
<p>An additional factor to consider is the chunksize for data parallelisation.
For lightweight computational tasks, the overhead associated with passing
data from <cite>multiprocessing.Pool.map()</cite> to the function being parallelized
can increase the time taken for all tasks to be completed. By setting
the <cite>self._chunksize</cite> argument, the overhead associated with passing data
to the tasks can be reduced. Note that there is only an advantage to using
chunksize when the time taken to pass the data from <cite>map</cite> to the function
call is within several orders of magnitude to that of the function call
itself. By default, we allow the Python multiprocessing library to determine
the chunk size automatically based on the size of the list being featurized.
You may want to specify a small chunk size for computationally-expensive
featurizers, which will enable better distribution of tasks across threads.
In contrast, for more lightweight featurizers, it is recommended that
the implementor trial a range of chunksize values to find the optimum.
As a general rule of thumb, if the featurize function takes 0.1 seconds or
less, a chunksize of around 30 will perform best.</p>
<p>## Documenting a BaseFeaturizer</p>
<p>The class documentation for each featurizer must contain a description of
the options and the features that will be computed. The options of the class
must all be defined in the <cite>__init__</cite> function of the class, and we
recommend documenting them using the
[Google style](<a class="reference external" href="https://google.github.io/styleguide/pyguide.html">https://google.github.io/styleguide/pyguide.html</a>).</p>
<p>For auto-generated documentation purposes, the first line of the featurizer
doc should come under the class declaration (not under __init__) and should
be a one line summary of the featurizer.</p>
<p>We recommend starting the class documentation with a high-level overview of
the features. For example, mention what kind of characteristics of the
material they describe and refer the reader to a paper that describes these
features well (use a hyperlink if possible, so that the readthedocs will
link to that paper). Then, describe each of the individual features in a
block named “Features”. It is necessary here to give the user enough
information for user to map a feature name what it means. The objective in
this part is to allow people to understand what each column of their
dataframe is without having to read the Python code. You do not need to
explain all of the math/algorithms behind each feature for them to be able
to reproduce the feature, just to get an idea what it is.</p>
<dl class="py method">
<dt id="matminer.featurizers.base.BaseFeaturizer.chunksize">
<em class="property">property </em><code class="sig-name descname">chunksize</code><a class="headerlink" href="#matminer.featurizers.base.BaseFeaturizer.chunksize" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.BaseFeaturizer.citations">
<em class="property">abstract </em><code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.BaseFeaturizer.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.BaseFeaturizer.feature_labels">
<em class="property">abstract </em><code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.BaseFeaturizer.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.BaseFeaturizer.featurize">
<em class="property">abstract </em><code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">x</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.BaseFeaturizer.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, which has to be implemented
in any derived featurizer subclass.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).</p>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.BaseFeaturizer.featurize_dataframe">
<code class="sig-name descname">featurize_dataframe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">df</span></em>, <em class="sig-param"><span class="n">col_id</span></em>, <em class="sig-param"><span class="n">ignore_errors</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">return_errors</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">multiindex</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">pbar</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.BaseFeaturizer.featurize_dataframe" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute features for all entries contained in input dataframe.</p>
<dl>
<dt>Args:</dt><dd><p>df (Pandas dataframe): Dataframe containing input data.
col_id (str or list of str): column label containing objects to</p>
<blockquote>
<div><p>featurize. Can be multiple labels if the featurize function
requires multiple inputs.</p>
</div></blockquote>
<dl class="simple">
<dt>ignore_errors (bool): Returns NaN for dataframe rows where</dt><dd><p>exceptions are thrown if True. If False, exceptions
are thrown as normal.</p>
</dd>
<dt>return_errors (bool). Returns the errors encountered for each</dt><dd><p>row in a separate <cite>XFeaturizer errors</cite> column if True. Requires
ignore_errors to be True.</p>
</dd>
<dt>inplace (bool): If True, adds columns to the original object in</dt><dd><p>memory and returns None. Else, returns the updated object.
Should be identical to pandas inplace behavior.</p>
</dd>
<dt>multiindex (bool): If True, use a Featurizer - Feature 2-level</dt><dd><p>index using the MultiIndex capabilities of pandas. If done
inplace, multiindex featurization will overwrite the original
dataframe’s column index.</p>
</dd>
</dl>
<p>pbar (bool): Shows a progress bar if True.</p>
</dd>
<dt>Returns:</dt><dd><p>updated dataframe.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.BaseFeaturizer.featurize_many">
<code class="sig-name descname">featurize_many</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">entries</span></em>, <em class="sig-param"><span class="n">ignore_errors</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">return_errors</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">pbar</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.BaseFeaturizer.featurize_many" title="Permalink to this definition">¶</a></dt>
<dd><p>Featurize a list of entries.</p>
<p>If <cite>featurize</cite> takes multiple inputs, supply inputs as a list of tuples.</p>
<p>Featurize_many supports entries as a list, tuple, numpy array,
Pandas Series, or Pandas DataFrame.</p>
<dl>
<dt>Args:</dt><dd><p>entries (list-like object): A list of entries to be featurized.
ignore_errors (bool): Returns NaN for entries where exceptions are</p>
<blockquote>
<div><p>thrown if True. If False, exceptions are thrown as normal.</p>
</div></blockquote>
<dl class="simple">
<dt>return_errors (bool): If True, returns the feature list as</dt><dd><p>determined by ignore_errors with traceback strings added
as an extra ‘feature’. Entries which featurize without
exceptions have this extra feature set to NaN.</p>
</dd>
</dl>
<p>pbar (bool): Show a progress bar for featurization if True.</p>
</dd>
<dt>Returns:</dt><dd><p>(list) features for each entry.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.BaseFeaturizer.featurize_wrapper">
<code class="sig-name descname">featurize_wrapper</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">x</span></em>, <em class="sig-param"><span class="n">return_errors</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">ignore_errors</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.BaseFeaturizer.featurize_wrapper" title="Permalink to this definition">¶</a></dt>
<dd><p>An exception wrapper for featurize, used in featurize_many and
featurize_dataframe. featurize_wrapper changes the behavior of featurize
when ignore_errors is True in featurize_many/dataframe.</p>
<dl>
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).
ignore_errors (bool): Returns NaN for entries where exceptions are</p>
<blockquote>
<div><p>thrown if True. If False, exceptions are thrown as normal.</p>
</div></blockquote>
<dl class="simple">
<dt>return_errors (bool): If True, returns the feature list as</dt><dd><p>determined by ignore_errors with traceback strings added
as an extra ‘feature’. Entries which featurize without
exceptions have this extra feature set to NaN.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.BaseFeaturizer.fit">
<code class="sig-name descname">fit</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">X</span></em>, <em class="sig-param"><span class="n">y</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">fit_kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.BaseFeaturizer.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Update the parameters of this featurizer based on available data</p>
<dl class="simple">
<dt>Args:</dt><dd><p>X - [list of tuples], training data</p>
</dd>
<dt>Returns:</dt><dd><p>self</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.BaseFeaturizer.fit_featurize_dataframe">
<code class="sig-name descname">fit_featurize_dataframe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">df</span></em>, <em class="sig-param"><span class="n">col_id</span></em>, <em class="sig-param"><span class="n">fit_args</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">*</span><span class="n">args</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.BaseFeaturizer.fit_featurize_dataframe" title="Permalink to this definition">¶</a></dt>
<dd><p>The dataframe equivalent of fit_transform. Takes a dataframe and
column id as input, fits the featurizer to that dataframe, and
returns a featurized dataframe. Accepts the same arguments as
featurize_dataframe.</p>
<dl>
<dt>Args:</dt><dd><p>df (Pandas dataframe): Dataframe containing input data.
col_id (str or list of str): column label containing objects to</p>
<blockquote>
<div><p>featurize. Can be multiple labels if the featurize function
requires multiple inputs.</p>
</div></blockquote>
<p>fit_args (list): list of arguments for fit function.</p>
</dd>
<dt>Returns:</dt><dd><p>updated dataframe based on featurizer fitted to that dataframe.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.BaseFeaturizer.implementors">
<em class="property">abstract </em><code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.BaseFeaturizer.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.BaseFeaturizer.n_jobs">
<em class="property">property </em><code class="sig-name descname">n_jobs</code><a class="headerlink" href="#matminer.featurizers.base.BaseFeaturizer.n_jobs" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.BaseFeaturizer.precheck">
<code class="sig-name descname">precheck</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">x</span></em><span class="sig-paren">)</span> &#x2192; <a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a><a class="headerlink" href="#matminer.featurizers.base.BaseFeaturizer.precheck" title="Permalink to this definition">¶</a></dt>
<dd><p>Precheck (provide an estimate of whether a featurizer will work or not)
for a single entry (e.g., a single composition). If the entry fails the
precheck, it will most likely fail featurization; if it passes, it is
likely (but not guaranteed) to featurize correctly.</p>
<dl class="simple">
<dt>Prechecks should be:</dt><dd><ul class="simple">
<li><p>accurate (but can be good estimates rather than ground truth)</p></li>
<li><p>fast to evaluate</p></li>
<li><dl class="simple">
<dt>unlikely to be obsolete via changes in the featurizer in the near</dt><dd><p>future</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
<p>This method should be overridden by any featurizer requiring its
use, as by default all entries will pass prechecking. Also, precheck
is a good opportunity to throw warnings about long runtimes (e.g., doing
nearest neighbors computations on a structure with many thousand sites).</p>
<p>See the documentation for precheck_dataframe for more information.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><a href="#id1"><span class="problematic" id="id2">*</span></a>x (Composition, Structure, etc.): Input to-be-featurized. Can be</dt><dd><p>a single input or multiple inputs.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>(bool): True, if passes the precheck. False, if fails.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.BaseFeaturizer.precheck_dataframe">
<code class="sig-name descname">precheck_dataframe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">df</span></em>, <em class="sig-param"><span class="n">col_id</span></em>, <em class="sig-param"><span class="n">return_frac</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">inplace</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span> &#x2192; [&lt;class ‘float’&gt;, &lt;class ‘pandas.core.frame.DataFrame’&gt;]<a class="headerlink" href="#matminer.featurizers.base.BaseFeaturizer.precheck_dataframe" title="Permalink to this definition">¶</a></dt>
<dd><p>Precheck an entire dataframe. Subclasses wanting to use precheck
functionality should not override this method, they should override
precheck (unless the entire df determines whether single entries pass
or fail a precheck).</p>
<p>Prechecking should be a quick and useful way to check that for a
particular dataframe (set of featurizer inputs), the featurizer is:</p>
<blockquote>
<div><ol class="arabic simple">
<li><p>in scope, and/or…</p></li>
<li><p>robust to errors and/or…</p></li>
<li><dl class="simple">
<dt>any other reason you would not practically want to use this</dt><dd><p>featurizer in on this dataframe.</p>
</dd>
</dl>
</li>
</ol>
</div></blockquote>
<p>By prechecking before featurizing, you can avoid applying featurizers
to data that will ultimately fail, return unreliable numbers, or
are out of scope. Prechecking is also a good time to throw/observe
warnings (such as long runtime warnings!).</p>
<dl>
<dt>Args:</dt><dd><p>df (pd.DataFrame): A dataframe
col_id (str or [str]): column label containing objects to featurize.</p>
<blockquote>
<div><p>Can be multiple labels if the featurize function requires
multiple inputs.</p>
</div></blockquote>
<dl class="simple">
<dt>return_frac (bool): If True, returns the fraction of entries</dt><dd><p>passing the precheck (e.g., 0.5). Else, returns a dataframe.</p>
</dd>
<dt>inplace (bool); Only relevant if return_frac=False. If inplace=True,</dt><dd><p>the input dataframe is modified in memory with a boolean column
for precheck. Otherwise, a new df with this column is returned.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>(bool, pd.DataFrame): If return_frac=True, returns the fraction of</dt><dd><p>entries passing the precheck. Else, returns the dataframe with
an extra boolean column added for the precheck.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.BaseFeaturizer.set_chunksize">
<code class="sig-name descname">set_chunksize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">chunksize</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.BaseFeaturizer.set_chunksize" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the chunksize used for Pool.map parallelisation.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.BaseFeaturizer.set_n_jobs">
<code class="sig-name descname">set_n_jobs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">n_jobs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.BaseFeaturizer.set_n_jobs" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the number of threads for this.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.BaseFeaturizer.transform">
<code class="sig-name descname">transform</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">X</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.BaseFeaturizer.transform" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute features for a list of inputs</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.base.MultipleFeaturizer">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.base.</code><code class="sig-name descname">MultipleFeaturizer</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">featurizers</span></em>, <em class="sig-param"><span class="n">iterate_over_entries</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.MultipleFeaturizer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Class to run multiple featurizers on the same input data.</p>
<p>All featurizers must take the same kind of data as input
to the featurize function.</p>
<dl>
<dt>Args:</dt><dd><p>featurizers (list of BaseFeaturizer): A list of featurizers to run.
iterate_over_entries (bool): Whether to iterate over the entries or</p>
<blockquote>
<div><p>featurizers. Iterating over entries will enable increased caching
but will only display a single progress bar for all featurizers.
If set to False, iteration will be performed over featurizers,
resulting in reduced caching but individual progress bars for each
featurizer.</p>
</div></blockquote>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.base.MultipleFeaturizer.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">featurizers</span></em>, <em class="sig-param"><span class="n">iterate_over_entries</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.MultipleFeaturizer.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.MultipleFeaturizer.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.MultipleFeaturizer.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.MultipleFeaturizer.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.MultipleFeaturizer.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.MultipleFeaturizer.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">x</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.MultipleFeaturizer.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, which has to be implemented
in any derived featurizer subclass.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).</p>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.MultipleFeaturizer.featurize_many">
<code class="sig-name descname">featurize_many</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">entries</span></em>, <em class="sig-param"><span class="n">ignore_errors</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">return_errors</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">pbar</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.MultipleFeaturizer.featurize_many" title="Permalink to this definition">¶</a></dt>
<dd><p>Featurize a list of entries.</p>
<p>If <cite>featurize</cite> takes multiple inputs, supply inputs as a list of tuples.</p>
<p>Featurize_many supports entries as a list, tuple, numpy array,
Pandas Series, or Pandas DataFrame.</p>
<dl>
<dt>Args:</dt><dd><p>entries (list-like object): A list of entries to be featurized.
ignore_errors (bool): Returns NaN for entries where exceptions are</p>
<blockquote>
<div><p>thrown if True. If False, exceptions are thrown as normal.</p>
</div></blockquote>
<dl class="simple">
<dt>return_errors (bool): If True, returns the feature list as</dt><dd><p>determined by ignore_errors with traceback strings added
as an extra ‘feature’. Entries which featurize without
exceptions have this extra feature set to NaN.</p>
</dd>
</dl>
<p>pbar (bool): Show a progress bar for featurization if True.</p>
</dd>
<dt>Returns:</dt><dd><p>(list) features for each entry.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.MultipleFeaturizer.featurize_wrapper">
<code class="sig-name descname">featurize_wrapper</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">x</span></em>, <em class="sig-param"><span class="n">return_errors</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">ignore_errors</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.MultipleFeaturizer.featurize_wrapper" title="Permalink to this definition">¶</a></dt>
<dd><p>An exception wrapper for featurize, used in featurize_many and
featurize_dataframe. featurize_wrapper changes the behavior of featurize
when ignore_errors is True in featurize_many/dataframe.</p>
<dl>
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).
ignore_errors (bool): Returns NaN for entries where exceptions are</p>
<blockquote>
<div><p>thrown if True. If False, exceptions are thrown as normal.</p>
</div></blockquote>
<dl class="simple">
<dt>return_errors (bool): If True, returns the feature list as</dt><dd><p>determined by ignore_errors with traceback strings added
as an extra ‘feature’. Entries which featurize without
exceptions have this extra feature set to NaN.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.MultipleFeaturizer.fit">
<code class="sig-name descname">fit</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">X</span></em>, <em class="sig-param"><span class="n">y</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">fit_kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.MultipleFeaturizer.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Update the parameters of this featurizer based on available data</p>
<dl class="simple">
<dt>Args:</dt><dd><p>X - [list of tuples], training data</p>
</dd>
<dt>Returns:</dt><dd><p>self</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.MultipleFeaturizer.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.MultipleFeaturizer.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.MultipleFeaturizer.set_n_jobs">
<code class="sig-name descname">set_n_jobs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">n_jobs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.MultipleFeaturizer.set_n_jobs" title="Permalink to this definition">¶</a></dt>
<dd><p>Set the number of threads for this.</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.base.StackedFeaturizer">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.base.</code><code class="sig-name descname">StackedFeaturizer</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">featurizer</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">model</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">name</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">class_names</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.StackedFeaturizer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Use the output of a machine learning model as features</p>
<p>For regression models, we use the single output class.</p>
<p>For classification models, we use the probability for the first N-1 classes where N is the
number of classes.</p>
<dl class="py method">
<dt id="matminer.featurizers.base.StackedFeaturizer.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">featurizer</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">model</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">name</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">class_names</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.StackedFeaturizer.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize featurizer</p>
<dl>
<dt>Args:</dt><dd><p>featurizer (BaseFeaturizer): Featurizer used to generate inputs to the model
model (BaseEstimator): Fitted machine learning model to be evaluated
name (str): [Optional] name of model, used when creating feature names</p>
<blockquote>
<div><p>class_names ([str]): Required for classification models, used when creating
feature names (scikit-learn does not specify the number of classes for
a classifier). Class names must be in the same order as the classes in the model
(e.g., class_names[0] must be the name of the class 0)</p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.StackedFeaturizer.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.StackedFeaturizer.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.StackedFeaturizer.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.StackedFeaturizer.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.StackedFeaturizer.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">x</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.StackedFeaturizer.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, which has to be implemented
in any derived featurizer subclass.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).</p>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.base.StackedFeaturizer.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.base.StackedFeaturizer.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-matminer.featurizers.composition">
<span id="matminer-featurizers-composition-module"></span><h2>matminer.featurizers.composition module<a class="headerlink" href="#module-matminer.featurizers.composition" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt id="matminer.featurizers.composition.AtomicOrbitals">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.composition.</code><code class="sig-name descname">AtomicOrbitals</code><a class="headerlink" href="#matminer.featurizers.composition.AtomicOrbitals" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Determine HOMO/LUMO features based on a composition.</p>
<p>The highest occupied molecular orbital (HOMO) and lowest unoccupied
molecular orbital (LUMO) are estiated from the atomic orbital energies
of the composition. The atomic orbital energies are from NIST:
<a class="reference external" href="https://www.nist.gov/pml/data/atomic-reference-data-electronic-structure-calculations">https://www.nist.gov/pml/data/atomic-reference-data-electronic-structure-calculations</a></p>
<p>Warning:
For compositions with inter-species fractions greater than 10,000 (e.g.
dilute alloys such as FeC0.00001) the composition will be truncated (to Fe
in this example). In such extreme cases, the truncation likely reflects the
true physics of the situation (i.e. that the dilute element does not
significantly contribute orbital character to the band structure), but the
user should be aware of this behavior.</p>
<dl class="py method">
<dt id="matminer.featurizers.composition.AtomicOrbitals.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.AtomicOrbitals.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.AtomicOrbitals.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.AtomicOrbitals.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.AtomicOrbitals.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.AtomicOrbitals.featurize" title="Permalink to this definition">¶</a></dt>
<dd><dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>comp: (Composition)</dt><dd><p>pymatgen Composition object</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>HOMO_character: (str) orbital symbol (‘s’, ‘p’, ‘d’, or ‘f’)
HOMO_element: (str) symbol of element for HOMO
HOMO_energy: (float in eV) absolute energy of HOMO
LUMO_character: (str) orbital symbol (‘s’, ‘p’, ‘d’, or ‘f’)
LUMO_element: (str) symbol of element for LUMO
LUMO_energy: (float in eV) absolute energy of LUMO
gap_AO: (float in eV)</p>
<blockquote>
<div><p>the estimated bandgap from HOMO and LUMO energeis</p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.AtomicOrbitals.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.AtomicOrbitals.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.composition.AtomicPackingEfficiency">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.composition.</code><code class="sig-name descname">AtomicPackingEfficiency</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">threshold</span><span class="o">=</span><span class="default_value">0.01</span></em>, <em class="sig-param"><span class="n">n_nearest</span><span class="o">=</span><span class="default_value">1, 3, 5</span></em>, <em class="sig-param"><span class="n">max_types</span><span class="o">=</span><span class="default_value">6</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.AtomicPackingEfficiency" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Packing efficiency based on a geometric theory of the amorphous packing
of hard spheres.</p>
<p>This featurizer computes two different kinds of the features. The first
relate to the distance between a composition and the composition of
the clusters of atoms expected to be efficiently packed based on a
theory from
<a href="#id36"><span class="problematic" id="id37">`Laws et al.&lt;http://www.nature.com/doifinder/10.1038/ncomms9123&gt;`_</span></a>.
The second corresponds to the packing efficiency of a system if all atoms
in the alloy are simultaneously as efficiently-packed as possible.</p>
<p>The packing efficiency in these models is based on the Atomic Packing
Efficiency (APE), which measures the difference between the ratio of
the radii of the central atom to its neighbors and the ideal ratio
of a cluster with the same number of atoms that has optimal packing
efficiency. If the difference between the ratios is too large, the APE is
positive. If the difference is too small, the APE is negative.</p>
<dl class="simple">
<dt>Features:</dt><dd><dl class="simple">
<dt>dist from {k} clusters <a href="#id32"><span class="problematic" id="id33">|APE|</span></a> &lt; {thr} - The distance between an</dt><dd><p>alloy composition and the k clusters that have a packing efficiency
below thr from ideal</p>
</dd>
<dt>mean simul. packing efficiency - Mean packing efficiency of all atoms.</dt><dd><p>The packing efficiency is measured with respect to ideal (0)</p>
</dd>
<dt>mean abs simul. packing efficiency - Mean absolute value of the</dt><dd><p>packing efficiencies. Closer to zero is more efficiently packed</p>
</dd>
</dl>
</dd>
<dt>References:</dt><dd><p>[1] K.J. Laws, D.B. Miracle, M. Ferry, A predictive structural model
for bulk metallic glasses, Nat. Commun. 6 (2015) 8123. doi:10.1038/ncomms9123.</p>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.composition.AtomicPackingEfficiency.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">threshold</span><span class="o">=</span><span class="default_value">0.01</span></em>, <em class="sig-param"><span class="n">n_nearest</span><span class="o">=</span><span class="default_value">1, 3, 5</span></em>, <em class="sig-param"><span class="n">max_types</span><span class="o">=</span><span class="default_value">6</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.AtomicPackingEfficiency.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the featurizer</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>threshold (float):Threshold to use for determining whether</dt><dd><p>a cluster is efficiently packed.</p>
</dd>
</dl>
<p>n_nearest ({int}): Number of nearest clusters to use when considering features
max_types (int): Maximum number of atom types to consider when</p>
<blockquote>
<div><p>looking for efficient clusters. The process for finding
efficient clusters very expensive for large numbers of types</p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.AtomicPackingEfficiency.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.AtomicPackingEfficiency.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.AtomicPackingEfficiency.compute_nearest_cluster_distance">
<code class="sig-name descname">compute_nearest_cluster_distance</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.AtomicPackingEfficiency.compute_nearest_cluster_distance" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the distance between a composition and that the nearest
efficiently-packed clusters.</p>
<p>Measures the mean <img class="math" src="_images/math/fcd0c45a70da7d2399b3fb666759b02c5edd251e.png" alt="L_2"/> distance between the alloy composition
and the <img class="math" src="_images/math/9630132210b904754c9ab272b61cb527d12263ca.png" alt="k"/>-nearest clusters with Atomic Packing Efficiencies
within the user-specified tolerance of 1. <img class="math" src="_images/math/9630132210b904754c9ab272b61cb527d12263ca.png" alt="k"/> is any of the
numbers defined in the “n_nearest” parameter of this class.</p>
<p>If there are less than <cite>k</cite> efficient clusters in the system, we use
the maximum distance betweeen any two compositions (1) for the
unmatched neighbors.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>comp (Composition): Composition of material to evaluate</p>
</dd>
<dt>Return:</dt><dd><p>[float] Average distances</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.AtomicPackingEfficiency.compute_simultaneous_packing_efficiency">
<code class="sig-name descname">compute_simultaneous_packing_efficiency</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.AtomicPackingEfficiency.compute_simultaneous_packing_efficiency" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the packing efficiency of the system when the neighbor
shell of each atom has the same composition as the alloy. When this
criterion is satisfied, it is possible for every atom in this system
to be simultaneously as efficiently-packed as possible.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>comp (Composition): Composition to be assessed</p>
</dd>
<dt>Returns</dt><dd><p>(float) Average APE of all atoms
(float) Average deviation of the APE of each atom from ideal (0)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.AtomicPackingEfficiency.create_cluster_lookup_tool">
<code class="sig-name descname">create_cluster_lookup_tool</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">elements</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.AtomicPackingEfficiency.create_cluster_lookup_tool" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the compositions of efficiently-packed clusters in a certain system
of elements</p>
<dl class="simple">
<dt>Args:</dt><dd><p>elements ([Element]): Elements in system</p>
</dd>
<dt>Return:</dt><dd><dl class="simple">
<dt>(NearNeighbors): Tool to find nearby clusters in this system. None</dt><dd><p>if there are no efficiently-packed clusters for this combination of elements</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.AtomicPackingEfficiency.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.AtomicPackingEfficiency.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.AtomicPackingEfficiency.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.AtomicPackingEfficiency.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, which has to be implemented
in any derived featurizer subclass.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).</p>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.AtomicPackingEfficiency.find_ideal_cluster_size">
<code class="sig-name descname">find_ideal_cluster_size</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">radius_ratio</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.AtomicPackingEfficiency.find_ideal_cluster_size" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the optimal cluster size for a certain radius ratio</p>
<p>Finds the number of nearest neighbors <img class="math" src="_images/math/5a939c5280da7202ca4531f175a7780ad5e1f80a.png" alt="n"/> that minimizes
<img class="math" src="_images/math/66343a54ffb585a31a82fe441ee9fc3cd8eb4368.png" alt="|1 - rp(n)/r|"/>, where <img class="math" src="_images/math/abdaecd41a1558846f03b71ab3d99436de6faf0d.png" alt="rp(n)"/> is the ideal radius
ratio for a certain <img class="math" src="_images/math/5a939c5280da7202ca4531f175a7780ad5e1f80a.png" alt="n"/> and <img class="math" src="_images/math/79a3d439d28652c547386f39b555d90d3aaf102d.png" alt="r"/> is the actual ratio.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>radius_ratio (float): <img class="math" src="_images/math/879aee5829f6469733f819b01853c5205ecc88b6.png" alt="r / r_{neighbor}"/></p>
</dd>
<dt>Returns:</dt><dd><p>(int) number of neighboring atoms for that will be the most
efficiently packed.
(float) Optimal APE</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.AtomicPackingEfficiency.get_ideal_radius_ratio">
<code class="sig-name descname">get_ideal_radius_ratio</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">n_neighbors</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.AtomicPackingEfficiency.get_ideal_radius_ratio" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the idea ratio between the central atom and neighboring
atoms for a neighbor with a certain number of nearest neighbors.</p>
<p>Based on work by <a class="reference external" href="https://www.jstage.jst.go.jp/article/matertrans/47/7/47_7_1737/_article/-char/en">Miracle, Lord, and Ranganathan</a>.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>n_neighbors (int): Number of atoms in 1st NN shell</p>
</dd>
<dt>Return:</dt><dd><p>(float) ideal radius ratio <img class="math" src="_images/math/879aee5829f6469733f819b01853c5205ecc88b6.png" alt="r / r_{neighbor}"/></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.AtomicPackingEfficiency.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.AtomicPackingEfficiency.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.composition.BandCenter">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.composition.</code><code class="sig-name descname">BandCenter</code><a class="headerlink" href="#matminer.featurizers.composition.BandCenter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Estimation of absolute position of band center using electronegativity.</p>
<dl class="simple">
<dt>Features</dt><dd><ul class="simple">
<li><p>Band center</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.composition.BandCenter.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.BandCenter.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.BandCenter.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.BandCenter.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.BandCenter.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.BandCenter.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>(Rough) estimation of absolution position of band center using
geometric mean of electronegativity.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>comp (Composition).</p>
</dd>
<dt>Returns:</dt><dd><p>(float) band center.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.BandCenter.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.BandCenter.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.composition.CationProperty">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.composition.</code><code class="sig-name descname">CationProperty</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data_source</span></em>, <em class="sig-param"><span class="n">features</span></em>, <em class="sig-param"><span class="n">stats</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.CationProperty" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.composition.ElementProperty" title="matminer.featurizers.composition.ElementProperty"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.composition.ElementProperty</span></code></a></p>
<p>Features based on properties of cations in a material</p>
<p>Requires that oxidation states have already been determined. Property
statistics weighted by composition.</p>
<p>Features: Based on the statistics of the data_source chosen, computed
by element stoichiometry. The format generally is:</p>
<p>“{data source} {statistic} {property}”</p>
<p>For example:</p>
<p>“DemlData range magn_moment” # Range of magnetic moment via Deml et al. data</p>
<p>For a list of all statistics, see the PropertyStats documentation; for a
list of all attributes available for a given data_source, see the
documentation for the data sources (e.g., PymatgenData, MagpieData,
MatscholarElementData, etc.).</p>
<dl class="py method">
<dt id="matminer.featurizers.composition.CationProperty.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.CationProperty.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.CationProperty.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.CationProperty.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.CationProperty.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.CationProperty.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get elemental property attributes</p>
<dl class="simple">
<dt>Args:</dt><dd><p>comp: Pymatgen composition object</p>
</dd>
<dt>Returns:</dt><dd><p>all_attributes: Specified property statistics of features</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.CationProperty.from_preset">
<em class="property">classmethod </em><code class="sig-name descname">from_preset</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">preset_name</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.CationProperty.from_preset" title="Permalink to this definition">¶</a></dt>
<dd><p>Return ElementProperty from a preset string
Args:</p>
<blockquote>
<div><dl class="simple">
<dt>preset_name: (str) can be one of “magpie”, “deml”, “matminer”,</dt><dd><p>“matscholar_el”, or “megnet_el”.</p>
</dd>
</dl>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>ElementProperty based on the preset name.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.composition.CohesiveEnergy">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.composition.</code><code class="sig-name descname">CohesiveEnergy</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">mapi_key</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.CohesiveEnergy" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Cohesive energy per atom using elemental cohesive energies and
formation energy.</p>
<p>Get cohesive energy per atom of a compound by adding known
elemental cohesive energies from the formation energy of the
compound.</p>
<dl class="simple">
<dt>Parameters:</dt><dd><dl class="simple">
<dt>mapi_key (str): Materials API key for looking up formation energy</dt><dd><p>by composition alone (if you don’t set the formation energy
yourself).</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.composition.CohesiveEnergy.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">mapi_key</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.CohesiveEnergy.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.CohesiveEnergy.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.CohesiveEnergy.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.CohesiveEnergy.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.CohesiveEnergy.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.CohesiveEnergy.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em>, <em class="sig-param"><span class="n">formation_energy_per_atom</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.CohesiveEnergy.featurize" title="Permalink to this definition">¶</a></dt>
<dd><dl>
<dt>Args:</dt><dd><p>comp: (str) compound composition, eg: “NaCl”
formation_energy_per_atom: (float) the formation energy per atom of</p>
<blockquote>
<div><p>your compound. If not set, will look up the most stable
formation energy from the Materials Project database.</p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.CohesiveEnergy.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.CohesiveEnergy.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.composition.CohesiveEnergyMP">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.composition.</code><code class="sig-name descname">CohesiveEnergyMP</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">mapi_key</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.CohesiveEnergyMP" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Cohesive energy per atom lookup using Materials Project</p>
<dl class="simple">
<dt>Parameters:</dt><dd><dl class="simple">
<dt>mapi_key (str): Materials API key for looking up cohesive energy</dt><dd><p>by composition alone.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.composition.CohesiveEnergyMP.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">mapi_key</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.CohesiveEnergyMP.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.CohesiveEnergyMP.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.CohesiveEnergyMP.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.CohesiveEnergyMP.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.CohesiveEnergyMP.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.CohesiveEnergyMP.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.CohesiveEnergyMP.featurize" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Args:</dt><dd><p>comp: (str) compound composition, eg: “NaCl”</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.CohesiveEnergyMP.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.CohesiveEnergyMP.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.composition.ElectronAffinity">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.composition.</code><code class="sig-name descname">ElectronAffinity</code><a class="headerlink" href="#matminer.featurizers.composition.ElectronAffinity" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Calculate average electron affinity times formal charge of anion elements.
Note: The formal charges must already be computed before calling <cite>featurize</cite>.
Generates average (electron affinity*formal charge) of anions.</p>
<dl class="py method">
<dt id="matminer.featurizers.composition.ElectronAffinity.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElectronAffinity.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.ElectronAffinity.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElectronAffinity.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.ElectronAffinity.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElectronAffinity.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.ElectronAffinity.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElectronAffinity.featurize" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Args:</dt><dd><p>comp: (Composition) Composition to be featurized</p>
</dd>
<dt>Returns:</dt><dd><p>avg_anion_affin (single-element list): average electron affinity*formal charge of anions</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.ElectronAffinity.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElectronAffinity.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.composition.ElectronegativityDiff">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.composition.</code><code class="sig-name descname">ElectronegativityDiff</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">stats</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElectronegativityDiff" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Features from electronegativity differences between anions and cations.</p>
<p>These features are computed by first determining the concentration-weighted
average electronegativity of the anions. For example, the average
electronegativity of the anions in CaCoSO is equal to 1/2 of that of S and 1/2 of that of O.
We then compute the difference between the electronegativity of each cation
and the average anion electronegativity.</p>
<p>The feature values are then determined based on the concentration-weighted statistics
in the same manner as ElementProperty features. For example, one value could be
the mean electronegativity difference over all the anions.</p>
<dl class="simple">
<dt>Parameters:</dt><dd><p>data_source (data class): source from which to retrieve element data
stats: Property statistics to compute</p>
</dd>
</dl>
<p>Generates average electronegativity difference between cations and anions</p>
<dl class="py method">
<dt id="matminer.featurizers.composition.ElectronegativityDiff.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">stats</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElectronegativityDiff.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.ElectronegativityDiff.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElectronegativityDiff.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.ElectronegativityDiff.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElectronegativityDiff.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.ElectronegativityDiff.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElectronegativityDiff.featurize" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Args:</dt><dd><p>comp: Pymatgen Composition object</p>
</dd>
<dt>Returns:</dt><dd><p>en_diff_stats (list of floats): Property stats of electronegativity difference</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.ElectronegativityDiff.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElectronegativityDiff.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.composition.ElementFraction">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.composition.</code><code class="sig-name descname">ElementFraction</code><a class="headerlink" href="#matminer.featurizers.composition.ElementFraction" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Class to calculate the atomic fraction of each element in a composition.</p>
<p>Generates a vector where each index represents an element in atomic number order.</p>
<dl class="py method">
<dt id="matminer.featurizers.composition.ElementFraction.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElementFraction.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.ElementFraction.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElementFraction.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.ElementFraction.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElementFraction.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.ElementFraction.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElementFraction.featurize" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Args:</dt><dd><p>comp: Pymatgen Composition object</p>
</dd>
<dt>Returns:</dt><dd><p>vector (list of floats): fraction of each element in a composition</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.ElementFraction.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElementFraction.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.composition.ElementProperty">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.composition.</code><code class="sig-name descname">ElementProperty</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data_source</span></em>, <em class="sig-param"><span class="n">features</span></em>, <em class="sig-param"><span class="n">stats</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElementProperty" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Class to calculate elemental property attributes.</p>
<p>To initialize quickly, use the from_preset() method.</p>
<p>Features: Based on the statistics of the data_source chosen, computed
by element stoichiometry. The format generally is:</p>
<p>“{data source} {statistic} {property}”</p>
<p>For example:</p>
<p>“PymetgenData range X”  # Range of electronegativity from Pymatgen data</p>
<p>For a list of all statistics, see the PropertyStats documentation; for a
list of all attributes available for a given data_source, see the
documentation for the data sources (e.g., PymatgenData, MagpieData,
MatscholarElementData, etc.).</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>data_source (AbstractData or str): source from which to retrieve</dt><dd><p>element property data (or use str for preset: “pymatgen”,
“magpie”, or “deml”)</p>
</dd>
<dt>features (list of strings): List of elemental properties to use</dt><dd><p>(these must be supported by data_source)</p>
</dd>
<dt>stats (list of strings): a list of weighted statistics to compute to for each</dt><dd><p>property (see PropertyStats for available stats)</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.composition.ElementProperty.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data_source</span></em>, <em class="sig-param"><span class="n">features</span></em>, <em class="sig-param"><span class="n">stats</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElementProperty.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.ElementProperty.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElementProperty.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.ElementProperty.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElementProperty.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.ElementProperty.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElementProperty.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get elemental property attributes</p>
<dl class="simple">
<dt>Args:</dt><dd><p>comp: Pymatgen composition object</p>
</dd>
<dt>Returns:</dt><dd><p>all_attributes: Specified property statistics of features</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.ElementProperty.from_preset">
<em class="property">classmethod </em><code class="sig-name descname">from_preset</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">preset_name</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElementProperty.from_preset" title="Permalink to this definition">¶</a></dt>
<dd><p>Return ElementProperty from a preset string
Args:</p>
<blockquote>
<div><dl class="simple">
<dt>preset_name: (str) can be one of “magpie”, “deml”, “matminer”,</dt><dd><p>“matscholar_el”, or “megnet_el”.</p>
</dd>
</dl>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>ElementProperty based on the preset name.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.ElementProperty.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ElementProperty.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.composition.IonProperty">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.composition.</code><code class="sig-name descname">IonProperty</code><span class="sig-paren">(</span><em class="sig-param">data_source=&lt;matminer.utils.data.PymatgenData object&gt;</em>, <em class="sig-param">fast=False</em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.IonProperty" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Ionic property attributes. Similar to ElementProperty.</p>
<dl class="py method">
<dt id="matminer.featurizers.composition.IonProperty.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">data_source=&lt;matminer.utils.data.PymatgenData object&gt;</em>, <em class="sig-param">fast=False</em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.IonProperty.__init__" title="Permalink to this definition">¶</a></dt>
<dd><dl>
<dt>Args:</dt><dd><blockquote>
<div><dl class="simple">
<dt>data_source - (OxidationStateMixin) - A AbstractData class that supports</dt><dd><p>the <cite>get_oxidation_state</cite> method.</p>
</dd>
</dl>
</div></blockquote>
<dl class="simple">
<dt>fast - (boolean) whether to assume elements exist in a single oxidation state,</dt><dd><p>which can dramatically accelerate the calculation of whether an ionic compound
is possible, but will miss heterovalent compounds like Fe3O4.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.IonProperty.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.IonProperty.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.IonProperty.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.IonProperty.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.IonProperty.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.IonProperty.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Ionic character attributes</p>
<dl class="simple">
<dt>Args:</dt><dd><p>comp: (Composition) Composition to be featurized</p>
</dd>
<dt>Returns:</dt><dd><p>cpd_possible (bool): Indicates if a neutral ionic compound is possible
max_ionic_char (float): Maximum ionic character between two atoms
avg_ionic_char (float): Average ionic character</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.IonProperty.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.IonProperty.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.composition.Meredig">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.composition.</code><code class="sig-name descname">Meredig</code><a class="headerlink" href="#matminer.featurizers.composition.Meredig" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Class to calculate features as defined in Meredig et. al.</p>
<dl>
<dt>Features:</dt><dd><p>Atomic fraction of each of the first 103 elements, in order of atomic number.
17 statistics of elemental properties;</p>
<blockquote>
<div><p>Mean atomic weight of constituent elements
Mean periodic table row and column number
Mean and range of atomic number
Mean and range of atomic radius
Mean and range of electronegativity
Mean number of valence electrons in each orbital
Fraction of total valence electrons in each orbital</p>
</div></blockquote>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.composition.Meredig.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.Meredig.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.Meredig.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.Meredig.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.Meredig.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.Meredig.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.Meredig.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.Meredig.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get elemental property attributes</p>
<dl class="simple">
<dt>Args:</dt><dd><p>comp: Pymatgen composition object</p>
</dd>
<dt>Returns:</dt><dd><p>all_attributes: Specified property statistics of features</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.Meredig.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.Meredig.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.composition.Miedema">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.composition.</code><code class="sig-name descname">Miedema</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">struct_types</span><span class="o">=</span><span class="default_value">'all'</span></em>, <em class="sig-param"><span class="n">ss_types</span><span class="o">=</span><span class="default_value">'min'</span></em>, <em class="sig-param"><span class="n">data_source</span><span class="o">=</span><span class="default_value">'Miedema'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.Miedema" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Formation enthalpies of intermetallic compounds, from Miedema et al.</p>
<p>Calculate the formation enthalpies of the intermetallic compound,
solid solution and amorphous phase of a given composition, based on
semi-empirical Miedema model (and some extensions), particularly for
transitional metal alloys.</p>
<p>Support elemental, binary and multicomponent alloys.
For elemental/binary alloys, the formulation is based on the original
works by Miedema et al. in 1980s;
For multicomponent alloys, the formulation is basically the linear
combination of sub-binary systems. This is reported to work well for
ternary alloys, but needs to be careful with quaternary alloys and more.</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>struct_types (str or [str]): default=’all’</dt><dd><p>‘inter’: intermetallic compound; ‘ss’: solid solution
‘amor’: amorphous phase; ‘all’: same for [‘inter’, ‘ss’, ‘amor’]
[‘inter’, ‘ss’]: amorphous phase and solid solution</p>
</dd>
<dt>ss_types (str or [str]): only for ss, default=’min’</dt><dd><p>‘fcc’: fcc solid solution; ‘bcc’: bcc solid solution
‘hcp’: hcp solid solution;
‘no_latt’: solid solution with no specific structure type
‘min’: min value of [‘fcc’, ‘bcc’, ‘hcp’, ‘no_latt’]
‘all’: same for [‘fcc’, ‘bcc’, ‘hcp’, ‘no_latt’]
[‘fcc’, ‘bcc’]: fcc and bcc solid solutions</p>
</dd>
<dt>data_source (str): source of dataset, default=’Miedema’</dt><dd><p>‘Miedema’: ‘Miedema.csv’ placed in “matminer/utils/data_files/”,
containing the following model parameters for 73 elements:
‘molar_volume’, ‘electron_density’, ‘electronegativity’
‘valence_electrons’, ‘a_const’, ‘R_const’, ‘H_trans’
‘compressibility’, ‘shear_modulus’, ‘melting_point’
‘structural_stability’. Please see the references for details.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl>
<dt>(list of floats) Miedema formation enthalpies (eV/atom) for input</dt><dd><p>struct_types:
-Miedema_deltaH_inter: for intermetallic compound
-Miedema_deltaH_ss: for solid solution, can include ‘fcc’, ‘bcc’,</p>
<blockquote>
<div><p>‘hcp’, ‘no_latt’, ‘min’ based on input ss_types</p>
</div></blockquote>
<p>-Miedema_deltaH_amor: for amorphous phase</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.composition.Miedema.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">struct_types</span><span class="o">=</span><span class="default_value">'all'</span></em>, <em class="sig-param"><span class="n">ss_types</span><span class="o">=</span><span class="default_value">'min'</span></em>, <em class="sig-param"><span class="n">data_source</span><span class="o">=</span><span class="default_value">'Miedema'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.Miedema.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.Miedema.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.Miedema.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.Miedema.deltaH_chem">
<code class="sig-name descname">deltaH_chem</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">elements</span></em>, <em class="sig-param"><span class="n">fracs</span></em>, <em class="sig-param"><span class="n">struct</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.Miedema.deltaH_chem" title="Permalink to this definition">¶</a></dt>
<dd><p>Chemical term of formation enthalpy
Args:</p>
<blockquote>
<div><p>elements (list of str): list of elements
fracs (list of floats): list of atomic fractions
struct (str): ‘inter’, ‘ss’ or ‘amor’</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>deltaH_chem (float): chemical term of formation enthalpy</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.Miedema.deltaH_elast">
<code class="sig-name descname">deltaH_elast</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">elements</span></em>, <em class="sig-param"><span class="n">fracs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.Miedema.deltaH_elast" title="Permalink to this definition">¶</a></dt>
<dd><p>Elastic term of formation enthalpy
Args:</p>
<blockquote>
<div><p>elements (list of str): list of elements
fracs (list of floats): list of atomic fractions</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>deltaH_elastic (float): elastic term of formation enthalpy</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.Miedema.deltaH_struct">
<code class="sig-name descname">deltaH_struct</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">elements</span></em>, <em class="sig-param"><span class="n">fracs</span></em>, <em class="sig-param"><span class="n">latt</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.Miedema.deltaH_struct" title="Permalink to this definition">¶</a></dt>
<dd><p>Structural term of formation enthalpy, only for solid solution
Args:</p>
<blockquote>
<div><p>elements (list of str): list of elements
fracs (list of floats): list of atomic fractions
latt (str): ‘fcc’, ‘bcc’, ‘hcp’ or ‘no_latt’</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>deltaH_struct (float): structural term of formation enthalpy</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.Miedema.deltaH_topo">
<code class="sig-name descname">deltaH_topo</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">elements</span></em>, <em class="sig-param"><span class="n">fracs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.Miedema.deltaH_topo" title="Permalink to this definition">¶</a></dt>
<dd><p>Topological term of formation enthalpy, only for amorphous phase
Args:</p>
<blockquote>
<div><p>elements (list of str): list of elements
fracs (list of floats): list of atomic fractions</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>deltaH_topo (float): topological term of formation enthalpy</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.Miedema.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.Miedema.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.Miedema.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.Miedema.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get Miedema formation enthalpies of target structures: inter, amor,
ss (can be further divided into ‘min’, ‘fcc’, ‘bcc’, ‘hcp’, ‘no_latt’</p>
<blockquote>
<div><p>for different lattice_types)</p>
</div></blockquote>
<dl class="simple">
<dt>Args:</dt><dd><p>comp: Pymatgen composition object</p>
</dd>
<dt>Returns:</dt><dd><p>miedema (list of floats): formation enthalpies of target structures</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.Miedema.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.Miedema.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.Miedema.precheck">
<code class="sig-name descname">precheck</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">c</span><span class="p">:</span> <span class="n">pymatgen.core.composition.Composition</span></em><span class="sig-paren">)</span> &#x2192; <a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a><a class="headerlink" href="#matminer.featurizers.composition.Miedema.precheck" title="Permalink to this definition">¶</a></dt>
<dd><p>Precheck a single entry. Miedema does not work for compositons
containing any elments for which the Miedema model has no parameters.
To precheck an entire dataframe (qnd automatically gather
the fraction of structures that will pass the precheck), please use
precheck_dataframe.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>c (pymatgen.Composition): The composition to precheck.</p>
</dd>
<dt>Returns:</dt><dd><p>(bool): If True, s passed the precheck; otherwise, it failed.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.composition.OxidationStates">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.composition.</code><code class="sig-name descname">OxidationStates</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">stats</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.OxidationStates" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Statistics about the oxidation states for each specie.
Features are concentration-weighted statistics of the oxidation states.</p>
<dl class="py method">
<dt id="matminer.featurizers.composition.OxidationStates.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">stats</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.OxidationStates.__init__" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Args:</dt><dd><p>stats - (list of string), which statistics compute</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.OxidationStates.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.OxidationStates.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.OxidationStates.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.OxidationStates.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.OxidationStates.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.OxidationStates.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, which has to be implemented
in any derived featurizer subclass.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).</p>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.OxidationStates.from_preset">
<em class="property">classmethod </em><code class="sig-name descname">from_preset</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">preset_name</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.OxidationStates.from_preset" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.OxidationStates.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.OxidationStates.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.composition.Stoichiometry">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.composition.</code><code class="sig-name descname">Stoichiometry</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">p_list</span><span class="o">=</span><span class="default_value">0, 2, 3, 5, 7, 10</span></em>, <em class="sig-param"><span class="n">num_atoms</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.Stoichiometry" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Calculate norms of stoichiometric attributes.</p>
<dl class="simple">
<dt>Parameters:</dt><dd><p>p_list (list of ints): list of norms to calculate
num_atoms (bool): whether to return number of atoms per formula unit</p>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.composition.Stoichiometry.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">p_list</span><span class="o">=</span><span class="default_value">0, 2, 3, 5, 7, 10</span></em>, <em class="sig-param"><span class="n">num_atoms</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.Stoichiometry.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.Stoichiometry.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.Stoichiometry.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.Stoichiometry.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.Stoichiometry.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.Stoichiometry.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.Stoichiometry.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get stoichiometric attributes
Args:</p>
<blockquote>
<div><p>comp: Pymatgen composition object
p_list (list of ints)</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>p_norm (list of floats): Lp norm-based stoichiometric attributes.</dt><dd><p>Returns number of atoms if no p-values specified.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.Stoichiometry.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.Stoichiometry.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.composition.TMetalFraction">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.composition.</code><code class="sig-name descname">TMetalFraction</code><a class="headerlink" href="#matminer.featurizers.composition.TMetalFraction" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Class to calculate fraction of magnetic transition metals in a composition.</p>
<dl class="simple">
<dt>Parameters:</dt><dd><p>data_source (data class): source from which to retrieve element data</p>
</dd>
</dl>
<p>Generates: Fraction of magnetic transition metal atoms in a compound</p>
<dl class="py method">
<dt id="matminer.featurizers.composition.TMetalFraction.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.TMetalFraction.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.TMetalFraction.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.TMetalFraction.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.TMetalFraction.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.TMetalFraction.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.TMetalFraction.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.TMetalFraction.featurize" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Args:</dt><dd><p>comp: Pymatgen Composition object</p>
</dd>
<dt>Returns:</dt><dd><p>frac_magn_atoms (single-element list): fraction of magnetic transitional metal atoms in a compound</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.TMetalFraction.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.TMetalFraction.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.composition.ValenceOrbital">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.composition.</code><code class="sig-name descname">ValenceOrbital</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">orbitals</span><span class="o">=</span><span class="default_value">'s', 'p', 'd', 'f'</span></em>, <em class="sig-param"><span class="n">props</span><span class="o">=</span><span class="default_value">'avg', 'frac'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ValenceOrbital" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Attributes of valence orbital shells</p>
<dl>
<dt>Args:</dt><dd><p>data_source (data object): source from which to retrieve element data
orbitals (list): orbitals to calculate
props (list): specifies whether to return average number of electrons in each orbital,</p>
<blockquote>
<div><p>fraction of electrons in each orbital, or both</p>
</div></blockquote>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.composition.ValenceOrbital.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">orbitals</span><span class="o">=</span><span class="default_value">'s', 'p', 'd', 'f'</span></em>, <em class="sig-param"><span class="n">props</span><span class="o">=</span><span class="default_value">'avg', 'frac'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ValenceOrbital.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.ValenceOrbital.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ValenceOrbital.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.ValenceOrbital.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ValenceOrbital.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.ValenceOrbital.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ValenceOrbital.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Weighted fraction of valence electrons in each orbital</p>
<dl class="simple">
<dt>Args:</dt><dd><p>comp: Pymatgen composition object</p>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>valence_attributes (list of floats): Average number and/or</dt><dd><p>fraction of valence electrons in specfied orbitals</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.ValenceOrbital.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.ValenceOrbital.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.composition.YangSolidSolution">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.composition.</code><code class="sig-name descname">YangSolidSolution</code><a class="headerlink" href="#matminer.featurizers.composition.YangSolidSolution" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Mixing thermochemistry and size mismatch terms of Yang and Zhang (2012)</p>
<p>This featurizer returns two different features developed by
.. Yang and Zhang <cite>https://linkinghub.elsevier.com/retrieve/pii/S0254058411009357</cite>
to predict whether metal alloys will form metallic glasses,
crystalline solid solutions, or intermetallics.
The first, Omega, is related to the balance between the mixing entropy and
mixing enthalpy of the liquid phase. The second, delta, is related to the
atomic size mismatch between the different elements of the material.</p>
<dl class="simple">
<dt>Features</dt><dd><p>Yang omega - Mixing thermochemistry feature, Omega
Yang delta - Atomic size mismatch term</p>
</dd>
<dt>References:</dt><dd></dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.composition.YangSolidSolution.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.YangSolidSolution.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.YangSolidSolution.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.YangSolidSolution.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.YangSolidSolution.compute_delta">
<code class="sig-name descname">compute_delta</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.YangSolidSolution.compute_delta" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute Yang’s delta parameter</p>
<p><img class="math" src="_images/math/b4111b2fd8976378ce69e158f8d25467da391510.png" alt="\sqrt{\sum^n_{i=1} c_i \left( 1 - \frac{r_i}{\bar{r}} \right)^2 }"/></p>
<p>where <img class="math" src="_images/math/30634b08f5df289a7acab4963cb6a3e6b99aba0f.png" alt="c_i"/> and <img class="math" src="_images/math/dbc4c429b48dd94e41ee866e1edbf0abededae3a.png" alt="r_i"/> are the fraction and radius of
element <img class="math" src="_images/math/5aa339d4daf45a810dda332e3c80a0698e526e04.png" alt="i"/>, and <img class="math" src="_images/math/5df994acd5cadc81c38f1a28e4ea3f8281b6c4cb.png" alt="\bar{r}"/> is the fraction-weighted
average of the radii. We use the radii compiled by
.. Miracle et al. <cite>https://www.tandfonline.com/doi/ref/10.1179/095066010X12646898728200?scroll=top</cite>.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>comp (Composition) - Composition to assess</p>
</dd>
<dt>Returns:</dt><dd><p>(float) delta</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.YangSolidSolution.compute_omega">
<code class="sig-name descname">compute_omega</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.YangSolidSolution.compute_omega" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute Yang’s mixing thermodynamics descriptor</p>
<p><img class="math" src="_images/math/7c4a5c45c86c3c6a4b3d73c5a58b549b14d201df.png" alt="\frac{T_m \Delta S_{mix}}{ |  \Delta H_{mix} | }"/></p>
<p>Where <img class="math" src="_images/math/60b74287bf19e27bc5a3e755457326add30fc669.png" alt="T_m"/> is average melting temperature,
<img class="math" src="_images/math/2fe6d1d220a6eb68148f55432fd2234ca529685c.png" alt="\Delta S_{mix}"/> is the ideal mixing entropy,
and <img class="math" src="_images/math/4fbce1d9d579dd5c678c54ab9967433263dbeaa2.png" alt="\Delta H_{mix}"/> is the average mixing enthalpies
of all pairs of elements in the alloy</p>
<dl class="simple">
<dt>Args:</dt><dd><p>comp (Composition) - Composition to featurizer</p>
</dd>
<dt>Returns:</dt><dd><p>(float) Omega</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.YangSolidSolution.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.YangSolidSolution.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.YangSolidSolution.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.YangSolidSolution.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, which has to be implemented
in any derived featurizer subclass.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).</p>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.YangSolidSolution.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.YangSolidSolution.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.composition.YangSolidSolution.precheck">
<code class="sig-name descname">precheck</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">c</span><span class="p">:</span> <span class="n">pymatgen.core.composition.Composition</span></em><span class="sig-paren">)</span> &#x2192; <a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a><a class="headerlink" href="#matminer.featurizers.composition.YangSolidSolution.precheck" title="Permalink to this definition">¶</a></dt>
<dd><p>Precheck a single entry. YangSolidSolution does not work for compositons
containing any binary elment combinations for which the model has no
parameters. We can nearly equivalently approximate this by checking
against the unary element list.</p>
<p>To precheck an entire dataframe (qnd automatically gather
the fraction of structures that will pass the precheck), please use
precheck_dataframe.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>c (pymatgen.Composition): The composition to precheck.</p>
</dd>
<dt>Returns:</dt><dd><p>(bool): If True, s passed the precheck; otherwise, it failed.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="matminer.featurizers.composition.has_oxidation_states">
<code class="sig-prename descclassname">matminer.featurizers.composition.</code><code class="sig-name descname">has_oxidation_states</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.has_oxidation_states" title="Permalink to this definition">¶</a></dt>
<dd><p>Check if a composition object has oxidation states for each element</p>
<p>TODO: Does this make sense to add to pymatgen? -wardlt</p>
<dl class="simple">
<dt>Args:</dt><dd><p>comp (Composition): Composition to check</p>
</dd>
<dt>Returns:</dt><dd><p>(boolean) Whether this composition object contains oxidation states</p>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="matminer.featurizers.composition.is_ionic">
<code class="sig-prename descclassname">matminer.featurizers.composition.</code><code class="sig-name descname">is_ionic</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.composition.is_ionic" title="Permalink to this definition">¶</a></dt>
<dd><p>Determines whether a compound is an ionic compound.</p>
<p>Looks at the oxidation states of each site and checks if both anions and cations exist</p>
<dl class="simple">
<dt>Args:</dt><dd><p>comp (Composition): Composition to check</p>
</dd>
<dt>Returns:</dt><dd><p>(bool) Whether the composition describes an ionic compound</p>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-matminer.featurizers.conversions">
<span id="matminer-featurizers-conversions-module"></span><h2>matminer.featurizers.conversions module<a class="headerlink" href="#module-matminer.featurizers.conversions" title="Permalink to this headline">¶</a></h2>
<p>This module defines featurizers that can convert between different data formats</p>
<p>Note that these featurizers do not produce machine learning-ready features.
Instead, they should be used to pre-process data, either through a standalone
transformation or as part of a Pipeline.</p>
<dl class="py class">
<dt id="matminer.featurizers.conversions.CompositionToOxidComposition">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.conversions.</code><code class="sig-name descname">CompositionToOxidComposition</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target_col_id</span><span class="o">=</span><span class="default_value">'composition_oxid'</span></em>, <em class="sig-param"><span class="n">overwrite_data</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">coerce_mixed</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">return_original_on_error</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.CompositionToOxidComposition" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.conversions.ConversionFeaturizer" title="matminer.featurizers.conversions.ConversionFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.conversions.ConversionFeaturizer</span></code></a></p>
<p>Utility featurizer to add oxidation states to a pymatgen Composition.</p>
<p>Oxidation states are determined using pymatgen’s guessing routines.
The expected input is a <cite>pymatgen.core.composition.Composition</cite> object.</p>
<p>Note that this Featurizer does not produce machine learning-ready features
but instead can be applied to pre-process data or as part of a Pipeline.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>target_col_id (str or None): The column in which the converted data will</dt><dd><p>be written. If the column already exists then an error will be
thrown unless <cite>overwrite_data</cite> is set to <cite>True</cite>. If <cite>target_col_id</cite>
begins with an underscore the data will be written to the column:
<cite>“{}_{}”.format(col_id, target_col_id[1:])</cite>, where <cite>col_id</cite> is the
column being featurized. If <cite>target_col_id</cite> is set to None then
the data will be written “in place” to the <cite>col_id</cite> column (this
will only work if <cite>overwrite_data=True</cite>).</p>
</dd>
<dt>overwrite_data (bool): Overwrite any data in <cite>target_column</cite> if it</dt><dd><p>exists.</p>
</dd>
<dt>coerce_mixed (bool): If a composition has both species containing</dt><dd><p>oxid states and not containing oxid states, strips all of the
oxid states and guesses the entire composition’s oxid states.</p>
</dd>
<dt>return_original_on_error: If the oxidation states cannot be</dt><dd><p>guessed and set to True, the composition without oxidation states
will be returned. If set to False, an error will be thrown.</p>
</dd>
<dt><a href="#id3"><span class="problematic" id="id4">**</span></a>kwargs: Parameters to control the settings for</dt><dd><p><cite>pymatgen.io.structure.Structure.add_oxidation_state_by_guess()</cite>.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.conversions.CompositionToOxidComposition.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target_col_id</span><span class="o">=</span><span class="default_value">'composition_oxid'</span></em>, <em class="sig-param"><span class="n">overwrite_data</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">coerce_mixed</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">return_original_on_error</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.CompositionToOxidComposition.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.CompositionToOxidComposition.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.CompositionToOxidComposition.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.CompositionToOxidComposition.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.CompositionToOxidComposition.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Add oxidation states to a Structure using pymatgen’s guessing routines.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>comp (<cite>pymatgen.core.composition.Composition</cite>): A composition.</p>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>(<cite>pymatgen.core.composition.Composition</cite>): A Composition object</dt><dd><p>decorated with oxidation states.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.CompositionToOxidComposition.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.CompositionToOxidComposition.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.conversions.CompositionToStructureFromMP">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.conversions.</code><code class="sig-name descname">CompositionToStructureFromMP</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target_col_id</span><span class="o">=</span><span class="default_value">'structure'</span></em>, <em class="sig-param"><span class="n">overwrite_data</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">mapi_key</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.CompositionToStructureFromMP" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.conversions.ConversionFeaturizer" title="matminer.featurizers.conversions.ConversionFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.conversions.ConversionFeaturizer</span></code></a></p>
<p>Featurizer to get a Structure object from Materials Project using the
composition alone. The most stable entry from Materials Project is selected,
or NaN if no entry is found in the Materials Project.</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>target_col_id (str or None): The column in which the converted data will</dt><dd><p>be written. If the column already exists then an error will be
thrown unless <cite>overwrite_data</cite> is set to <cite>True</cite>. If <cite>target_col_id</cite>
begins with an underscore the data will be written to the column:
<cite>“{}_{}”.format(col_id, target_col_id[1:])</cite>, where <cite>col_id</cite> is the
column being featurized. If <cite>target_col_id</cite> is set to None then
the data will be written “in place” to the <cite>col_id</cite> column (this
will only work if <cite>overwrite_data=True</cite>).</p>
</dd>
<dt>overwrite_data (bool): Overwrite any data in <cite>target_column</cite> if it</dt><dd><p>exists.</p>
</dd>
</dl>
<p>map_key (str): Materials API key</p>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.conversions.CompositionToStructureFromMP.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target_col_id</span><span class="o">=</span><span class="default_value">'structure'</span></em>, <em class="sig-param"><span class="n">overwrite_data</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">mapi_key</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.CompositionToStructureFromMP.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.CompositionToStructureFromMP.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.CompositionToStructureFromMP.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.CompositionToStructureFromMP.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">comp</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.CompositionToStructureFromMP.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the most stable structure from Materials Project
Args:</p>
<blockquote>
<div><p>comp (<cite>pymatgen.core.composition.Composition</cite>): A composition.</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>(<cite>pymatgen.core.structure.Structure</cite>): A Structure object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.CompositionToStructureFromMP.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.CompositionToStructureFromMP.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.conversions.ConversionFeaturizer">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.conversions.</code><code class="sig-name descname">ConversionFeaturizer</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target_col_id</span></em>, <em class="sig-param"><span class="n">overwrite_data</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.ConversionFeaturizer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Abstract class to perform data conversions.</p>
<p>Featurizers subclassing this class do not produce machine learning-ready
features but instead are used to pre-process data. As Featurizers,
the conversion process can take advantage of the parallelisation implemented
in ScikitLearn.</p>
<p>Note that <cite>feature_labels</cite> are set dynamically and may depend on the column
id of the data being featurized. As such, <cite>feature_labels</cite> may differ
before and after featurization.</p>
<p>ConversionFeaturizers differ from other Featurizers in that the user can
can specify the column in which to write the converted data. The output
column is controlled through <cite>target_col_id</cite>. ConversionFeaturizers also
have the ability to overwrite data in existing columns. This is
controlled by the <cite>overwrite_data</cite> option. “in place” conversion of data can
be achieved by setting <cite>target_col_id=None</cite> and <cite>overwrite_data=True</cite>. See
the docstring below for more details.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>target_col_id (str or None): The column in which the converted data will</dt><dd><p>be written. If the column already exists then an error will be
thrown unless <cite>overwrite_data</cite> is set to <cite>True</cite>. If <cite>target_col_id</cite>
begins with an underscore the data will be written to the column:
<cite>“{}_{}”.format(col_id, target_col_id[1:])</cite>, where <cite>col_id</cite> is the
column being featurized. If <cite>target_col_id</cite> is set to None then
the data will be written “in place” to the <cite>col_id</cite> column (this
will only work if <cite>overwrite_data=True</cite>).</p>
</dd>
<dt>overwrite_data (bool): Overwrite any data in <cite>target_col_id</cite> if it</dt><dd><p>exists.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.conversions.ConversionFeaturizer.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target_col_id</span></em>, <em class="sig-param"><span class="n">overwrite_data</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.ConversionFeaturizer.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.ConversionFeaturizer.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.ConversionFeaturizer.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.ConversionFeaturizer.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.ConversionFeaturizer.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.ConversionFeaturizer.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">x</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.ConversionFeaturizer.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, which has to be implemented
in any derived featurizer subclass.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).</p>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.ConversionFeaturizer.featurize_dataframe">
<code class="sig-name descname">featurize_dataframe</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">df</span></em>, <em class="sig-param"><span class="n">col_id</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.ConversionFeaturizer.featurize_dataframe" title="Permalink to this definition">¶</a></dt>
<dd><p>Perform the data conversion and set the target column dynamically.</p>
<p><cite>target_col_id</cite>, and accordingly <cite>feature_labels</cite>, may depend on the
column id of the data being featurized. As such, <cite>target_col_id</cite> is
first set dynamically before the <cite>BaseFeaturizer.featurize_dataframe()</cite>
super method is called.</p>
<dl>
<dt>Args:</dt><dd><p>df (Pandas.DataFrame): Dataframe containing input data.
col_id (str or list of str): column label containing objects to</p>
<blockquote>
<div><p>featurize. Can be multiple labels if the featurize function
requires multiple inputs.</p>
</div></blockquote>
<dl class="simple">
<dt><a href="#id5"><span class="problematic" id="id6">**</span></a>kwargs: Additional keyword arguments that will be passed through</dt><dd><p>to <cite>BseFeaturizer.featurize_dataframe()</cite>.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>(Pandas.Dataframe): The updated dataframe.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.ConversionFeaturizer.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.ConversionFeaturizer.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.conversions.DictToObject">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.conversions.</code><code class="sig-name descname">DictToObject</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target_col_id</span><span class="o">=</span><span class="default_value">'_object'</span></em>, <em class="sig-param"><span class="n">overwrite_data</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.DictToObject" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.conversions.ConversionFeaturizer" title="matminer.featurizers.conversions.ConversionFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.conversions.ConversionFeaturizer</span></code></a></p>
<p>Utility featurizer to decode a dict to Python object via MSON.</p>
<p>Note that this Featurizer does not produce machine learning-ready features
but instead can be applied to pre-process data or as part of a Pipeline.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>target_col_id (str or None): The column in which the converted data will</dt><dd><p>be written. If the column already exists then an error will be
thrown unless <cite>overwrite_data</cite> is set to <cite>True</cite>. If <cite>target_col_id</cite>
begins with an underscore the data will be written to the column:
<cite>“{}_{}”.format(col_id, target_col_id[1:])</cite>, where <cite>col_id</cite> is the
column being featurized. If <cite>target_col_id</cite> is set to None then
the data will be written “in place” to the <cite>col_id</cite> column (this
will only work if <cite>overwrite_data=True</cite>).</p>
</dd>
<dt>overwrite_data (bool): Overwrite any data in <cite>target_column</cite> if it</dt><dd><p>exists.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.conversions.DictToObject.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target_col_id</span><span class="o">=</span><span class="default_value">'_object'</span></em>, <em class="sig-param"><span class="n">overwrite_data</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.DictToObject.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.DictToObject.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.DictToObject.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.DictToObject.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dict_data</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.DictToObject.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert a string to a pymatgen Composition.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>dict_data (dict): A MSONable dictionary. E.g. Produced from</dt><dd><p><cite>pymatgen.core.structure.Structure.as_dict()</cite>.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>(object): An object with the type specified by <cite>dict_data</cite>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.DictToObject.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.DictToObject.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.conversions.JsonToObject">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.conversions.</code><code class="sig-name descname">JsonToObject</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target_col_id</span><span class="o">=</span><span class="default_value">'_object'</span></em>, <em class="sig-param"><span class="n">overwrite_data</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.JsonToObject" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.conversions.ConversionFeaturizer" title="matminer.featurizers.conversions.ConversionFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.conversions.ConversionFeaturizer</span></code></a></p>
<p>Utility featurizer to decode json data to a Python object via MSON.</p>
<p>Note that this Featurizer does not produce machine learning-ready features
but instead can be applied to pre-process data or as part of a Pipeline.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>target_col_id (str or None): The column in which the converted data will</dt><dd><p>be written. If the column already exists then an error will be
thrown unless <cite>overwrite_data</cite> is set to <cite>True</cite>. If <cite>target_col_id</cite>
begins with an underscore the data will be written to the column:
<cite>“{}_{}”.format(col_id, target_col_id[1:])</cite>, where <cite>col_id</cite> is the
column being featurized. If <cite>target_col_id</cite> is set to None then
the data will be written “in place” to the <cite>col_id</cite> column (this
will only work if <cite>overwrite_data=True</cite>).</p>
</dd>
<dt>overwrite_data (bool): Overwrite any data in <cite>target_column</cite> if it</dt><dd><p>exists.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.conversions.JsonToObject.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target_col_id</span><span class="o">=</span><span class="default_value">'_object'</span></em>, <em class="sig-param"><span class="n">overwrite_data</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.JsonToObject.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.JsonToObject.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.JsonToObject.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.JsonToObject.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">json_data</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.JsonToObject.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert a string to a pymatgen Composition.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>json_data (dict): MSONable json data. E.g. Produced from</dt><dd><p><cite>pymatgen.core.structure.Structure.to_json()</cite>.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>(object): An object with the type specified by <cite>json_data</cite>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.JsonToObject.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.JsonToObject.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.conversions.StrToComposition">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.conversions.</code><code class="sig-name descname">StrToComposition</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">reduce</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">target_col_id</span><span class="o">=</span><span class="default_value">'composition'</span></em>, <em class="sig-param"><span class="n">overwrite_data</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.StrToComposition" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.conversions.ConversionFeaturizer" title="matminer.featurizers.conversions.ConversionFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.conversions.ConversionFeaturizer</span></code></a></p>
<p>Utility featurizer to convert a string to a Composition</p>
<p>The expected input is a composition in string form (e.g. “Fe2O3”).</p>
<p>Note that this Featurizer does not produce machine learning-ready features
but instead can be applied to pre-process data or as part of a Pipeline.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>reduce (bool): Whether to return a reduced</dt><dd><p><cite>pymatgen.core.composition.Composition</cite> object.</p>
</dd>
<dt>target_col_id (str or None): The column in which the converted data will</dt><dd><p>be written. If the column already exists then an error will be
thrown unless <cite>overwrite_data</cite> is set to <cite>True</cite>. If <cite>target_col_id</cite>
begins with an underscore the data will be written to the column:
<cite>“{}_{}”.format(col_id, target_col_id[1:])</cite>, where <cite>col_id</cite> is the
column being featurized. If <cite>target_col_id</cite> is set to None then
the data will be written “in place” to the <cite>col_id</cite> column (this
will only work if <cite>overwrite_data=True</cite>).</p>
</dd>
<dt>overwrite_data (bool): Overwrite any data in <cite>target_column</cite> if it</dt><dd><p>exists.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.conversions.StrToComposition.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">reduce</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">target_col_id</span><span class="o">=</span><span class="default_value">'composition'</span></em>, <em class="sig-param"><span class="n">overwrite_data</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.StrToComposition.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.StrToComposition.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.StrToComposition.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.StrToComposition.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">string_composition</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.StrToComposition.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert a string to a pymatgen Composition.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>string_composition (str): A chemical formula as a string (e.g.</dt><dd><p>“Fe2O3”).</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>(<cite>pymatgen.core.composition.Composition</cite>): A composition object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.StrToComposition.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.StrToComposition.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.conversions.StructureToComposition">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.conversions.</code><code class="sig-name descname">StructureToComposition</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">reduce</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">target_col_id</span><span class="o">=</span><span class="default_value">'composition'</span></em>, <em class="sig-param"><span class="n">overwrite_data</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.StructureToComposition" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.conversions.ConversionFeaturizer" title="matminer.featurizers.conversions.ConversionFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.conversions.ConversionFeaturizer</span></code></a></p>
<p>Utility featurizer to convert a Structure to a Composition.</p>
<p>The expected input is a <cite>pymatgen.core.structure.Structure</cite> object.</p>
<p>Note that this Featurizer does not produce machine learning-ready features
but instead can be applied to pre-process data or as part of a Pipeline.</p>
<dl>
<dt>Args:</dt><dd><p>reduce (bool): Whether to return a reduced Composition object.
target_col_id (str or None): The column in which the converted data will</p>
<blockquote>
<div><p>be written. If the column already exists then an error will be
thrown unless <cite>overwrite_data</cite> is set to <cite>True</cite>. If <cite>target_col_id</cite>
begins with an underscore the data will be written to the column:
<cite>“{}_{}”.format(col_id, target_col_id[1:])</cite>, where <cite>col_id</cite> is the
column being featurized. If <cite>target_col_id</cite> is set to None then
the data will be written “in place” to the <cite>col_id</cite> column (this
will only work if <cite>overwrite_data=True</cite>).</p>
</div></blockquote>
<dl class="simple">
<dt>overwrite_data (bool): Overwrite any data in <cite>target_column</cite> if it</dt><dd><p>exists.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.conversions.StructureToComposition.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">reduce</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">target_col_id</span><span class="o">=</span><span class="default_value">'composition'</span></em>, <em class="sig-param"><span class="n">overwrite_data</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.StructureToComposition.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.StructureToComposition.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.StructureToComposition.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.StructureToComposition.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">structure</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.StructureToComposition.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert a string to a pymatgen Composition.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>structure (<cite>pymatgen.core.structure.Structure</cite>): A structure.</p>
</dd>
<dt>Returns:</dt><dd><p>(<cite>pymatgen.core.composition.Composition</cite>): A Composition object.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.StructureToComposition.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.StructureToComposition.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.conversions.StructureToIStructure">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.conversions.</code><code class="sig-name descname">StructureToIStructure</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target_col_id</span><span class="o">=</span><span class="default_value">'istructure'</span></em>, <em class="sig-param"><span class="n">overwrite_data</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.StructureToIStructure" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.conversions.ConversionFeaturizer" title="matminer.featurizers.conversions.ConversionFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.conversions.ConversionFeaturizer</span></code></a></p>
<p>Utility featurizer to convert a Structure to an immutable IStructure.</p>
<p>This is useful if you are using features that employ caching.</p>
<p>The expected input is a <cite>pymatgen.core.structure.Structure</cite> object.</p>
<p>Note that this Featurizer does not produce machine learning-ready features
but instead can be applied to pre-process data or as part of a Pipeline.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>target_col_id (str or None): The column in which the converted data will</dt><dd><p>be written. If the column already exists then an error will be
thrown unless <cite>overwrite_data</cite> is set to <cite>True</cite>. If <cite>target_col_id</cite>
begins with an underscore the data will be written to the column:
<cite>“{}_{}”.format(col_id, target_col_id[1:])</cite>, where <cite>col_id</cite> is the
column being featurized. If <cite>target_col_id</cite> is set to None then
the data will be written “in place” to the <cite>col_id</cite> column (this
will only work if <cite>overwrite_data=True</cite>).</p>
</dd>
<dt>overwrite_data (bool): Overwrite any data in <cite>target_column</cite> if it</dt><dd><p>exists.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.conversions.StructureToIStructure.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target_col_id</span><span class="o">=</span><span class="default_value">'istructure'</span></em>, <em class="sig-param"><span class="n">overwrite_data</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.StructureToIStructure.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.StructureToIStructure.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.StructureToIStructure.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.StructureToIStructure.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">structure</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.StructureToIStructure.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert a pymatgen Structure to an immutable IStructure,</p>
<dl class="simple">
<dt>Args:</dt><dd><p>structure (<cite>pymatgen.core.structure.Structure</cite>): A structure.</p>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>(<cite>pymatgen.core.structure.IStructure</cite>): An immutable IStructure</dt><dd><p>object.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.StructureToIStructure.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.StructureToIStructure.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.conversions.StructureToOxidStructure">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.conversions.</code><code class="sig-name descname">StructureToOxidStructure</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target_col_id</span><span class="o">=</span><span class="default_value">'structure_oxid'</span></em>, <em class="sig-param"><span class="n">overwrite_data</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">return_original_on_error</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.StructureToOxidStructure" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.conversions.ConversionFeaturizer" title="matminer.featurizers.conversions.ConversionFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.conversions.ConversionFeaturizer</span></code></a></p>
<p>Utility featurizer to add oxidation states to a pymatgen Structure.</p>
<p>Oxidation states are determined using pymatgen’s guessing routines.
The expected input is a <cite>pymatgen.core.structure.Structure</cite> object.</p>
<p>Note that this Featurizer does not produce machine learning-ready features
but instead can be applied to pre-process data or as part of a Pipeline.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>target_col_id (str or None): The column in which the converted data will</dt><dd><p>be written. If the column already exists then an error will be
thrown unless <cite>overwrite_data</cite> is set to <cite>True</cite>. If <cite>target_col_id</cite>
begins with an underscore the data will be written to the column:
<cite>“{}_{}”.format(col_id, target_col_id[1:])</cite>, where <cite>col_id</cite> is the
column being featurized. If <cite>target_col_id</cite> is set to None then
the data will be written “in place” to the <cite>col_id</cite> column (this
will only work if <cite>overwrite_data=True</cite>).</p>
</dd>
<dt>overwrite_data (bool): Overwrite any data in <cite>target_column</cite> if it</dt><dd><p>exists.</p>
</dd>
<dt>return_original_on_error: If the oxidation states cannot be</dt><dd><p>guessed and set to True, the structure without oxidation states will
be returned. If set to False, an error will be thrown.</p>
</dd>
<dt><a href="#id7"><span class="problematic" id="id8">**</span></a>kwargs: Parameters to control the settings for</dt><dd><p><cite>pymatgen.io.structure.Structure.add_oxidation_state_by_guess()</cite>.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.conversions.StructureToOxidStructure.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target_col_id</span><span class="o">=</span><span class="default_value">'structure_oxid'</span></em>, <em class="sig-param"><span class="n">overwrite_data</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">return_original_on_error</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.StructureToOxidStructure.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.StructureToOxidStructure.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.StructureToOxidStructure.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.StructureToOxidStructure.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">structure</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.StructureToOxidStructure.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Add oxidation states to a Structure using pymatgen’s guessing routines.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>structure (<cite>pymatgen.core.structure.Structure</cite>): A structure.</p>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>(<cite>pymatgen.core.structure.Structure</cite>): A Structure object decorated</dt><dd><p>with oxidation states.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.conversions.StructureToOxidStructure.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.conversions.StructureToOxidStructure.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-matminer.featurizers.deprecated">
<span id="matminer-featurizers-deprecated-module"></span><h2>matminer.featurizers.deprecated module<a class="headerlink" href="#module-matminer.featurizers.deprecated" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt id="matminer.featurizers.deprecated.CrystalSiteFingerprint">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.deprecated.</code><code class="sig-name descname">CrystalSiteFingerprint</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.deprecated.CrystalSiteFingerprint" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>A local order parameter fingerprint for periodic crystals.</p>
<p>A site fingerprint intended for periodic crystals. The fingerprint represents
the value of various order parameters for the site; each value is the product
two quantities: (i) the value of the order parameter itself and (ii) a factor
that describes how consistent the number of neighbors is with that order
parameter. Note that we can include only factor (ii) using the “wt” order
parameter which is always set to 1. Also note that the cation-anion flag
works only if the structures are oxidation-state decorated (e.g., use
pymatgen’s BVAnalyzer or matminer’s structure_to_oxidstructure()).</p>
<dl class="py method">
<dt id="matminer.featurizers.deprecated.CrystalSiteFingerprint.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.deprecated.CrystalSiteFingerprint.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.deprecated.CrystalSiteFingerprint.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.deprecated.CrystalSiteFingerprint.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.deprecated.CrystalSiteFingerprint.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.deprecated.CrystalSiteFingerprint.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.deprecated.CrystalSiteFingerprint.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">struct</span></em>, <em class="sig-param"><span class="n">idx</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.deprecated.CrystalSiteFingerprint.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get crystal fingerprint of site with given index in input
structure.
Args:</p>
<blockquote>
<div><p>struct (Structure): Pymatgen Structure object.
idx (int): index of target site in structure.</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>list of weighted order parameters of target site.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.deprecated.CrystalSiteFingerprint.from_preset">
<em class="property">static </em><code class="sig-name descname">from_preset</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">preset</span></em>, <em class="sig-param"><span class="n">cation_anion</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.deprecated.CrystalSiteFingerprint.from_preset" title="Permalink to this definition">¶</a></dt>
<dd><p>Use preset parameters to get the fingerprint
Args:</p>
<blockquote>
<div><p>preset (str): name of preset (“cn” or “ops”)
cation_anion (bool): whether to only consider cation&lt;-&gt;anion bonds</p>
<blockquote>
<div><p>(bonds with zero charge are also allowed)</p>
</div></blockquote>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.deprecated.CrystalSiteFingerprint.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.deprecated.CrystalSiteFingerprint.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-matminer.featurizers.dos">
<span id="matminer-featurizers-dos-module"></span><h2>matminer.featurizers.dos module<a class="headerlink" href="#module-matminer.featurizers.dos" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt id="matminer.featurizers.dos.DOSFeaturizer">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.dos.</code><code class="sig-name descname">DOSFeaturizer</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">contributors</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">decay_length</span><span class="o">=</span><span class="default_value">0.1</span></em>, <em class="sig-param"><span class="n">sampling_resolution</span><span class="o">=</span><span class="default_value">100</span></em>, <em class="sig-param"><span class="n">gaussian_smear</span><span class="o">=</span><span class="default_value">0.05</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.DOSFeaturizer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Significant character and contribution of the density of state from a
CompleteDos, object. Contributors are the atomic orbitals from each site
within the structure. This underlines the importance of dos.structure.</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>contributors (int):</dt><dd><p>Sets the number of top contributors to the DOS that are
returned as features. (i.e. contributors=1 will only return the
main cb and main vb orbital)</p>
</dd>
<dt>decay_length (float in eV):</dt><dd><p>The dos is sampled by an exponential decay function. this parameter
sets the decay length of the exponential. Three times the decay
length corresponds to 10% sampling strength. There is a hard cutoff
at five times the decay length (1% sampling strength)</p>
</dd>
<dt>sampling_resolution (int):</dt><dd><p>Number of points to sample DOS</p>
</dd>
<dt>gaussian_smear (float in eV):</dt><dd><p>Gaussian smearing (sigma) around each sampled point in the DOS</p>
</dd>
</dl>
</dd>
<dt>Returns (featurize returns [float] and featurize_labels returns [str]):</dt><dd><p>xbm_score_i (float): fractions of ith contributor orbital
xbm_location_i (str): fractional coordinate of ith contributor/site
xbm_character_i (str): character of ith contributor (s, p, d, f)
xbm_specie_i (str): elemental specie of ith contributor (ex: ‘Ti’)
xbm_hybridization (int): the amount of hybridization at the band edge</p>
<blockquote>
<div><p>characterized by an entropy score (x ln x). the hybridization score
is larger for a greater number of significant contributors</p>
</div></blockquote>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.dos.DOSFeaturizer.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">contributors</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">decay_length</span><span class="o">=</span><span class="default_value">0.1</span></em>, <em class="sig-param"><span class="n">sampling_resolution</span><span class="o">=</span><span class="default_value">100</span></em>, <em class="sig-param"><span class="n">gaussian_smear</span><span class="o">=</span><span class="default_value">0.05</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.DOSFeaturizer.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.dos.DOSFeaturizer.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.DOSFeaturizer.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.dos.DOSFeaturizer.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.DOSFeaturizer.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Returns ([str]): list of names of the features. See the docs for the</dt><dd><p>featurize method for more information.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.dos.DOSFeaturizer.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dos</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.DOSFeaturizer.featurize" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>dos (pymatgen CompleteDos or their dict):</dt><dd><p>The density of states to featurize. Must be a complete DOS,
(i.e. contains PDOS and structure, in addition to total DOS)
and must contain the structure.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.dos.DOSFeaturizer.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.DOSFeaturizer.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.dos.DopingFermi">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.dos.</code><code class="sig-name descname">DopingFermi</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dopings</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">eref</span><span class="o">=</span><span class="default_value">'midgap'</span></em>, <em class="sig-param"><span class="n">T</span><span class="o">=</span><span class="default_value">300</span></em>, <em class="sig-param"><span class="n">return_eref</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.DopingFermi" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>The fermi level (w.r.t. selected reference energy) associated with a
specified carrier concentration (1/cm3) and temperature. This featurizar
requires the total density of states and structure. The Structure
as dos.structure (e.g. in CompleteDos) is required by FermiDos class.</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>dopings ([float]): list of doping concentrations 1/cm3. Note that a</dt><dd><p>negative concentration is treated as electron majority carrier
(n-type) and positive for holes (p-type)</p>
</dd>
<dt>eref (str or int or float): energy alignment reference. Defaults</dt><dd><p>to midgap (equilibrium fermi). A fixed number can also be used.
str options: “midgap”, “vbm”, “cbm”, “dos_fermi”, “band_center”</p>
</dd>
</dl>
<p>T (float): absolute temperature in Kelvin
return_eref: if True, instead of aligning the fermi levels based</p>
<blockquote>
<div><p>on eref, it (eref) will be explicitly returned as a feature</p>
</div></blockquote>
</dd>
<dt>Returns (featurize returns [float] and featurize_labels returns [str]):</dt><dd><dl class="simple">
<dt>examples:</dt><dd><dl class="simple">
<dt>fermi_c-1e+20T300 (float): the fermi level for the electron</dt><dd><p>concentration of 1e20 and the temperature of 300K.</p>
</dd>
<dt>fermi_c1e+18T600 (float): fermi level for the hole concentration</dt><dd><p>of 1e18 and the temperature of 600K.</p>
</dd>
<dt>midgap eref (float): if return_eref==True then eref (midgap here)</dt><dd><p>energy is returned. In this case, fermi levels are absolute as
opposed to relative to eref (i.e. if not return_eref)</p>
</dd>
</dl>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.dos.DopingFermi.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dopings</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">eref</span><span class="o">=</span><span class="default_value">'midgap'</span></em>, <em class="sig-param"><span class="n">T</span><span class="o">=</span><span class="default_value">300</span></em>, <em class="sig-param"><span class="n">return_eref</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.DopingFermi.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.dos.DopingFermi.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.DopingFermi.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.dos.DopingFermi.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.DopingFermi.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Returns ([str]): list of names of the features generated by featurize</dt><dd><p>example: “fermi_c-1e+20T300” that is the fermi level for the
electron concentration of 1e20 (c-1e+20) and temperature of 300K.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.dos.DopingFermi.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dos</span></em>, <em class="sig-param"><span class="n">bandgap</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.DopingFermi.featurize" title="Permalink to this definition">¶</a></dt>
<dd><dl>
<dt>Args:</dt><dd><p>dos (pymatgen Dos, CompleteDos or FermiDos):
bandgap (float): for example the experimentally measured band gap</p>
<blockquote>
<div><p>or one that is calculated via more accurate methods than the
one used to generate dos. dos will be scissored to have the
same electronic band gap as bandgap.</p>
</div></blockquote>
</dd>
<dt>Returns ([float]): features are fermi levels in eV at the given</dt><dd><p>concentrations and temperature + eref in eV if return_eref</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.dos.DopingFermi.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.DopingFermi.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.dos.DosAsymmetry">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.dos.</code><code class="sig-name descname">DosAsymmetry</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">decay_length</span><span class="o">=</span><span class="default_value">0.5</span></em>, <em class="sig-param"><span class="n">sampling_resolution</span><span class="o">=</span><span class="default_value">100</span></em>, <em class="sig-param"><span class="n">gaussian_smear</span><span class="o">=</span><span class="default_value">0.05</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.DosAsymmetry" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Quantifies the asymmetry of the DOS near the Fermi level.</p>
<p>The DOS asymmetry is defined the natural logarithm of the quotient of the
total DOS above the Fermi level and the total DOS below the Fermi level. A
positive number indicates that there are more states directly above the
Fermi level than below the Fermi level. This featurizer is only meant for
metals and semi-metals.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>decay_length (float in eV):</dt><dd><p>The dos is sampled by an exponential decay function. this parameter
sets the decay length of the exponential. Three times the decay
length corresponds to 10% sampling strength. There is a hard cutoff
at five times the decay length (1% sampling strength)</p>
</dd>
<dt>sampling_resolution (int):</dt><dd><p>Number of points to sample DOS</p>
</dd>
<dt>gaussian_smear (float in eV):</dt><dd><p>Gaussian smearing (sigma) around each sampled point in the DOS</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.dos.DosAsymmetry.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">decay_length</span><span class="o">=</span><span class="default_value">0.5</span></em>, <em class="sig-param"><span class="n">sampling_resolution</span><span class="o">=</span><span class="default_value">100</span></em>, <em class="sig-param"><span class="n">gaussian_smear</span><span class="o">=</span><span class="default_value">0.05</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.DosAsymmetry.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.dos.DosAsymmetry.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.DosAsymmetry.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.dos.DosAsymmetry.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.DosAsymmetry.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the labels for each of the features.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.dos.DosAsymmetry.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dos</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.DosAsymmetry.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates the DOS asymmetry.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>dos (Dos): A pymatgen Dos object.</p>
</dd>
<dt>Returns:</dt><dd><p>A float describing the asymmetry of the DOS.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.dos.DosAsymmetry.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.DosAsymmetry.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.dos.Hybridization">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.dos.</code><code class="sig-name descname">Hybridization</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">decay_length</span><span class="o">=</span><span class="default_value">0.1</span></em>, <em class="sig-param"><span class="n">sampling_resolution</span><span class="o">=</span><span class="default_value">100</span></em>, <em class="sig-param"><span class="n">gaussian_smear</span><span class="o">=</span><span class="default_value">0.05</span></em>, <em class="sig-param"><span class="n">species</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.Hybridization" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>quantify s/p/d/f orbital character and their hybridizations at band edges</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>decay_length (float in eV):</dt><dd><p>The dos is sampled by an exponential decay function. this parameter
sets the decay length of the exponential. Three times the decay
length corresponds to 10% sampling strength. There is a hard cutoff
at five times the decay length (1% sampling strength)</p>
</dd>
<dt>sampling_resolution (int):</dt><dd><p>Number of points to sample DOS</p>
</dd>
<dt>gaussian_smear (float in eV):</dt><dd><p>Gaussian smearing (sigma) around each sampled point in the DOS</p>
</dd>
<dt>species ([str]): the species for which orbital contributions are</dt><dd><p>separately returned.</p>
</dd>
</dl>
</dd>
<dt>Returns (featurize returns [float] and featurize_labels returns [str]):</dt><dd><p>set of orbitals contributions and hybridizations. If species, then also
individual contributions from given species. Examples:</p>
<blockquote>
<div><p>cbm_s (float): s-orbital character of the cbm up to energy_cutoff
vbm_sp (float): sp-hybridization at the vbm edge. Minimum is 0</p>
<blockquote>
<div><p>or no hybridization (e.g. all s or vbm_s==1) and 1.0 is
maximum hybridization (i.e. vbm_s==0.5, vbm_p==0.5)</p>
</div></blockquote>
<p>cbm_Si_p (float): p-orbital character of Si</p>
</div></blockquote>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.dos.Hybridization.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">decay_length</span><span class="o">=</span><span class="default_value">0.1</span></em>, <em class="sig-param"><span class="n">sampling_resolution</span><span class="o">=</span><span class="default_value">100</span></em>, <em class="sig-param"><span class="n">gaussian_smear</span><span class="o">=</span><span class="default_value">0.05</span></em>, <em class="sig-param"><span class="n">species</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.Hybridization.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.dos.Hybridization.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.Hybridization.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.dos.Hybridization.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.Hybridization.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns ([str]): feature names starting with the extrema (cbm or vbm)
followed by either s,p,d,f orbital to show normalized contribution
or a pair showing their hybridization or contribution of an element.
See the class docs for examples.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.dos.Hybridization.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dos</span></em>, <em class="sig-param"><span class="n">decay_length</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.Hybridization.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>takes in the density of state and return the orbitals contributions
and hybridizations.</p>
<dl>
<dt>Args:</dt><dd><p>dos (pymatgen CompleteDos): note that dos.structure is required
decay_length (float or None): if set, it overrides the instance</p>
<blockquote>
<div><p>variable self.decay_length.</p>
</div></blockquote>
</dd>
</dl>
<p>Returns ([float]): features, see class doc for more info</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.dos.Hybridization.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.Hybridization.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.dos.SiteDOS">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.dos.</code><code class="sig-name descname">SiteDOS</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">decay_length</span><span class="o">=</span><span class="default_value">0.1</span></em>, <em class="sig-param"><span class="n">sampling_resolution</span><span class="o">=</span><span class="default_value">100</span></em>, <em class="sig-param"><span class="n">gaussian_smear</span><span class="o">=</span><span class="default_value">0.05</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.SiteDOS" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>report the fractional s/p/d/f dos for a particular site. a CompleteDos
object is required because knowledge of the structure is needed. this
featurizer will work for metals as well as semiconductors. if the dos is a
semiconductor, cbm and vbm will correspond to the two respective band
edges. if the dos is a metal, then cbm and vbm correspond to above and
below the fermi level, respectively.</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>decay_length (float in eV):</dt><dd><p>the dos is sampled by an exponential decay function. this parameter
sets the decay length of the exponential. three times the
decay_length corresponds to 10% sampling strength. there is a hard
cutoff at five times the decay length (1% sampling strength)</p>
</dd>
<dt>sampling_resolution (int):</dt><dd><p>number of points to sample dos</p>
</dd>
<dt>gaussian_smear (float in eV):</dt><dd><p>Gaussian smearing (sigma) around each sampled point in dos</p>
</dd>
</dl>
</dd>
<dt>Returns (list of floats):</dt><dd><p>cbm_score_i (float): fractional score for i in {s,p,d,f}
cbm_score_total (float): the total sum of all the {s,p,d,f} scores</p>
<blockquote>
<div><p>this is useful information when comparing the relative
contributions from multiples sites</p>
</div></blockquote>
<p>vbm_score_i (float): fractional score for i in {s,p,d,f}
vbm_score_total (float): the total sum of all the {s,p,d,f} scores</p>
<blockquote>
<div><p>this is useful information when comparing the relative
contributions from multiples sites</p>
</div></blockquote>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.dos.SiteDOS.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">decay_length</span><span class="o">=</span><span class="default_value">0.1</span></em>, <em class="sig-param"><span class="n">sampling_resolution</span><span class="o">=</span><span class="default_value">100</span></em>, <em class="sig-param"><span class="n">gaussian_smear</span><span class="o">=</span><span class="default_value">0.05</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.SiteDOS.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.dos.SiteDOS.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.SiteDOS.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.dos.SiteDOS.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.SiteDOS.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Returns (list of str): list of names of the features. See the docs for</dt><dd><p>the featurizer class for more information.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.dos.SiteDOS.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dos</span></em>, <em class="sig-param"><span class="n">idx</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.SiteDOS.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>get dos scores for given site index</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>dos (pymatgen CompleteDos or their dict):</dt><dd><p>dos to featurize, must contain pdos and structure</p>
</dd>
</dl>
<p>idx (int): index of target site in structure.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.dos.SiteDOS.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.SiteDOS.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="matminer.featurizers.dos.get_cbm_vbm_scores">
<code class="sig-prename descclassname">matminer.featurizers.dos.</code><code class="sig-name descname">get_cbm_vbm_scores</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dos</span></em>, <em class="sig-param"><span class="n">decay_length</span></em>, <em class="sig-param"><span class="n">sampling_resolution</span></em>, <em class="sig-param"><span class="n">gaussian_smear</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.get_cbm_vbm_scores" title="Permalink to this definition">¶</a></dt>
<dd><p>Quantifies the contribution of all atomic orbitals (s/p/d/f) from all
crystal sites to the conduction band minimum (CBM) and the valence band
maximum (VBM). An exponential decay function is used to sample the DOS.
An example use may be sorting the output based on cbm_score or vbm_score.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>dos (pymatgen CompleteDos or their dict):</dt><dd><p>The density of states to featurize. Must be a complete DOS,
(i.e. contains PDOS and structure, in addition to total DOS)</p>
</dd>
<dt>decay_length (float in eV):</dt><dd><p>The dos is sampled by an exponential decay function. this parameter
sets the decay length of the exponential. Three times the decay
length corresponds to 10% sampling strength. There is a hard cutoff
at five times the decay length (1% sampling strength)</p>
</dd>
<dt>sampling_resolution (int):</dt><dd><p>Number of points to sample DOS</p>
</dd>
<dt>gaussian_smear (float in eV):</dt><dd><p>Gaussian smearing (sigma) around each sampled point in the DOS</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>orbital_scores [(dict)]:</dt><dd><p>A list of how much each orbital contributes to the partial
density of states near the band edge. Dictionary items are:
.. cbm_score: (float) fractional contribution to conduction band
.. vbm_score: (float) fractional contribution to valence band
.. species: (pymatgen Specie) the Specie of the orbital
.. character: (str) is the orbital character s, p, d, or f
.. location: [(float)] fractional coordinates of the orbital</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py function">
<dt id="matminer.featurizers.dos.get_site_dos_scores">
<code class="sig-prename descclassname">matminer.featurizers.dos.</code><code class="sig-name descname">get_site_dos_scores</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">dos</span></em>, <em class="sig-param"><span class="n">idx</span></em>, <em class="sig-param"><span class="n">decay_length</span></em>, <em class="sig-param"><span class="n">sampling_resolution</span></em>, <em class="sig-param"><span class="n">gaussian_smear</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.dos.get_site_dos_scores" title="Permalink to this definition">¶</a></dt>
<dd><p>Quantifies the contribution of all atomic orbitals (s/p/d/f) from a
particular crystal site to the conduction band minimum (CBM) and the
valence band maximum (VBM). An exponential decay function is used to sample
the DOS. if the dos is a metal, then CBM and VBM indicate the orbital
scores above and below the fermi energy, respectively.</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>dos (pymatgen CompleteDos or their dict):</dt><dd><p>The density of states to featurize. Must be a complete DOS,
(i.e. contains PDOS and structure, in addition to total DOS)</p>
</dd>
<dt>decay_length (float in eV):</dt><dd><p>The dos is sampled by an exponential decay function. this parameter
sets the decay length of the exponential. Three times the decay
length corresponds to 10% sampling strength. There is a hard cutoff
at five times the decay length (1% sampling strength)</p>
</dd>
<dt>sampling_resolution (int):</dt><dd><p>Number of points to sample DOS</p>
</dd>
<dt>gaussian_smear (float in eV):</dt><dd><p>Gaussian smearing (sigma) around each sampled point in the DOS</p>
</dd>
<dt>idx (int):</dt><dd><p>site index for which to gather dos s/p/d/f scores</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl>
<dt>orbital_scores (dict):</dt><dd><p>a dictionary of the fractional s/p/d/f orbital scores from the
total dos accumulated from that site. dictionary structure:</p>
<blockquote>
<div><dl class="simple">
<dt>{cbm: {s: (float), …, f: (float), total: (float)},</dt><dd><p>vbm: {s: (float), …, f: (float), total: (float)}}</p>
</dd>
</dl>
</div></blockquote>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-matminer.featurizers.function">
<span id="matminer-featurizers-function-module"></span><h2>matminer.featurizers.function module<a class="headerlink" href="#module-matminer.featurizers.function" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt id="matminer.featurizers.function.FunctionFeaturizer">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.function.</code><code class="sig-name descname">FunctionFeaturizer</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">expressions</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">multi_feature_depth</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">postprocess</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">combo_function</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">latexify_labels</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.function.FunctionFeaturizer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Features from functions applied to existing features, e.g. “1/x”</p>
<p>This featurizer must be fit either by calling .fit_featurize_dataframe
or by calling .fit followed by featurize_dataframe.</p>
<p>This class featurizes a dataframe according to a set
of expressions representing functions to apply to
existing features. The approach here has uses a sympy-based
parsing of string expressions, rather than explicit
python functions.  The primary reason this has been
done is to provide for better support for book-keeping
(e. g. with feature labels), substitution, and elimination
of symbolic redundancy, which sympy is well-suited for.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>expressions ([str]): list of sympy-parseable expressions</dt><dd><p>representing a function of a single variable x, e. g.
[“1 / x”, “x ** 2”], defaults to the list above</p>
</dd>
<dt>multi_feature_depth (int): how many features to include if using</dt><dd><p>multiple fields for functionalization, e. g. 2 will
include pairwise combined features</p>
</dd>
<dt>postprocess (function or type): type to cast functional outputs</dt><dd><p>to, if, for example, you want to include the possibility of
complex numbers in your outputs, use postprocess=np.complex,
defaults to float</p>
</dd>
<dt>combo_function (function): function to combine multi-features,</dt><dd><p>defaults to np.prod (i.e. cumulative product of expressions),
note that a combo function must cleanly process sympy
expressions and <strong>takes a list of arbitrary length as input</strong>,
other options include np.sum</p>
</dd>
<dt>latexify_labels (bool): whether to render labels in latex,</dt><dd><p>defaults to False</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.function.FunctionFeaturizer.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">expressions</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">multi_feature_depth</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">postprocess</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">combo_function</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">latexify_labels</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.function.FunctionFeaturizer.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.function.FunctionFeaturizer.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.function.FunctionFeaturizer.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.function.FunctionFeaturizer.exp_dict">
<em class="property">property </em><code class="sig-name descname">exp_dict</code><a class="headerlink" href="#matminer.featurizers.function.FunctionFeaturizer.exp_dict" title="Permalink to this definition">¶</a></dt>
<dd><p>Generates a dictionary of expressions keyed by number of
variables in each expression</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>Dictionary of expressions keyed by number of variables</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.function.FunctionFeaturizer.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.function.FunctionFeaturizer.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Returns:</dt><dd><p>Set of feature labels corresponding to expressions</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.function.FunctionFeaturizer.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">*</span><span class="n">args</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.function.FunctionFeaturizer.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, essentially iterates over all
of the functions in self.function_list to generate
features for each argument.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><a href="#id9"><span class="problematic" id="id10">*</span></a>args: list of numbers to generate functional output</dt><dd><p>features</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>list of functional outputs corresponding to input args</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.function.FunctionFeaturizer.fit">
<code class="sig-name descname">fit</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">X</span></em>, <em class="sig-param"><span class="n">y</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">fit_kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.function.FunctionFeaturizer.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Sets the feature labels.  Not intended to be used by a user,
only intended to be invoked as part of featurize_dataframe</p>
<dl class="simple">
<dt>Args:</dt><dd><p>X (DataFrame or array-like): data to fit to</p>
</dd>
<dt>Returns:</dt><dd><p>Set of feature labels corresponding to expressions</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.function.FunctionFeaturizer.generate_string_expressions">
<code class="sig-name descname">generate_string_expressions</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">input_variable_names</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.function.FunctionFeaturizer.generate_string_expressions" title="Permalink to this definition">¶</a></dt>
<dd><p>Method to generate string expressions for input strings,
mainly used to generate columns names for featurize_dataframe</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>input_variable_names ([str]): strings corresponding to</dt><dd><p>functional input variable names</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>List of string expressions generated by substitution of
variable names into functions</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.function.FunctionFeaturizer.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.function.FunctionFeaturizer.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="matminer.featurizers.function.generate_expressions_combinations">
<code class="sig-prename descclassname">matminer.featurizers.function.</code><code class="sig-name descname">generate_expressions_combinations</code><span class="sig-paren">(</span><em class="sig-param">expressions</em>, <em class="sig-param">combo_depth=2</em>, <em class="sig-param">combo_function=&lt;function prod&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.function.generate_expressions_combinations" title="Permalink to this definition">¶</a></dt>
<dd><p>This function takes a list of strings representing functions
of x, converts them to sympy expressions, and combines
them according to the combo_depth parameter.  Also filters
resultant expressions for any redundant ones determined
by sympy expression equivalence.</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>expressions (strings): all of the sympy-parseable strings</dt><dd><p>to be converted to expressions and combined, e. g.
[“1 / x”, “x ** 2”], must be functions of x</p>
</dd>
</dl>
<p>combo_depth (int): the number of independent variables to consider
combo_function (method): the function which combines the</p>
<blockquote>
<div><p>the respective expressions provided, defaults to np.prod,
i. e. the cumulative product of the expressions</p>
</div></blockquote>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>list of unique non-trivial expressions for featurization</dt><dd><p>of inputs</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-matminer.featurizers.site">
<span id="matminer-featurizers-site-module"></span><h2>matminer.featurizers.site module<a class="headerlink" href="#module-matminer.featurizers.site" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt id="matminer.featurizers.site.AGNIFingerprints">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.site.</code><code class="sig-name descname">AGNIFingerprints</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">directions</span><span class="o">=</span><span class="default_value">None, 'x', 'y', 'z'</span></em>, <em class="sig-param"><span class="n">etas</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">8</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AGNIFingerprints" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Product integral of RDF and Gaussian window function, from <a class="reference external" href="http://pubs.acs.org/doi/abs/10.1021/acs.jpcc.6b10908">Botu et al</a>.</p>
<p>Integral of the product of the radial distribution function and a
Gaussian window function. Originally used by
<a class="reference external" href="http://pubs.acs.org/doi/abs/10.1021/acs.jpcc.6b10908">Botu et al</a> to fit empiricial
potentials. These features come in two forms: atomic fingerprints and
direction-resolved fingerprints.
Atomic fingerprints describe the local environment of an atom and are
computed using the function:
<img class="math" src="_images/math/c833e66a1aa68dba7be89cbcaed7e749967ab5c9.png" alt="A_i(\eta) = \sum\limits_{i \ne j} e^{-(\frac{r_{ij}}{\eta})^2} f(r_{ij})"/>
where <img class="math" src="_images/math/5aa339d4daf45a810dda332e3c80a0698e526e04.png" alt="i"/> is the index of the atom, <img class="math" src="_images/math/e3fc28292267f066fee7718c64f4bbfece521f24.png" alt="j"/> is the index of a neighboring atom, <img class="math" src="_images/math/97db043c7cba573ac389b71add78b048077e8a13.png" alt="\eta"/> is a scaling function,
<img class="math" src="_images/math/774df2c68904c92b7c0db636d12ca2ca6ee21263.png" alt="r_{ij}"/> is the distance between atoms <img class="math" src="_images/math/5aa339d4daf45a810dda332e3c80a0698e526e04.png" alt="i"/> and <img class="math" src="_images/math/e3fc28292267f066fee7718c64f4bbfece521f24.png" alt="j"/>, and <img class="math" src="_images/math/079ba1f3373823268ba0e33b400423434ca7c0a6.png" alt="f(r)"/> is a cutoff function where
<img class="math" src="_images/math/e8c0d853d9741d1d80356648c449808dd3fdeb3c.png" alt="f(r) = 0.5[\cos(\frac{\pi r_{ij}}{R_c}) + 1]"/> if <img class="math" src="_images/math/0239bb38c93a6213c5ac6ba1dabd3e0a003ed371.png" alt="r &lt; R_c"/> and <img class="math" src="_images/math/31fdf41b39df23c95e52c5aef07f59d9adf82f3c.png" alt="0"/> otherwise.
The direction-resolved fingerprints are computed using
<img class="math" src="_images/math/cfca9af3a86afde1b9efe235bfc3e48f90890d82.png" alt="V_i^k(\eta) = \sum\limits_{i \ne j} \frac{r_{ij}^k}{r_{ij}} e^{-(\frac{r_{ij}}{\eta})^2} f(r_{ij})"/>
where <img class="math" src="_images/math/a9bb86de6492860286fd9038b2070f5870e63cb6.png" alt="r_{ij}^k"/> is the <img class="math" src="_images/math/c279824074ed8acc9cd963e930735f542462edec.png" alt="k^{th}"/> component of <img class="math" src="_images/math/5a53aff7886e588bf17ff209a8fa05351cda8798.png" alt="\bold{r}_i - \bold{r}_j"/>.
Parameters:
TODO: Differentiate between different atom types (maybe as another class)</p>
<dl class="py method">
<dt id="matminer.featurizers.site.AGNIFingerprints.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">directions</span><span class="o">=</span><span class="default_value">None, 'x', 'y', 'z'</span></em>, <em class="sig-param"><span class="n">etas</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">8</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AGNIFingerprints.__init__" title="Permalink to this definition">¶</a></dt>
<dd><dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>directions (iterable): List of directions for the fingerprints. Can</dt><dd><p>be one or more of ‘None`, ‘x’, ‘y’, or ‘z’</p>
</dd>
</dl>
<p>etas (iterable of floats): List of which window widths to compute
cutoff (float): Cutoff distance (Angstroms)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.AGNIFingerprints.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AGNIFingerprints.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.AGNIFingerprints.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AGNIFingerprints.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.AGNIFingerprints.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">struct</span></em>, <em class="sig-param"><span class="n">idx</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AGNIFingerprints.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, which has to be implemented
in any derived featurizer subclass.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).</p>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.AGNIFingerprints.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AGNIFingerprints.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.site.AngularFourierSeries">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.site.</code><code class="sig-name descname">AngularFourierSeries</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">bins</span></em>, <em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">10.0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AngularFourierSeries" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Compute the angular Fourier series (AFS), including both angular and radial info</p>
<p>The AFS is the product of pairwise distance function (g_n, g_n’) between two pairs
of atoms (sharing the common central site) and the cosine of the angle
between the two pairs. The AFS is a 2-dimensional feature (the axes are g_n,
g_n’).</p>
<p>Examples of distance functionals are square functions, Gaussian, trig
functions, and Bessel functions. An example for Gaussian:</p>
<blockquote>
<div><p>lambda d: exp( -(d - d_n)**2 ), where d_n is the coefficient for g_n</p>
</div></blockquote>
<p>See <a class="reference internal" href="matminer.featurizers.utils.html#module-matminer.featurizers.utils.grdf" title="matminer.featurizers.utils.grdf"><code class="xref py py-func docutils literal notranslate"><span class="pre">grdf()</span></code></a> for a full list of available binning functions.</p>
<dl class="simple">
<dt>There are two preset conditions:</dt><dd><p>gaussian: bin functions are gaussians
histogram: bin functions are rectangular functions</p>
</dd>
<dt>Features:</dt><dd><p>AFS ([gn], [gn’]) - Angular Fourier Series between binning functions (g1 and g2)</p>
</dd>
<dt>Args:</dt><dd><dl class="simple">
<dt>bins:   ([AbstractPairwise]) a list of binning functions that</dt><dd><p>implement the AbstractPairwise base class</p>
</dd>
<dt>cutoff: (float) maximum distance to look for neighbors. The</dt><dd><p>featurizer will run slowly for large distance cutoffs
because of the number of neighbor pairs scales as
the square of the number of neighbors</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.site.AngularFourierSeries.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">bins</span></em>, <em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">10.0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AngularFourierSeries.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.AngularFourierSeries.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AngularFourierSeries.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.AngularFourierSeries.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AngularFourierSeries.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.AngularFourierSeries.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">struct</span></em>, <em class="sig-param"><span class="n">idx</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AngularFourierSeries.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get AFS of the input structure.
Args:</p>
<blockquote>
<div><p>struct (Structure): Pymatgen Structure object.
idx (int): index of target site in structure struct.</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>Flattened list of AFS values. the list order is:</dt><dd><p>g_n g_n’</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.AngularFourierSeries.from_preset">
<em class="property">static </em><code class="sig-name descname">from_preset</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">preset</span></em>, <em class="sig-param"><span class="n">width</span><span class="o">=</span><span class="default_value">0.5</span></em>, <em class="sig-param"><span class="n">spacing</span><span class="o">=</span><span class="default_value">0.5</span></em>, <em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">10</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AngularFourierSeries.from_preset" title="Permalink to this definition">¶</a></dt>
<dd><dl>
<dt>Preset bin functions for this featurizer. Example use:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">AFS</span> <span class="o">=</span> <span class="n">AngularFourierSeries</span><span class="o">.</span><span class="n">from_preset</span><span class="p">(</span><span class="s1">&#39;gaussian&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">AFS</span><span class="o">.</span><span class="n">featurize</span><span class="p">(</span><span class="n">struct</span><span class="p">,</span> <span class="n">idx</span><span class="p">)</span>
</pre></div>
</div>
</dd>
<dt>Args:</dt><dd><p>preset (str): shape of bin (either ‘gaussian’ or ‘histogram’)
width (float): bin width. std dev for gaussian, width for histogram
spacing (float): the spacing between bin centers
cutoff (float): maximum distance to look for neighbors</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.AngularFourierSeries.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AngularFourierSeries.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.site.AverageBondAngle">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.site.</code><code class="sig-name descname">AverageBondAngle</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">method</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AverageBondAngle" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Determines the average bond angles of a specific site with
its nearest neighbors using one of pymatgen’s NearNeighbor
classes. Neighbors that are adjacent to each other are stored
and angle between them are computed. ‘Average bond angle’ of
a site is the mean bond angle between all its nearest neighbors.</p>
<dl class="py method">
<dt id="matminer.featurizers.site.AverageBondAngle.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">method</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AverageBondAngle.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize featurizer</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>method (NearNeighbor) - subclass under NearNeighbor used to compute nearest</dt><dd><p>neighbors</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.AverageBondAngle.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AverageBondAngle.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.AverageBondAngle.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AverageBondAngle.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.AverageBondAngle.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">strc</span></em>, <em class="sig-param"><span class="n">idx</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AverageBondAngle.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get average bond length of a site and all its nearest
neighbors.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>strc (Structure): Pymatgen Structure object
idx (int): index of target site in structure object</p>
</dd>
<dt>Returns:</dt><dd><p>average bond length (list)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.AverageBondAngle.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AverageBondAngle.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.site.AverageBondLength">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.site.</code><code class="sig-name descname">AverageBondLength</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">method</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AverageBondLength" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Determines the average bond length between one specific site
and all its nearest neighbors using one of pymatgen’s NearNeighbor
classes. These nearest neighbor calculators return weights related
to the proximity of each neighbor to this site. ‘Average bond
length’ of a site is the weighted average of the distance between
site and all its nearest neighbors.</p>
<dl class="py method">
<dt id="matminer.featurizers.site.AverageBondLength.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">method</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AverageBondLength.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize featurizer</p>
<dl class="simple">
<dt>Args:</dt><dd><p>method (NearNeighbor) - subclass under NearNeighbor used to compute nearest neighbors</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.AverageBondLength.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AverageBondLength.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.AverageBondLength.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AverageBondLength.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.AverageBondLength.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">strc</span></em>, <em class="sig-param"><span class="n">idx</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AverageBondLength.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get weighted average bond length of a site and all its nearest
neighbors.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>strc (Structure): Pymatgen Structure object
idx (int): index of target site in structure object</p>
</dd>
<dt>Returns:</dt><dd><p>average bond length (list)</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.AverageBondLength.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.AverageBondLength.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.site.BondOrientationalParameter">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.site.</code><code class="sig-name descname">BondOrientationalParameter</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">max_l</span><span class="o">=</span><span class="default_value">10</span></em>, <em class="sig-param"><span class="n">compute_w</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">compute_w_hat</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.BondOrientationalParameter" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Averages of spherical harmonics of local neighbors</p>
<p>Bond Orientational Parameters (BOPs) describe the local environment around an atom by
considering the local symmetry of the bonds as computed using spherical harmonics.
To create descriptors that are invariant to rotating the coordinate system, we use the
average of all spherical harmonics of a certain degree - following the approach of
<a class="reference external" href="https://link.aps.org/doi/10.1103/PhysRevB.28.784">Steinhardt et al.</a>.
We weigh the contributions of each neighbor with the solid angle of the Voronoi tessellation
(see <cite>Mickel et al. &lt;https://aip.scitation.org/doi/abs/10.1063/1.4774084&gt;_</cite> for further
discussion). The weighing scheme makes these descriptors vary smoothly with small distortions
of a crystal structure.</p>
<p>In addition to the average spherical harmonics, this class can also compute the <img class="math" src="_images/math/1fbee781f84569077719a167b64e12064360fac1.png" alt="W"/> and
<img class="math" src="_images/math/a6c088a730612710750ee31f9cc780d5f94f9989.png" alt="\hat{W}"/> parameters proposed by <a class="reference external" href="https://link.aps.org/doi/10.1103/PhysRevB.28.784">Steinhardt et al.</a>.</p>
<dl class="simple">
<dt>Attributes:</dt><dd><p>BOOP Q l=&lt;n&gt; - Average spherical harmonic for a certain degree, n.
BOOP W l=&lt;n&gt; - W parameter for a certain degree of spherical harmonic, n.
BOOP What l=&lt;n&gt; - <img class="math" src="_images/math/a6c088a730612710750ee31f9cc780d5f94f9989.png" alt="\hat{W}"/> parameter for a certain degree of spherical harmonic, n.</p>
</dd>
<dt>References:</dt><dd><p><a class="reference external" href="https://link.aps.org/doi/10.1103/PhysRevB.28.784">Steinhardt et al., _PRB_ (1983)</a>
<a class="reference external" href="http://link.aps.org/doi/10.1103/PhysRevB.95.144110">Seko et al., _PRB_ (2017)</a></p>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.site.BondOrientationalParameter.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">max_l</span><span class="o">=</span><span class="default_value">10</span></em>, <em class="sig-param"><span class="n">compute_w</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">compute_w_hat</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.BondOrientationalParameter.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the featurizer</p>
<dl class="simple">
<dt>Args:</dt><dd><p>max_l (int) - Maximum spherical harmonic to consider
compute_w (bool) - Whether to compute Ws as well
compute_w_hat (bool) - Whether to compute What</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.BondOrientationalParameter.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.BondOrientationalParameter.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.BondOrientationalParameter.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.BondOrientationalParameter.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.BondOrientationalParameter.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">strc</span></em>, <em class="sig-param"><span class="n">idx</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.BondOrientationalParameter.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, which has to be implemented
in any derived featurizer subclass.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).</p>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.BondOrientationalParameter.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.BondOrientationalParameter.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.site.ChemEnvSiteFingerprint">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.site.</code><code class="sig-name descname">ChemEnvSiteFingerprint</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cetypes</span></em>, <em class="sig-param"><span class="n">strategy</span></em>, <em class="sig-param"><span class="n">geom_finder</span></em>, <em class="sig-param"><span class="n">max_csm</span><span class="o">=</span><span class="default_value">8</span></em>, <em class="sig-param"><span class="n">max_dist_fac</span><span class="o">=</span><span class="default_value">1.41</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.ChemEnvSiteFingerprint" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Resemblance of given sites to ideal environments</p>
<p>Site fingerprint computed from pymatgen’s ChemEnv package
that provides resemblance percentages of a given site
to ideal environments.
Args:</p>
<blockquote>
<div><dl class="simple">
<dt>cetypes ([str]): chemical environments (CEs) to be</dt><dd><p>considered.</p>
</dd>
</dl>
<p>strategy (ChemenvStrategy): ChemEnv neighbor-finding strategy.
geom_finder (LocalGeometryFinder): ChemEnv local geometry finder.
max_csm (float): maximum continuous symmetry measure (CSM;</p>
<blockquote>
<div><p>default of 8 taken from chemenv). Note that any CSM
larger than max_csm will be set to max_csm in order
to avoid negative values (i.e., all features are
constrained to be between 0 and 1).</p>
</div></blockquote>
<p>max_dist_fac (float): maximum distance factor (default: 1.41).</p>
</div></blockquote>
<dl class="py method">
<dt id="matminer.featurizers.site.ChemEnvSiteFingerprint.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cetypes</span></em>, <em class="sig-param"><span class="n">strategy</span></em>, <em class="sig-param"><span class="n">geom_finder</span></em>, <em class="sig-param"><span class="n">max_csm</span><span class="o">=</span><span class="default_value">8</span></em>, <em class="sig-param"><span class="n">max_dist_fac</span><span class="o">=</span><span class="default_value">1.41</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.ChemEnvSiteFingerprint.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.ChemEnvSiteFingerprint.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.ChemEnvSiteFingerprint.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.ChemEnvSiteFingerprint.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.ChemEnvSiteFingerprint.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.ChemEnvSiteFingerprint.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">struct</span></em>, <em class="sig-param"><span class="n">idx</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.ChemEnvSiteFingerprint.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get ChemEnv fingerprint of site with given index in input
structure.
Args:</p>
<blockquote>
<div><p>struct (Structure): Pymatgen Structure object.
idx (int): index of target site in structure struct.</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(numpy array): resemblance fraction of target site to ideal</dt><dd><p>local environments.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.ChemEnvSiteFingerprint.from_preset">
<em class="property">static </em><code class="sig-name descname">from_preset</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">preset</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.ChemEnvSiteFingerprint.from_preset" title="Permalink to this definition">¶</a></dt>
<dd><p>Use a standard collection of CE types and
choose your ChemEnv neighbor-finding strategy.
Args:</p>
<blockquote>
<div><dl class="simple">
<dt>preset (str): preset types (“simple” or</dt><dd><p>“multi_weights”).</p>
</dd>
</dl>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>ChemEnvSiteFingerprint object from a preset.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.ChemEnvSiteFingerprint.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.ChemEnvSiteFingerprint.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.site.ChemicalSRO">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.site.</code><code class="sig-name descname">ChemicalSRO</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">nn</span></em>, <em class="sig-param"><span class="n">includes</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">excludes</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">sort</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.ChemicalSRO" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Chemical short range ordering, deviation of local site and nominal structure compositions</p>
<p>Chemical SRO features to evaluate the deviation
of local chemistry with the nominal composition of the structure.</p>
<p>A local bonding preference is computed using
f_el = N_el/(sum of N_el) - c_el,
where N_el is the number of each element type in the neighbors around
the target site, sum of N_el is the sum of all possible element types
(coordination number), and c_el is the composition of the specific
element in the entire structure.
A positive f_el indicates the “bonding” with the specific element
is favored, at least in the target site;
A negative f_el indicates the “bonding” is not favored, at least
in the target site.</p>
<p>Note that ChemicalSRO is only featurized for elements identified by
“fit” (see following), thus “fit” must be called before “featurize”,
or else an error will be raised.</p>
<dl class="simple">
<dt>Features:</dt><dd><dl class="simple">
<dt>CSRO__[nn method]_[element] - The Chemical SRO of a site computed based</dt><dd><p>on neighbors determined with a certain  NN-detection method for
a certain element.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.site.ChemicalSRO.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">nn</span></em>, <em class="sig-param"><span class="n">includes</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">excludes</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">sort</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.ChemicalSRO.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the featurizer</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>nn (NearestNeighbor): instance of one of pymatgen’s NearestNeighbor</dt><dd><p>classes.</p>
</dd>
</dl>
<p>includes (array-like or str): elements included to calculate CSRO.
excludes (array-like or str): elements excluded to calculate CSRO.
sort (bool): whether to sort elements by mendeleev number.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.ChemicalSRO.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.ChemicalSRO.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.ChemicalSRO.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.ChemicalSRO.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.ChemicalSRO.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">struct</span></em>, <em class="sig-param"><span class="n">idx</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.ChemicalSRO.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get CSRO features of site with given index in input structure.
Args:</p>
<blockquote>
<div><p>struct (Structure): Pymatgen Structure object.
idx (int): index of target site in structure.</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>(list of floats): Chemical SRO features for each element.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.ChemicalSRO.fit">
<code class="sig-name descname">fit</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">X</span></em>, <em class="sig-param"><span class="n">y</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.ChemicalSRO.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Identify elements to be included in the following featurization,
by intersecting the elements present in the passed structures with
those explicitly included (or excluded) in __init__. Only elements
in the <a href="#id38"><span class="problematic" id="id39">self.el_list_</span></a> will be featurized.
Besides, compositions of the passed structures will also be “stored”
in a dict of <a href="#id40"><span class="problematic" id="id41">self.el_amt_dict_</span></a>, avoiding repeated calculation of
composition when featurizing multiple sites in the same structure.
Args:</p>
<blockquote>
<div><dl>
<dt>X (array-like): containing Pymatgen structures and sites, supports</dt><dd><p>multiple choices:
-2D array-like object:</p>
<blockquote>
<div><dl class="simple">
<dt>e.g. [[struct, site], [struct, site], …]</dt><dd><p>np.array([[struct, site], [struct, site], …])</p>
</dd>
</dl>
</div></blockquote>
<dl class="simple">
<dt>-Pandas dataframe:</dt><dd><p>e.g. df[[‘struct’, ‘site’]]</p>
</dd>
</dl>
</dd>
</dl>
<p>y : unused (added for consistency with overridden method signature)</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>self</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.ChemicalSRO.from_preset">
<em class="property">static </em><code class="sig-name descname">from_preset</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">preset</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.ChemicalSRO.from_preset" title="Permalink to this definition">¶</a></dt>
<dd><p>Use one of the standard instances of a given NearNeighbor class.
Args:</p>
<blockquote>
<div><dl class="simple">
<dt>preset (str): preset type (“VoronoiNN”, “JmolNN”,</dt><dd><p>“MiniumDistanceNN”, “MinimumOKeeffeNN”,
or “MinimumVIRENN”).</p>
</dd>
</dl>
<p><a href="#id13"><span class="problematic" id="id14">**</span></a>kwargs: allow to pass args to the NearNeighbor class.</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>ChemicalSRO from a preset.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.ChemicalSRO.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.ChemicalSRO.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.site.CoordinationNumber">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.site.</code><code class="sig-name descname">CoordinationNumber</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">nn</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">use_weights</span><span class="o">=</span><span class="default_value">'none'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.CoordinationNumber" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Number of first nearest neighbors of a site.</p>
<p>Determines the number of nearest neighbors of a site using one of
pymatgen’s NearNeighbor classes. These nearest neighbor calculators
can return weights related to the proximity of each neighbor to this
site. It is possible to take these weights into account to prevent
the coordination number from changing discontinuously with small
perturbations of a structure, either by summing the total weights
or using the normalization method presented by
[Ward et al.](<a class="reference external" href="http://link.aps.org/doi/10.1103/PhysRevB.96.014107">http://link.aps.org/doi/10.1103/PhysRevB.96.014107</a>)</p>
<dl class="simple">
<dt>Features:</dt><dd><dl class="simple">
<dt>CN_[method] - Coordination number computed using a certain method</dt><dd><p>for calculating nearest neighbors.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.site.CoordinationNumber.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">nn</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">use_weights</span><span class="o">=</span><span class="default_value">'none'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.CoordinationNumber.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the featurizer</p>
<dl>
<dt>Args:</dt><dd><p>nn (NearestNeighbor) - Method used to determine coordination number
use_weights (string) - Method used to account for weights of neighbors:</p>
<blockquote>
<div><p>‘none’ - Do not use weights when computing coordination number
‘sum’ - Use sum of weights as the coordination number
‘effective’ - Compute the ‘effective coordination number’, which</p>
<blockquote>
<div><p>is computed as <img class="math" src="_images/math/d188c978e0d95e667be9c33955b35c80ec4c786f.png" alt="\frac{(\sum_n w_n)^2)}{\sum_n w_n^2}"/></p>
</div></blockquote>
</div></blockquote>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.CoordinationNumber.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.CoordinationNumber.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.CoordinationNumber.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.CoordinationNumber.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.CoordinationNumber.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">struct</span></em>, <em class="sig-param"><span class="n">idx</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.CoordinationNumber.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get coordintion number of site with given index in input
structure.
Args:</p>
<blockquote>
<div><p>struct (Structure): Pymatgen Structure object.
idx (int): index of target site in structure struct.</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>[float] - Coordination number</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.CoordinationNumber.from_preset">
<em class="property">static </em><code class="sig-name descname">from_preset</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">preset</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.CoordinationNumber.from_preset" title="Permalink to this definition">¶</a></dt>
<dd><p>Use one of the standard instances of a given NearNeighbor class.
Args:</p>
<blockquote>
<div><dl class="simple">
<dt>preset (str): preset type (“VoronoiNN”, “JmolNN”,</dt><dd><p>“MiniumDistanceNN”, “MinimumOKeeffeNN”,
or “MinimumVIRENN”).</p>
</dd>
</dl>
<p><a href="#id15"><span class="problematic" id="id16">**</span></a>kwargs: allow to pass args to the NearNeighbor class.</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>CoordinationNumber from a preset.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.CoordinationNumber.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.CoordinationNumber.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.site.CrystalNNFingerprint">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.site.</code><code class="sig-name descname">CrystalNNFingerprint</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">op_types</span></em>, <em class="sig-param"><span class="n">chem_info</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.CrystalNNFingerprint" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>A local order parameter fingerprint for periodic crystals.</p>
<p>The fingerprint represents the value of various order parameters for the
site. The “wt” order parameter describes how consistent a site is with a
certain coordination number. The remaining order parameters are computed
by multiplying the “wt” for that coordination number with the OP value.</p>
<p>The chem_info parameter can be used to also get chemical descriptors that
describe differences in some chemical parameter (e.g., electronegativity)
between the central site and the site neighbors.</p>
<dl class="py method">
<dt id="matminer.featurizers.site.CrystalNNFingerprint.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">op_types</span></em>, <em class="sig-param"><span class="n">chem_info</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.CrystalNNFingerprint.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the CrystalNNFingerprint. Use the from_preset() function to
use default params.
Args:</p>
<blockquote>
<div><dl class="simple">
<dt>op_types (dict): a dict of coordination number (int) to a list of str</dt><dd><p>representing the order parameter types</p>
</dd>
<dt>chem_info (dict): a dict of chemical properties (e.g., atomic mass)</dt><dd><p>to dictionaries that map an element to a value
(e.g., chem_info[“Pauling scale”][“O”] = 3.44)</p>
</dd>
</dl>
<p><a href="#id17"><span class="problematic" id="id18">**</span></a>kwargs: other settings to be passed into CrystalNN class</p>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.CrystalNNFingerprint.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.CrystalNNFingerprint.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.CrystalNNFingerprint.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.CrystalNNFingerprint.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.CrystalNNFingerprint.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">struct</span></em>, <em class="sig-param"><span class="n">idx</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.CrystalNNFingerprint.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get crystal fingerprint of site with given index in input
structure.
Args:</p>
<blockquote>
<div><p>struct (Structure): Pymatgen Structure object.
idx (int): index of target site in structure.</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>list of weighted order parameters of target site.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.CrystalNNFingerprint.from_preset">
<em class="property">static </em><code class="sig-name descname">from_preset</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">preset</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.CrystalNNFingerprint.from_preset" title="Permalink to this definition">¶</a></dt>
<dd><p>Use preset parameters to get the fingerprint
Args:</p>
<blockquote>
<div><p>preset (str): name of preset (“cn” or “ops”)
<a href="#id19"><span class="problematic" id="id20">**</span></a>kwargs: other settings to be passed into CrystalNN class</p>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.CrystalNNFingerprint.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.CrystalNNFingerprint.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.site.EwaldSiteEnergy">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.site.</code><code class="sig-name descname">EwaldSiteEnergy</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">accuracy</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.EwaldSiteEnergy" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Compute site energy from Coulombic interactions</p>
<dl class="simple">
<dt>User notes:</dt><dd><ul class="simple">
<li><p>This class uses that <cite>charges that are already-defined for the structure</cite>.</p></li>
<li><p>Ewald summations can be expensive. If you evaluating every site in many
large structures, run all of the sites for each structure at the same time.
We cache the Ewald result for the structure that was run last, so looping
over sites and then structures is faster than structures than sites.</p></li>
</ul>
</dd>
<dt>Features:</dt><dd><p>ewald_site_energy - Energy for the site computed from Coulombic interactions</p>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.site.EwaldSiteEnergy.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">accuracy</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.EwaldSiteEnergy.__init__" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Args:</dt><dd><p>accuracy (int): Accuracy of Ewald summation, number of decimal places</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.EwaldSiteEnergy.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.EwaldSiteEnergy.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.EwaldSiteEnergy.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.EwaldSiteEnergy.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.EwaldSiteEnergy.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">strc</span></em>, <em class="sig-param"><span class="n">idx</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.EwaldSiteEnergy.featurize" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Args:</dt><dd><p>struct (Structure): Pymatgen Structure object.
idx (int): index of target site in structure.</p>
</dd>
<dt>Returns:</dt><dd><p>([float]) - Electrostatic energy of the site</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.EwaldSiteEnergy.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.EwaldSiteEnergy.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.site.GaussianSymmFunc">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.site.</code><code class="sig-name descname">GaussianSymmFunc</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">etas_g2</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">etas_g4</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">zetas_g4</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">gammas_g4</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">6.5</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.GaussianSymmFunc" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Gaussian symmetry function features suggested by Behler et al.</p>
<p>The function is based on pair distances and angles, to approximate the functional
dependence of local energies, originally used in the fitting of
machine-learning potentials.
The symmetry functions can be divided to a set of radial functions
(g2 function), and a set of angular functions (g4 function).
The number of symmetry functions returned are based on parameters
of etas_g2, etas_g4, zetas_g4 and gammas_g4.
See the original papers for more details:
“Atom-centered symmetry functions for constructing high-dimensional
neural network potentials”, J Behler, J Chem Phys 134, 074106 (2011).
The cutoff function is taken as the polynomial form (cosine_cutoff)
to give a smoothed truncation.
A Fortran and a different Python version can be found in the code
Amp: Atomistic Machine-learning Package
(<a class="reference external" href="https://bitbucket.org/andrewpeterson/amp">https://bitbucket.org/andrewpeterson/amp</a>).
Args:</p>
<blockquote>
<div><dl class="simple">
<dt>etas_g2 (list of floats): etas used in radial functions.</dt><dd><p>(default: [0.05, 4., 20., 80.])</p>
</dd>
<dt>etas_g4 (list of floats): etas used in angular functions.</dt><dd><p>(default: [0.005])</p>
</dd>
<dt>zetas_g4 (list of floats): zetas used in angular functions.</dt><dd><p>(default: [1., 4.])</p>
</dd>
<dt>gammas_g4 (list of floats): gammas used in angular functions.</dt><dd><p>(default: [+1., -1.])</p>
</dd>
</dl>
<p>cutoff (float): cutoff distance. (default: 6.5)</p>
</div></blockquote>
<dl class="py method">
<dt id="matminer.featurizers.site.GaussianSymmFunc.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">etas_g2</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">etas_g4</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">zetas_g4</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">gammas_g4</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">6.5</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.GaussianSymmFunc.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.GaussianSymmFunc.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.GaussianSymmFunc.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.GaussianSymmFunc.cosine_cutoff">
<em class="property">static </em><code class="sig-name descname">cosine_cutoff</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">rs</span></em>, <em class="sig-param"><span class="n">cutoff</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.GaussianSymmFunc.cosine_cutoff" title="Permalink to this definition">¶</a></dt>
<dd><p>Polynomial cutoff function to give a smoothed truncation of the Gaussian
symmetry functions.
Args:</p>
<blockquote>
<div><p>rs (ndarray): distances to elements
cutoff (float): cutoff distance.</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>(ndarray) cutoff function.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.GaussianSymmFunc.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.GaussianSymmFunc.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.GaussianSymmFunc.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">struct</span></em>, <em class="sig-param"><span class="n">idx</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.GaussianSymmFunc.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get Gaussian symmetry function features of site with given index
in input structure.
Args:</p>
<blockquote>
<div><p>struct (Structure): Pymatgen Structure object.
idx (int): index of target site in structure.</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>(list of floats): Gaussian symmetry function features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.GaussianSymmFunc.g2">
<em class="property">static </em><code class="sig-name descname">g2</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">eta</span></em>, <em class="sig-param"><span class="n">rs</span></em>, <em class="sig-param"><span class="n">cutoff</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.GaussianSymmFunc.g2" title="Permalink to this definition">¶</a></dt>
<dd><p>Gaussian radial symmetry function of the center atom,
given an eta parameter.
Args:</p>
<blockquote>
<div><p>eta: radial function parameter.
rs: distances from the central atom to each neighbor
cutoff (float): cutoff distance.</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>(float) Gaussian radial symmetry function.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.GaussianSymmFunc.g4">
<em class="property">static </em><code class="sig-name descname">g4</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">etas</span></em>, <em class="sig-param"><span class="n">zetas</span></em>, <em class="sig-param"><span class="n">gammas</span></em>, <em class="sig-param"><span class="n">neigh_dist</span></em>, <em class="sig-param"><span class="n">neigh_coords</span></em>, <em class="sig-param"><span class="n">cutoff</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.GaussianSymmFunc.g4" title="Permalink to this definition">¶</a></dt>
<dd><p>Gaussian angular symmetry function of the center atom,
given a set of eta, zeta and gamma parameters.
Args:</p>
<blockquote>
<div><p>eta ([float]): angular function parameters.
zeta ([float]): angular function parameters.
gamma ([float]): angular function parameters.
neigh_coords (list of [floats]): coordinates of neighboring atoms, with respect</p>
<blockquote>
<div><p>to the central atom</p>
</div></blockquote>
<p>cutoff (float): cutoff parameter.</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>(float) Gaussian angular symmetry function for all combinations of eta, zeta, gamma</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.GaussianSymmFunc.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.GaussianSymmFunc.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.site.GeneralizedRadialDistributionFunction">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.site.</code><code class="sig-name descname">GeneralizedRadialDistributionFunction</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">bins</span></em>, <em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">20.0</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'GRDF'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.GeneralizedRadialDistributionFunction" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Compute the general radial distribution function (GRDF) for a site.</p>
<p>The GRDF is a radial measure of crystal order around a site. There are two
featurizing modes:</p>
<ol class="arabic simple">
<li><dl class="simple">
<dt>GRDF: (recommended) - n_bins length vector</dt><dd><p>In GRDF mode, The GRDF is computed by considering all sites around a
central site (i.e., no sites are omitted when computing the GRDF). The
features output from this mode will be vectors with length n_bins.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>pairwise GRDF: (advanced users) - n_bins x n_sites matrix</dt><dd><p>In this mode, GRDFs are are still computed around a central site, but
only one other site (and their translational equivalents) are used to
compute a GRDF (e.g. site 1 with site 2 and the translational
equivalents of site 2). This results in a a n_sites x n_bins matrix of
features. Requires <cite>fit</cite> for determining the max number of sites for</p>
</dd>
</dl>
</li>
</ol>
<p>The GRDF is a generalization of the partial radial distribution function
(PRDF). In contrast with the PRDF, the bins of the GRDF are not mutually-
exclusive and need not carry a constant weight of 1. The PRDF is a case of
the GRDF when the bins are rectangular functions. Examples of other
functions to use with the GRDF are Gaussian, trig, and Bessel functions.</p>
<p>See <a class="reference internal" href="matminer.featurizers.utils.html#module-matminer.featurizers.utils.grdf" title="matminer.featurizers.utils.grdf"><code class="xref py py-func docutils literal notranslate"><span class="pre">grdf()</span></code></a> for a full list of available binning functions.</p>
<dl>
<dt>There are two preset conditions:</dt><dd><p>gaussian: bin functions are gaussians
histogram: bin functions are rectangular functions</p>
</dd>
<dt>Args:</dt><dd><dl class="simple">
<dt>bins:   ([AbstractPairwise]) List of pairwise binning functions. Each of these functions</dt><dd><p>must implement the AbstractPairwise class.</p>
</dd>
</dl>
<p>cutoff: (float) maximum distance to look for neighbors
mode:   (str) the featurizing mode. supported options are:</p>
<blockquote>
<div><p>‘GRDF’ and ‘pairwise_GRDF’</p>
</div></blockquote>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.site.GeneralizedRadialDistributionFunction.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">bins</span></em>, <em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">20.0</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'GRDF'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.GeneralizedRadialDistributionFunction.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.GeneralizedRadialDistributionFunction.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.GeneralizedRadialDistributionFunction.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.GeneralizedRadialDistributionFunction.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.GeneralizedRadialDistributionFunction.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.GeneralizedRadialDistributionFunction.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">struct</span></em>, <em class="sig-param"><span class="n">idx</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.GeneralizedRadialDistributionFunction.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get GRDF of the input structure.
Args:</p>
<blockquote>
<div><p>struct (Structure): Pymatgen Structure object.
idx (int): index of target site in structure struct.</p>
</div></blockquote>
<dl>
<dt>Returns:</dt><dd><dl class="simple">
<dt>Flattened list of GRDF values. For each run mode the list order is:</dt><dd><p>GRDF:          bin#
pairwise GRDF: site2# bin#</p>
</dd>
</dl>
<p>The site2# corresponds to a pymatgen site index and bin#
corresponds to one of the bin functions</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.GeneralizedRadialDistributionFunction.fit">
<code class="sig-name descname">fit</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">X</span></em>, <em class="sig-param"><span class="n">y</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">fit_kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.GeneralizedRadialDistributionFunction.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Determine the maximum number of sites in X to assign correct feature
labels</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>X - [list of tuples], training data</dt><dd><p>tuple values should be (struc, idx)</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><p>self</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.GeneralizedRadialDistributionFunction.from_preset">
<em class="property">static </em><code class="sig-name descname">from_preset</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">preset</span></em>, <em class="sig-param"><span class="n">width</span><span class="o">=</span><span class="default_value">1.0</span></em>, <em class="sig-param"><span class="n">spacing</span><span class="o">=</span><span class="default_value">1.0</span></em>, <em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">10</span></em>, <em class="sig-param"><span class="n">mode</span><span class="o">=</span><span class="default_value">'GRDF'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.GeneralizedRadialDistributionFunction.from_preset" title="Permalink to this definition">¶</a></dt>
<dd><dl>
<dt>Preset bin functions for this featurizer. Example use:</dt><dd><div class="doctest highlight-default notranslate"><div class="highlight"><pre><span></span><span class="gp">&gt;&gt;&gt; </span><span class="n">GRDF</span> <span class="o">=</span> <span class="n">GeneralizedRadialDistributionFunction</span><span class="o">.</span><span class="n">from_preset</span><span class="p">(</span><span class="s1">&#39;gaussian&#39;</span><span class="p">)</span>
<span class="gp">&gt;&gt;&gt; </span><span class="n">GRDF</span><span class="o">.</span><span class="n">featurize</span><span class="p">(</span><span class="n">struct</span><span class="p">,</span> <span class="n">idx</span><span class="p">)</span>
</pre></div>
</div>
</dd>
<dt>Args:</dt><dd><p>preset (str): shape of bin (either ‘gaussian’ or ‘histogram’)
width (float): bin width. std dev for gaussian, width for histogram
spacing (float): the spacing between bin centers
cutoff (float): maximum distance to look for neighbors
mode (str): featurizing mode. either ‘GRDF’ or ‘pairwise_GRDF’</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.GeneralizedRadialDistributionFunction.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.GeneralizedRadialDistributionFunction.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.site.IntersticeDistribution">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.site.</code><code class="sig-name descname">IntersticeDistribution</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">6.5</span></em>, <em class="sig-param"><span class="n">interstice_types</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">stats</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">radius_type</span><span class="o">=</span><span class="default_value">'MiracleRadius'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.IntersticeDistribution" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Interstice distribution in the neighboring cluster around an atom site.</p>
<p>The interstices are categorized to distance, area and volume interstices.
Each of these metrics is a measures of the relative amount of empty space
around each atom as determined using atomic sphere models. The distance
interstice is the fraction of a bonding line unoccupied by the atom spheres;
The area interstice is the unoccupied area within the triangulated surface
formed by atom triplets in convex hull formed by neighbors, and the volume
interstice is the unoccupied portion of a tetrahedron formed between the
central atom and neighbor atom triplets. Please refer to the original paper
for more details (Wang et al. Nat Commun 10, 5537 (2019))</p>
<p>For amorphous alloys (metallic glasses), the coordination environments are
anisotropic, which can be reflected in the inequality of the interstices
present around an atom. To describe the anisotropy, here we derive statistics
of the interstices to featurize the interstice distribution around the atom.
Other methods can be grouping the interstices into histogram grids of fixed
bins and the features are then a vector of the values of the histograms.</p>
<p>User note:
This class is particularly designed for featuring the site-specific packing
heterogeneity in metallic glasses, especially the all-metallic-element ones.
If non-metallic-elements are present in the structures, the interstice
estimates may have larger deviation from actual values (despite this
deviation is systematic and thus the interstice estimates can still be
used to represent the packing heterogeneity).</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>cutoff (float): cutoff distance in determining the potential</dt><dd><p>neighbors for Voronoi tessellation analysis. (default: 6.5)</p>
</dd>
<dt>interstice_types (str or [str]): interstice distribution types,</dt><dd><p>support sub-list of [‘dist’, ‘area’, ‘vol’].</p>
</dd>
</dl>
<p>stats ([str]): statistics of distance/area/volume interstices.
radius_type (str): source of radius estimate. (default: “MiracleRadius”)</p>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.site.IntersticeDistribution.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">6.5</span></em>, <em class="sig-param"><span class="n">interstice_types</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">stats</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">radius_type</span><span class="o">=</span><span class="default_value">'MiracleRadius'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.IntersticeDistribution.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.IntersticeDistribution.analyze_area_interstice">
<em class="property">static </em><code class="sig-name descname">analyze_area_interstice</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">nn_coords</span></em>, <em class="sig-param"><span class="n">nn_rs</span></em>, <em class="sig-param"><span class="n">convex_hull_simplices</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.IntersticeDistribution.analyze_area_interstice" title="Permalink to this definition">¶</a></dt>
<dd><p>Analyze the area interstices in the neighbor convex hull facets.
Args:</p>
<blockquote>
<div><p>nn_coords (array-like, shape (N, 3)): Nearest Neighbors’ coordinates
nn_rs ([float]): Nearest Neighbors’ radii.
convex_hull_simplices (array-like, shape (M, 3)): Indices of points</p>
<blockquote>
<div><p>forming the simplicial facets of convex hull.</p>
</div></blockquote>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>area_interstice_list ([float]): Area interstice list.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.IntersticeDistribution.analyze_dist_interstices">
<em class="property">static </em><code class="sig-name descname">analyze_dist_interstices</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">center_r</span></em>, <em class="sig-param"><span class="n">nn_rs</span></em>, <em class="sig-param"><span class="n">nn_dists</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.IntersticeDistribution.analyze_dist_interstices" title="Permalink to this definition">¶</a></dt>
<dd><p>Analyze the distance interstices between center atom and neighbors.
Args:</p>
<blockquote>
<div><p>center_r (float): central atom’s radius.
nn_rs ([float]): Nearest Neighbors’ radii.
nn_dists ([float]): Nearest Neighbors’ distances.</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>dist_interstice_list ([float]): Distance interstice list.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.IntersticeDistribution.analyze_vol_interstice">
<em class="property">static </em><code class="sig-name descname">analyze_vol_interstice</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">center_coords</span></em>, <em class="sig-param"><span class="n">nn_coords</span></em>, <em class="sig-param"><span class="n">center_r</span></em>, <em class="sig-param"><span class="n">nn_rs</span></em>, <em class="sig-param"><span class="n">convex_hull_simplices</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.IntersticeDistribution.analyze_vol_interstice" title="Permalink to this definition">¶</a></dt>
<dd><p>Analyze the volume interstices in the tetrahedra formed by center
atom and neighbor convex hull triplets.
Args:</p>
<blockquote>
<div><p>center_coords ([float]): Central atomic coordinates.
nn_coords (array-like, shape (N, 3)): Nearest Neighbors’ coordinates
center_r (float): central atom’s radius.
nn_rs ([float]): Nearest Neighbors’ radii.
convex_hull_simplices (array-like, shape (M, 3)): Indices of points</p>
<blockquote>
<div><p>forming the simplicial facets of convex hull.</p>
</div></blockquote>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>volume_interstice_list ([float]): Volume interstice list.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.IntersticeDistribution.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.IntersticeDistribution.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.IntersticeDistribution.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.IntersticeDistribution.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.IntersticeDistribution.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">struct</span></em>, <em class="sig-param"><span class="n">idx</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.IntersticeDistribution.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get interstice distribution fingerprints of site with given index in
input structure.
Args:</p>
<blockquote>
<div><p>struct (Structure): Pymatgen Structure object.
idx (int): index of target site in structure.</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>interstice_fps ([float]): Interstice distribution fingerprints.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.IntersticeDistribution.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.IntersticeDistribution.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.site.LocalPropertyDifference">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.site.</code><code class="sig-name descname">LocalPropertyDifference</code><span class="sig-paren">(</span><em class="sig-param">data_source=&lt;matminer.utils.data.MagpieData object&gt;</em>, <em class="sig-param">weight='area'</em>, <em class="sig-param">properties=('Electronegativity'</em>, <em class="sig-param">)</em>, <em class="sig-param">signed=False</em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.LocalPropertyDifference" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Differences in elemental properties between site and its neighboring sites.</p>
<p>Uses the Voronoi tessellation of the structure to determine the
neighbors of the site, and assigns each neighbor (<img class="math" src="_images/math/5a939c5280da7202ca4531f175a7780ad5e1f80a.png" alt="n"/>) a
weight (<img class="math" src="_images/math/da96c58cf8c1bcc132a2252a23859c4ceaadf9f9.png" alt="A_n"/>) that corresponds to the area of the facet
on the tessellation corresponding to that neighbor.
The local property difference is then computed by
<img class="math" src="_images/math/44b79d2c935cf226fc2707938207c023c1351cb0.png" alt="\frac{\sum_n {A_n |p_n - p_0|}}{\sum_n {A_n}}"/>
where <img class="math" src="_images/math/3fd88e5561f4bce2bf17dc50624a4756051bf38a.png" alt="p_n"/> is the property (e.g., atomic number) of a neighbor
and <img class="math" src="_images/math/8968e4fd6ebe2d668010ed37e2172340daaefa2d.png" alt="p_0"/> is the property of a site. If signed parameter is assigned
True, signed difference of the properties is returned instead of absolute
difference.</p>
<dl class="simple">
<dt>Features:</dt><dd><ul class="simple">
<li><dl class="simple">
<dt>“local property difference in [property]” - Weighted average</dt><dd><p>of differences between an elemental property of a site and
that of each of its neighbors, weighted by size of face on
Voronoi tessellation</p>
</dd>
</dl>
</li>
</ul>
</dd>
<dt>References:</dt><dd><p><a class="reference external" href="http://link.aps.org/doi/10.1103/PhysRevB.96.024104">Ward et al. _PRB_ 2017</a></p>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.site.LocalPropertyDifference.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">data_source=&lt;matminer.utils.data.MagpieData object&gt;</em>, <em class="sig-param">weight='area'</em>, <em class="sig-param">properties=('Electronegativity'</em>, <em class="sig-param">)</em>, <em class="sig-param">signed=False</em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.LocalPropertyDifference.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the featurizer</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>data_source (AbstractData) - Class from which to retrieve</dt><dd><p>elemental properties</p>
</dd>
<dt>weight (str) - What aspect of each voronoi facet to use to</dt><dd><p>weigh each neighbor (see VoronoiNN)</p>
</dd>
</dl>
<p>properties ([str]) - List of properties to use (default=[‘Electronegativity’])
signed (bool) - whether to return absolute difference or signed difference of</p>
<blockquote>
<div><p>properties(default=False (absolute difference))</p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.LocalPropertyDifference.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.LocalPropertyDifference.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.LocalPropertyDifference.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.LocalPropertyDifference.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.LocalPropertyDifference.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">strc</span></em>, <em class="sig-param"><span class="n">idx</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.LocalPropertyDifference.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, which has to be implemented
in any derived featurizer subclass.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).</p>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.LocalPropertyDifference.from_preset">
<em class="property">static </em><code class="sig-name descname">from_preset</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">preset</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.LocalPropertyDifference.from_preset" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a new LocalPropertyDifference class according to a preset</p>
<dl class="simple">
<dt>Args:</dt><dd><p>preset (str) - Name of preset</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.LocalPropertyDifference.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.LocalPropertyDifference.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.site.OPSiteFingerprint">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.site.</code><code class="sig-name descname">OPSiteFingerprint</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target_motifs</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">dr</span><span class="o">=</span><span class="default_value">0.1</span></em>, <em class="sig-param"><span class="n">ddr</span><span class="o">=</span><span class="default_value">0.01</span></em>, <em class="sig-param"><span class="n">ndr</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">dop</span><span class="o">=</span><span class="default_value">0.001</span></em>, <em class="sig-param"><span class="n">dist_exp</span><span class="o">=</span><span class="default_value">2</span></em>, <em class="sig-param"><span class="n">zero_ops</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.OPSiteFingerprint" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Local structure order parameters computed from a site’s neighbor env.</p>
<p>For each order parameter, we determine
the neighbor shell that complies with the expected
coordination number. For example, we find the 4 nearest
neighbors for the tetrahedral OP, the 6 nearest for the
octahedral OP, and the 8 nearest neighbors for the bcc OP.
If we don’t find such a shell, the OP is either set to zero
or evaluated with the shell of the next largest observed
coordination number.
Args:</p>
<blockquote>
<div><dl class="simple">
<dt>target_motifs (dict): target op or motif type where keys</dt><dd><p>are corresponding coordination numbers
(e.g., {4: “tetrahedral”}).</p>
</dd>
<dt>dr (float): width for binning neighbors in unit of relative</dt><dd><p>distances (= distance/nearest neighbor
distance).  The binning is necessary to make the
neighbor-finding step robust against small numerical
variations in neighbor distances (default: 0.1).</p>
</dd>
</dl>
<p>ddr (float): variation of width for finding stable OP values.
ndr (int): number of width variations for each variation direction</p>
<blockquote>
<div><p>(e.g., ndr = 0 only uses the input dr, whereas
ndr=1 tests dr = dr - ddr, dr, and dr + ddr.</p>
</div></blockquote>
<dl class="simple">
<dt>dop (float): binning width to compute histogram for each OP</dt><dd><p>if ndr &gt; 0.</p>
</dd>
<dt>dist_exp (boolean): exponent for distance factor to multiply</dt><dd><p>order parameters with that penalizes (large)
variations in distances in a given motif.
0 will switch the option off
(default: 2).</p>
</dd>
<dt>zero_ops (boolean): set an OP to zero if there is no neighbor</dt><dd><p>shell that complies with the expected
coordination number of a given OP
(e.g., CN=4 for tetrahedron;
default: True).</p>
</dd>
</dl>
</div></blockquote>
<dl class="py method">
<dt id="matminer.featurizers.site.OPSiteFingerprint.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">target_motifs</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">dr</span><span class="o">=</span><span class="default_value">0.1</span></em>, <em class="sig-param"><span class="n">ddr</span><span class="o">=</span><span class="default_value">0.01</span></em>, <em class="sig-param"><span class="n">ndr</span><span class="o">=</span><span class="default_value">1</span></em>, <em class="sig-param"><span class="n">dop</span><span class="o">=</span><span class="default_value">0.001</span></em>, <em class="sig-param"><span class="n">dist_exp</span><span class="o">=</span><span class="default_value">2</span></em>, <em class="sig-param"><span class="n">zero_ops</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.OPSiteFingerprint.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.OPSiteFingerprint.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.OPSiteFingerprint.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.OPSiteFingerprint.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.OPSiteFingerprint.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.OPSiteFingerprint.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">struct</span></em>, <em class="sig-param"><span class="n">idx</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.OPSiteFingerprint.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get OP fingerprint of site with given index in input
structure.
Args:</p>
<blockquote>
<div><p>struct (Structure): Pymatgen Structure object.
idx (int): index of target site in structure.</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>opvals (numpy array): order parameters of target site.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.OPSiteFingerprint.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.OPSiteFingerprint.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.site.SOAP">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.site.</code><code class="sig-name descname">SOAP</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.SOAP" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Smooth overlap of atomic positions (interface via DScribe).</p>
<p>Class for generating a partial power spectrum from Smooth Overlap of Atomic
Orbitals (SOAP). This implementation uses real (tesseral) spherical
harmonics as the angular basis set and provides two orthonormalized
alternatives for the radial basis functions: spherical primitive gaussian
type orbitals (“gto”) or the polynomial basis set (“polynomial”). By
default the faster gto-basis is used. Please see the DScribe SOAP
documentation for more details.</p>
<p>Note that SOAP is only featurized for elements identified by “fit” (see
following), thus “fit” must be called before “featurize”, or else an error
will be raised.</p>
<p>Based originally on the following publications:</p>
<dl class="simple">
<dt>“On representing chemical environments, Albert P. Bartók, Risi</dt><dd><p>Kondor, and Gábor Csányi, Phys. Rev. B 87, 184115, (2013),
<a class="reference external" href="https://doi.org/10.1103/PhysRevB.87.184115">https://doi.org/10.1103/PhysRevB.87.184115</a></p>
</dd>
<dt>“Comparing molecules and solids across structural and alchemical</dt><dd><p>space”, Sandip De, Albert P. Bartók, Gábor Csányi and Michele Ceriotti,
Phys.  Chem. Chem. Phys. 18, 13754 (2016),
<a class="reference external" href="https://doi.org/10.1039/c6cp00415f">https://doi.org/10.1039/c6cp00415f</a></p>
</dd>
</dl>
<p>Implementation (and some documentation) originally based on DScribe:
<a class="reference external" href="https://github.com/SINGROUP/dscribe">https://github.com/SINGROUP/dscribe</a>.</p>
<dl>
<dt>“DScribe: Library of descriptors for machine learning in materials science”,</dt><dd><p>Himanen, L., J{“a}ger, M. O.J., Morooka, E. V., Federici
Canova, F., Ranawat, Y. S., Gao, D. Z., Rinke, P. and Foster, A. S.
Computer Physics Communications, 106949 (2019),
<a class="reference external" href="https://doi.org/10.1016/j.cpc.2019.106949">https://doi.org/10.1016/j.cpc.2019.106949</a></p>
</dd>
<dt>Args:</dt><dd><dl class="simple">
<dt>rcut (float): A cutoff for local region in angstroms. Should be</dt><dd><p>bigger than 1 angstrom.</p>
</dd>
</dl>
<p>nmax (int): The number of radial basis functions.
lmax (int): The maximum degree of spherical harmonics.
sigma (float): The standard deviation of the gaussians used to expand the</p>
<blockquote>
<div><p>atomic density.</p>
</div></blockquote>
<p>rbf (str): The radial basis functions to use. The available options are:</p>
<blockquote>
<div><ul class="simple">
<li><p>“gto”: Spherical gaussian type orbitals defined as <img class="math" src="_images/math/9c175e2552d4e2d9c682055a2ce7d1fbb538116d.png" alt="g_{nl}(r) = \sum_{n'=1}^{n_\mathrm{max}}\,\beta_{nn'l} r^l e^{-\alpha_{n'l}r^2}"/></p></li>
<li><p>“polynomial”: Polynomial basis defined as <img class="math" src="_images/math/b7fdf13f2e4d24f9f2d55a77dcebbf52c8ad19c7.png" alt="g_{n}(r) = \sum_{n'=1}^{n_\mathrm{max}}\,\beta_{nn'} (r-r_\mathrm{cut})^{n'+2}"/></p></li>
</ul>
</div></blockquote>
<dl class="simple">
<dt>periodic (bool): Determines whether the system is considered to be</dt><dd><p>periodic.</p>
</dd>
<dt>crossover (bool): Determines if crossover of atomic types should</dt><dd><p>be included in the power spectrum. If enabled, the power
spectrum is calculated over all unique species combinations Z
and Z’. If disabled, the power spectrum does not contain
cross-species information and is only run over each unique
species Z. Turned on by default to correspond to the original
definition</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.site.SOAP.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">rcut</span></em>, <em class="sig-param"><span class="n">nmax</span></em>, <em class="sig-param"><span class="n">lmax</span></em>, <em class="sig-param"><span class="n">sigma</span></em>, <em class="sig-param"><span class="n">periodic</span></em>, <em class="sig-param"><span class="n">rbf</span><span class="o">=</span><span class="default_value">'gto'</span></em>, <em class="sig-param"><span class="n">crossover</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.SOAP.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.SOAP.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.SOAP.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.SOAP.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.SOAP.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.SOAP.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">struct</span></em>, <em class="sig-param"><span class="n">idx</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.SOAP.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, which has to be implemented
in any derived featurizer subclass.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).</p>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.SOAP.fit">
<code class="sig-name descname">fit</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">X</span></em>, <em class="sig-param"><span class="n">y</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.SOAP.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit the SOAP featurizer to a dataframe.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>X ([SiteCollection]): For example, a list of pymatgen Structures.
y : unused (added for consistency with overridden method signature)</p>
</dd>
<dt>Returns:</dt><dd><p>self</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.SOAP.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.SOAP.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.site.SiteElementalProperty">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.site.</code><code class="sig-name descname">SiteElementalProperty</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data_source</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">properties</span><span class="o">=</span><span class="default_value">'Number'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.SiteElementalProperty" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Elemental properties of atom on a certain site</p>
<dl class="simple">
<dt>Features:</dt><dd><p>site [property] - Elemental property for this site</p>
</dd>
<dt>References:</dt><dd><p><a class="reference external" href="http://link.aps.org/doi/10.1103/PhysRevB.95.144110">Seko et al., _PRB_ (2017)</a>
<a class="reference external" href="http://dx.doi.org/10.1021/acs.chemmater.7b00156">Schmidt et al., _Chem Mater_. (2017)</a></p>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.site.SiteElementalProperty.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">data_source</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">properties</span><span class="o">=</span><span class="default_value">'Number'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.SiteElementalProperty.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the featurizer</p>
<dl class="simple">
<dt>Args:</dt><dd><p>data_source (AbstractData): Tool used to look up elemental properties
properties ([string]): List of properties to use for features</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.SiteElementalProperty.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.SiteElementalProperty.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.SiteElementalProperty.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.SiteElementalProperty.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.SiteElementalProperty.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">strc</span></em>, <em class="sig-param"><span class="n">idx</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.SiteElementalProperty.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, which has to be implemented
in any derived featurizer subclass.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).</p>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.SiteElementalProperty.from_preset">
<em class="property">static </em><code class="sig-name descname">from_preset</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">preset</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.SiteElementalProperty.from_preset" title="Permalink to this definition">¶</a></dt>
<dd><p>Create the class with pre-defined settings</p>
<dl class="simple">
<dt>Args:</dt><dd><p>preset (string): Desired preset</p>
</dd>
<dt>Returns:</dt><dd><p>SiteElementalProperty initialized with desired settings</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.SiteElementalProperty.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.SiteElementalProperty.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.site.VoronoiFingerprint">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.site.</code><code class="sig-name descname">VoronoiFingerprint</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">6.5</span></em>, <em class="sig-param"><span class="n">use_symm_weights</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">symm_weights</span><span class="o">=</span><span class="default_value">'solid_angle'</span></em>, <em class="sig-param"><span class="n">stats_vol</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">stats_area</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">stats_dist</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.VoronoiFingerprint" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Voronoi tessellation-based features around target site.</p>
<p>Calculate the following sets of features based on Voronoi tessellation
analysis around the target site:
Voronoi indices</p>
<blockquote>
<div><p>n_i denotes the number of i-edged facets, and i is in the range of 3-10.
e.g.
for bcc lattice, the Voronoi indices are [0,6,0,8,…];
for fcc/hcp lattice, the Voronoi indices are [0,12,0,0,…];
for icosahedra, the Voronoi indices are [0,0,12,0,…];</p>
</div></blockquote>
<dl>
<dt>i-fold symmetry indices</dt><dd><p>computed as n_i/sum(n_i), and i is in the range of 3-10.
reflect the strength of i-fold symmetry in local sites.
e.g.
for bcc lattice, the i-fold symmetry indices are [0,6/14,0,8/14,…]</p>
<blockquote>
<div><p>indicating both 4-fold and a stronger 6-fold symmetries are present;</p>
</div></blockquote>
<dl class="simple">
<dt>for fcc/hcp lattice, the i-fold symmetry factors are [0,1,0,0,…],</dt><dd><p>indicating only 4-fold symmetry is present;</p>
</dd>
<dt>for icosahedra, the Voronoi indices are [0,0,1,0,…],</dt><dd><p>indicating only 5-fold symmetry is present;</p>
</dd>
</dl>
</dd>
<dt>Weighted i-fold symmetry indices</dt><dd><p>if use_weights = True</p>
</dd>
<dt>Voronoi volume</dt><dd><p>total volume of the Voronoi polyhedron around the target site</p>
</dd>
<dt>Voronoi volume statistics of sub_polyhedra formed by each facet + center</dt><dd><p>stats_vol = [‘mean’, ‘std_dev’, ‘minimum’, ‘maximum’]</p>
</dd>
<dt>Voronoi area</dt><dd><p>total area of the Voronoi polyhedron around the target site</p>
</dd>
<dt>Voronoi area statistics of the facets</dt><dd><p>stats_area = [‘mean’, ‘std_dev’, ‘minimum’, ‘maximum’]</p>
</dd>
<dt>Voronoi nearest-neighboring distance statistics</dt><dd><p>stats_dist = [‘mean’, ‘std_dev’, ‘minimum’, ‘maximum’]</p>
</dd>
<dt>Args:</dt><dd><dl class="simple">
<dt>cutoff (float): cutoff distance in determining the potential</dt><dd><p>neighbors for Voronoi tessellation analysis.
(default: 6.5)</p>
</dd>
<dt>use_symm_weights(bool): whether to use weights to derive weighted</dt><dd><p>i-fold symmetry indices.</p>
</dd>
<dt>symm_weights(str): weights to be used in weighted i-fold symmetry</dt><dd><p>indices.
Supported options: ‘solid_angle’, ‘area’, ‘volume’,
‘face_dist’. (default: ‘solid_angle’)</p>
</dd>
</dl>
<p>stats_vol (list of str): volume statistics types.
stats_area (list of str): area statistics types.
stats_dist (list of str): neighboring distance statistics types.</p>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.site.VoronoiFingerprint.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">6.5</span></em>, <em class="sig-param"><span class="n">use_symm_weights</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">symm_weights</span><span class="o">=</span><span class="default_value">'solid_angle'</span></em>, <em class="sig-param"><span class="n">stats_vol</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">stats_area</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">stats_dist</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.VoronoiFingerprint.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.VoronoiFingerprint.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.VoronoiFingerprint.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.VoronoiFingerprint.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.VoronoiFingerprint.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.VoronoiFingerprint.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">struct</span></em>, <em class="sig-param"><span class="n">idx</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.VoronoiFingerprint.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get Voronoi fingerprints of site with given index in input structure.
Args:</p>
<blockquote>
<div><p>struct (Structure): Pymatgen Structure object.
idx (int): index of target site in structure.</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list of floats): Voronoi fingerprints.</dt><dd><p>-Voronoi indices
-i-fold symmetry indices
-weighted i-fold symmetry indices (if use_symm_weights = True)
-Voronoi volume
-Voronoi volume statistics
-Voronoi area
-Voronoi area statistics
-Voronoi dist statistics</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.site.VoronoiFingerprint.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.VoronoiFingerprint.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py function">
<dt id="matminer.featurizers.site.get_wigner_coeffs">
<code class="sig-prename descclassname">matminer.featurizers.site.</code><code class="sig-name descname">get_wigner_coeffs</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">l</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.site.get_wigner_coeffs" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the list of non-zero Wigner 3j triplets</p>
<dl class="simple">
<dt>Args:</dt><dd><p>l (int): Desired l</p>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>List of tuples that contain:</dt><dd><ul class="simple">
<li><p>((int)) m coordinates of the triplet</p></li>
<li><p>(float) Wigner coefficient</p></li>
</ul>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</div>
<div class="section" id="module-matminer.featurizers.structure">
<span id="matminer-featurizers-structure-module"></span><h2>matminer.featurizers.structure module<a class="headerlink" href="#module-matminer.featurizers.structure" title="Permalink to this headline">¶</a></h2>
<dl class="py class">
<dt id="matminer.featurizers.structure.BagofBonds">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">BagofBonds</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">coulomb_matrix</span><span class="o">=</span><span class="default_value">SineCoulombMatrix(flatten=False)</span></em>, <em class="sig-param"><span class="n">token</span><span class="o">=</span><span class="default_value">' - '</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.BagofBonds" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Compute a Bag of Bonds vector, as first described by Hansen et al. (2015).</p>
<p>The Bag of Bonds approach is based creating an even-length vector from a
Coulomb matrix output. Practically, it represents the Coloumbic interactions
between each possible set of sites in a structure as a vector.</p>
<p>BagofBonds must be fit to an iterable of structures using the “fit” method
before featurization can occur. This is because the bags and the maximum
lengths of each bag must be set prior to featurization. We recommend
fitting and featurizing on the same data to maintain consistency
between generated feature sets. This can be done using the fit_transform
method (for lists of structures) or the fit_featurize_dataframe method
(for dataframes).</p>
<p>BagofBonds is based on a method by Hansen et. al “Machine Learning
Predictions of Molecular Properties: Accurate Many-Body Potentials and
Nonlocality in Chemical Space” (2015).</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>coulomb_matrix (BaseFeaturizer): A featurizer object containing a</dt><dd><p>“featurize” method which returns a matrix of size nsites x nsites.
Good choices are CoulombMatrix() or SineCoulombMatrix(), with the
flatten=False parameter set.</p>
</dd>
<dt>token (str): The string used to separate species in a bond, including</dt><dd><p>spaces. The token must contain at least one space and cannot have
alphabetic characters in it, and should be padded by spaces. For
example, for the bond Cs+ - Cl-, the token is ‘ - ‘. This determines
how bonds are represented in the dataframe.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.structure.BagofBonds.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">coulomb_matrix</span><span class="o">=</span><span class="default_value">SineCoulombMatrix(flatten=False)</span></em>, <em class="sig-param"><span class="n">token</span><span class="o">=</span><span class="default_value">' - '</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.BagofBonds.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.BagofBonds.bag">
<code class="sig-name descname">bag</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">s</span></em>, <em class="sig-param"><span class="n">return_baglens</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.BagofBonds.bag" title="Permalink to this definition">¶</a></dt>
<dd><p>Convert a structure into a bag of bonds, where each bag has no padded
zeros. using this function will give the ‘raw’ bags, which when
concatenated, will have different lengths.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>s (Structure): A pymatgen Structure or IStructure object. May also</dt><dd><p>work with a</p>
</dd>
<dt>return_baglens (bool): If True, returns the bag of bonds with as</dt><dd><p>a dictionary with the number of bonds as values in place
of the vectors of coulomb matrix vals. If False, calculates
Coulomb matrix values and returns ‘raw’ bags.</p>
</dd>
</dl>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>(dict) A bag of bonds, where the keys are sorted tuples of pymatgen</dt><dd><p>Site objects representing bonds or sites, and the values are the
Coulomb matrix values for that bag.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.BagofBonds.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.BagofBonds.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.BagofBonds.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.BagofBonds.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.BagofBonds.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">s</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.BagofBonds.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Featurizes a structure according to the bag of bonds method.
Specifically, each structure is first bagged by flattening the
Coulomb matrix for the structure. Then, it is zero-padded according to
the maximum number of bonds in each bag, for the set of bags that
BagofBonds was fit with.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>s (Structure): A pymatgen structure object</p>
</dd>
<dt>Returns:</dt><dd><p>(list): The Bag of Bonds vector for the input structure</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.BagofBonds.fit">
<code class="sig-name descname">fit</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">X</span></em>, <em class="sig-param"><span class="n">y</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.BagofBonds.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Define the bags using a list of structures.</p>
<p>Both the names of the bags (e.g., Cs-Cl) and the maximum lengths of
the bags are set with fit.</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>X (Series/list): An iterable of pymatgen Structure</dt><dd><p>objects which will be used to determine the allowed bond
types and bag lengths.</p>
</dd>
</dl>
<p>y : unused (added for consistency with overridden method signature)</p>
</dd>
<dt>Returns:</dt><dd><p>self</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.BagofBonds.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.BagofBonds.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.structure.BondFractions">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">BondFractions</code><span class="sig-paren">(</span><em class="sig-param">nn=&lt;pymatgen.analysis.local_env.CrystalNN object&gt;</em>, <em class="sig-param">bbv=0</em>, <em class="sig-param">no_oxi=False</em>, <em class="sig-param">approx_bonds=False</em>, <em class="sig-param">token=' - '</em>, <em class="sig-param">allowed_bonds=None</em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.BondFractions" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Compute the fraction of each bond in a structure, based on NearestNeighbors.</p>
<p>For example, in a structure with 2 Li-O bonds and 3 Li-P bonds:</p>
<p>Li-0: 0.4
Li-P: 0.6</p>
<p>Features:</p>
<p>BondFractions must be fit with iterable of structures before featurization in
order to define the allowed bond types (features). To do this, pass a list
of allowed_bonds. Otherwise, fit based on a list of structures. If
allowed_bonds is defined and BondFractions is also fit, the intersection
of the two lists of possible bonds is used.</p>
<p>For dataframes containing structures of various compositions, a unified
dataframe is returned which has the collection of all possible bond types
gathered from all structures as columns. To approximate bonds based on
chemical rules (ie, for a structure which you’d like to featurize but has
bonds not in the allowed set), use approx_bonds = True.</p>
<p>BondFractions is based on the “sum over bonds” in the Bag of Bonds approach,
based on a method by Hansen et. al “Machine Learning Predictions of Molecular
Properties: Accurate Many-Body Potentials and Nonlocality in Chemical Space”
(2015).</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>nn (NearestNeighbors): A Pymatgen nearest neighbors derived object. For</dt><dd><p>example, pymatgen.analysis.local_env.VoronoiNN().</p>
</dd>
<dt>bbv (float): The ‘bad bond values’, values substituted for</dt><dd><p>structure-bond combinations which can not physically exist, but
exist in the unified dataframe. For example, if a dataframe contains
structures of BaLiP and BaTiO3, determines the value to place in
the Li-P column for the BaTiO3 row; by default, is 0.</p>
</dd>
<dt>no_oxi (bool): If True, the featurizer will be agnostic to oxidation</dt><dd><p>states, which prevents oxidation states from  differentiating
bonds. For example, if True, Ca - O is identical to Ca2+ - O2-,
Ca3+ - O-, etc., and all of them will be included in Ca - O column.</p>
</dd>
<dt>approx_bonds (bool): If True, approximates the fractions of bonds not</dt><dd><p>in allowed_bonds (forbidden bonds) with similar allowed bonds.
Chemical rules are used to determine which bonds are most ‘similar’;
particularly, the Euclidean distance between the 2-tuples of the
bonds in Mendeleev no. space is minimized for the approximate
bond chosen.</p>
</dd>
<dt>token (str): The string used to separate species in a bond, including</dt><dd><p>spaces. The token must contain at least one space and cannot have
alphabetic characters in it, and should be padded by spaces. For
example, for the bond Cs+ - Cl-, the token is ‘ - ‘. This determines
how bonds are represented in the dataframe.</p>
</dd>
<dt>allowed_bonds ([str]): A listlike object containing bond types as</dt><dd><p>strings. For example, Cs - Cl, or Li+ - O2-. Ions and elements
will still have distinct bonds if (1) the bonds list originally
contained them and (2) no_oxi is False. These must match the
token specified.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.structure.BondFractions.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">nn=&lt;pymatgen.analysis.local_env.CrystalNN object&gt;</em>, <em class="sig-param">bbv=0</em>, <em class="sig-param">no_oxi=False</em>, <em class="sig-param">approx_bonds=False</em>, <em class="sig-param">token=' - '</em>, <em class="sig-param">allowed_bonds=None</em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.BondFractions.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.BondFractions.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.BondFractions.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.BondFractions.enumerate_all_bonds">
<code class="sig-name descname">enumerate_all_bonds</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">structures</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.BondFractions.enumerate_all_bonds" title="Permalink to this definition">¶</a></dt>
<dd><p>Identify all the unique, possible bonds types of all structures present,
and create the ‘unified’ bonds list.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>structures (list/ndarray): List of pymatgen Structures</p>
</dd>
<dt>Returns:</dt><dd><p>A tuple of unique, possible bond types for an entire list of
structures. This tuple is used to form the unified feature labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.BondFractions.enumerate_bonds">
<code class="sig-name descname">enumerate_bonds</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">s</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.BondFractions.enumerate_bonds" title="Permalink to this definition">¶</a></dt>
<dd><p>Lists out all the bond possibilities in a single structure.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>s (Structure): A pymatgen structure</p>
</dd>
<dt>Returns:</dt><dd><p>A list of bond types in ‘Li-O’ form, where the order of the
elements in each bond type is alphabetic.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.BondFractions.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.BondFractions.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Returns the list of allowed bonds. Throws an error if the featurizer
has not been fit.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.BondFractions.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">s</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.BondFractions.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Quantify the fractions of each bond type in a structure.</p>
<p>For collections of structures, bonds types which are not found in a
particular structure (e.g., Li-P in BaTiO3) are represented as NaN.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>s (Structure): A pymatgen Structure object</p>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) The feature list of bond fractions, in the order of the</dt><dd><p>alphabetized corresponding bond names.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.BondFractions.fit">
<code class="sig-name descname">fit</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">X</span></em>, <em class="sig-param"><span class="n">y</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.BondFractions.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Define the bond types allowed to be returned during each featurization.
Bonds found during featurization which are not allowed will be omitted
from the returned dataframe or matrix.</p>
<p>Fit BondFractions by either passing an iterable of structures to
training_data or by defining the bonds explicitly with allowed_bonds
in __init__.</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>X (Series/list): An iterable of pymatgen Structure</dt><dd><p>objects which will be used to determine the allowed bond
types.</p>
</dd>
</dl>
<p>y : unused (added for consistency with overridden method signature)</p>
</dd>
<dt>Returns:</dt><dd><p>self</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.BondFractions.from_preset">
<em class="property">static </em><code class="sig-name descname">from_preset</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">preset</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.BondFractions.from_preset" title="Permalink to this definition">¶</a></dt>
<dd><p>Use one of the standard instances of a given NearNeighbor class.
Pass args to __init__, such as allowed_bonds, using this method as well.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>preset (str): preset type (“CrystalNN”, “VoronoiNN”, “JmolNN”,
“MiniumDistanceNN”, “MinimumOKeeffeNN”, or “MinimumVIRENN”).</p>
</dd>
<dt>Returns:</dt><dd><p>CoordinationNumber from a preset.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.BondFractions.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.BondFractions.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.structure.CGCNNFeaturizer">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">CGCNNFeaturizer</code><span class="sig-paren">(</span><em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.CGCNNFeaturizer" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Features generated by training a Crystal Graph Convolutional Neural Network
(CGCNN) model.</p>
<dl class="simple">
<dt>This featurizer requires a CGCNN model that can either be:</dt><dd><ol class="arabic simple">
<li><p>from a pretrained model, currently only supports the models from
the CGCNN repo (12/10/18): <a class="reference external" href="https://github.com/txie-93/cgcnn">https://github.com/txie-93/cgcnn</a>;</p></li>
<li><p>train a CGCNN model based on the X (structures) and y (target) from
fresh start;</p></li>
<li><p>similar to 2), but train a model from a warm_start model that can
either be a pretrained model or saved checkpoints.</p></li>
</ol>
</dd>
</dl>
<p>Please see the fit function for more details.</p>
<p>After obtaining a CGCNN model, we will featurize the structures by taking
the crystal feature vector obtained after pooling as the features.</p>
<p>This featurizer requires installing cgcnn and torch. We wrap and refractor
some of the classes and functions from the original cgcnn to make them
work better for matminer. Please also see utils/cgcnn for more details.</p>
<dl class="simple">
<dt>Features:</dt><dd><ul class="simple">
<li><p>Features for the structures extracted from CGCNN model after pooling.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.structure.CGCNNFeaturizer.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">task</span><span class="o">=</span><span class="default_value">'classification'</span></em>, <em class="sig-param"><span class="n">atom_init_fea</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">pretrained_name</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">warm_start_file</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">warm_start_latest</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">save_model_to_dir</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">save_checkpoint_to_dir</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">checkpoint_interval</span><span class="o">=</span><span class="default_value">100</span></em>, <em class="sig-param"><span class="n">del_checkpoint</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">cgcnn_kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.CGCNNFeaturizer.__init__" title="Permalink to this definition">¶</a></dt>
<dd><dl>
<dt>Args:</dt><dd><dl>
<dt>task (str):</dt><dd><p>Task type, “classification” or “regression”.</p>
</dd>
<dt>atom_init_fea (dict):</dt><dd><p>A dict of {atom type: atom feature}. If not provided, will use
the default atom features from the CGCNN repo.</p>
</dd>
<dt>pretrained_name (str):</dt><dd><p>CGCNN pretrained model name, if None don’t use pre-trained model</p>
</dd>
<dt>warm_start_file (str):</dt><dd><p>The warm start model file, if None, don’t warm start.</p>
</dd>
<dt>warm_start_latest(bool):</dt><dd><p>Warm start from the latest model or best model.
This is set because we customize our checkpoints to contain both
best model and latest model. And if the warm start model does
not contain these two options, will just use the static_dict
given in the model/checkpoints to warm start.</p>
</dd>
<dt>save_model_to_dir (str):</dt><dd><p>Whether to save the best model to disk, if None, don’t save,
otherwise, save the best model to ‘save_model_to_dir’ path.</p>
</dd>
<dt>save_checkpoint_to_dir (str):</dt><dd><p>Whether to save checkpoint during training, if None, don’t save,
otherwise, save the it to ‘save_checkpoint_to_dir’ path.</p>
</dd>
<dt>checkpoint_interval (int):</dt><dd><p>Save checkpoint every n epochs if save_checkpoint_to_dir is not
None. If the epochs is less than this checkpoint_interval, will
reset the checkpoint_interval as int(epochs/2).</p>
</dd>
<dt>del_checkpoint (bool):</dt><dd><p>Whether to delete checkpoints if training ends successfully.</p>
</dd>
<dt><a href="#id22"><span class="problematic" id="id23">**</span></a>cgcnn_kwargs (optional): settings of CGCNN, containing:</dt><dd><dl>
<dt>CrystalGraphConvNet model kwargs:</dt><dd><dl class="simple">
<dt>-atom_fea_len (int): Number of hidden atom features in conv</dt><dd><p>layers, default 64.</p>
</dd>
</dl>
<p>-n_conv (int): Number of conv layers, default 3.
-h_fea_len (int): Number of hidden features after pooling,</p>
<blockquote>
<div><p>default 128.</p>
</div></blockquote>
<p>-n_epochs (int): Number of total epochs to run, default 30.
-print_freq (bool): Print frequency, default 10.
-test (bool): Whether to save test predictions
-task (str): “classification” or “regression”,</p>
<blockquote>
<div><p>default “classification”.</p>
</div></blockquote>
</dd>
<dt>Dataset (CIFDataWrapper) kwargs:</dt><dd><dl class="simple">
<dt>-max_num_nbr (int): The maximum number of neighbors while</dt><dd><p>constructing the crystal graph, default 12</p>
</dd>
<dt>-radius (float): The cutoff radius for searching neighbors,</dt><dd><p>default 8</p>
</dd>
<dt>-dmin (float): The minimum distance for constructing</dt><dd><p>GaussianDistance, default 0</p>
</dd>
<dt>-step (float): The step size for constructing</dt><dd><p>GaussianDistance, default 0.2</p>
</dd>
<dt>-random_seed (int): Random seed for shuffling the dataset,</dt><dd><p>default 123</p>
</dd>
</dl>
</dd>
<dt>DataLoader kwargs:</dt><dd><p>batch_size (int): Mini-batch size, default 256
num_workers (int): Number of data loading workers, default 0
train_size (int): Number of training data to be loaded,</p>
<blockquote>
<div><p>default none</p>
</div></blockquote>
<dl class="simple">
<dt>val_size (int): Number of validation data to be loaded,</dt><dd><p>default 1000</p>
</dd>
<dt>test_size (int): Number of test data to be loaded,</dt><dd><p>default 1000</p>
</dd>
<dt>“return_test” (bool): Whether to return the test dataset</dt><dd><p>loader. default True</p>
</dd>
</dl>
</dd>
<dt>Optimizer kwargs:</dt><dd><dl class="simple">
<dt>-optim (str): Choose an optimizer, “SGD” or “Adam”,</dt><dd><p>default “SGD”.</p>
</dd>
</dl>
<p>-lr (float): Initial learning rate, default 0.01
-momentum (float): Momentum, default 0.9
-weight_decay (float): Weight decay (default: 0)</p>
</dd>
<dt>Scheduler MultiStepLR kwargs:</dt><dd><dl class="simple">
<dt>-gamma (float): Multiplicative factor of learning rate</dt><dd><p>decay, default: 0.1.</p>
</dd>
<dt>-lr_milestones (list): List of epoch indices.</dt><dd><p>Must be increasing.</p>
</dd>
</dl>
</dd>
</dl>
<p>These input cgcnn_kwargs will be processed and grouped in
_initialize_kwargs.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.CGCNNFeaturizer.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.CGCNNFeaturizer.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.CGCNNFeaturizer.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.CGCNNFeaturizer.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.CGCNNFeaturizer.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">strc</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.CGCNNFeaturizer.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the feature vector after pooling layer of the CGCNN model obtained
from fit.
Args:</p>
<blockquote>
<div><p>strc (Structure): Structure object</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>Features extracted after the pooling layer in CGCNN model</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.CGCNNFeaturizer.fit">
<code class="sig-name descname">fit</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">X</span></em>, <em class="sig-param"><span class="n">y</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.CGCNNFeaturizer.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Get a CGCNN model that can either be:
1) from a pretrained model, currently only supports the models from</p>
<blockquote>
<div><p>the CGCNN repo;</p>
</div></blockquote>
<ol class="arabic simple" start="2">
<li><p>train a CGCNN model based on the X (structures) and y (target) from
fresh start;</p></li>
<li><p>similar to 2), but train a model from a warm_start model that can
either be a pretrained model or saved checkpoints.</p></li>
</ol>
<p>Note that to use CGCNNFeaturizer, a target y is needed!
Args:</p>
<blockquote>
<div><dl class="simple">
<dt>X (Series/list):</dt><dd><p>An iterable of pymatgen Structure objects.</p>
</dd>
<dt>y (Series/list):</dt><dd><p>Target property that CGCNN is designed to predict.</p>
</dd>
</dl>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>self</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.CGCNNFeaturizer.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.CGCNNFeaturizer.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.CGCNNFeaturizer.latest_model">
<em class="property">property </em><code class="sig-name descname">latest_model</code><a class="headerlink" href="#matminer.featurizers.structure.CGCNNFeaturizer.latest_model" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the latest model</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.CGCNNFeaturizer.model">
<em class="property">property </em><code class="sig-name descname">model</code><a class="headerlink" href="#matminer.featurizers.structure.CGCNNFeaturizer.model" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the best model</p>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.structure.ChemicalOrdering">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">ChemicalOrdering</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">shells</span><span class="o">=</span><span class="default_value">1, 2, 3</span></em>, <em class="sig-param"><span class="n">weight</span><span class="o">=</span><span class="default_value">'area'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.ChemicalOrdering" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>How much the ordering of species in the structure differs from random</p>
<p>These parameters describe how much the ordering of all species in a
structure deviates from random using a Warren-Cowley-like ordering
parameter. The first step of this calculation is to determine the nearest
neighbor shells of each site. Then, for each shell a degree of order for
each type is determined by computing:</p>
<p><img class="math" src="_images/math/78eabc361f3a54daeed501fff0d3353b946504ae.png" alt="\alpha (t,s) = 1 - \frac{\sum_n w_n \delta (t - t_n)}{x_t \sum_n w_n}"/></p>
<p>where <img class="math" src="_images/math/045fae7cfacaf28d29d6bdb81ab5ef61b1b07808.png" alt="w_n"/> is the weight associated with a certain neighbor,
<img class="math" src="_images/math/f7d483a2293790337ad1ba440b6430f82ad420d6.png" alt="t_p"/> is the type of the neighbor, and <img class="math" src="_images/math/8f3c6b529aaa515fdbc6c51b4fc4f73293dbfc09.png" alt="x_t"/> is the fraction
of type t in the structure. For atoms that are randomly dispersed in a
structure, this formula yields 0 for all types. For structures where
each site is surrounded only by atoms of another type, this formula
yields large values of <img class="math" src="_images/math/f0d949c29370cd898ed70f8f32a09349764a1ebb.png" alt="alpha"/>.</p>
<p>The mean absolute value of this parameter across all sites is used
as a feature.</p>
<dl class="simple">
<dt>Features:</dt><dd><dl class="simple">
<dt>mean ordering parameter shell [n] - Mean ordering parameter for</dt><dd><p>atoms in the n&lt;sup&gt;th&lt;/sup&gt; neighbor shell</p>
</dd>
</dl>
</dd>
<dt>References:</dt><dd><p><a class="reference external" href="http://link.aps.org/doi/10.1103/PhysRevB.96.024104">Ward et al. _PRB_ 2017</a></p>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.structure.ChemicalOrdering.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">shells</span><span class="o">=</span><span class="default_value">1, 2, 3</span></em>, <em class="sig-param"><span class="n">weight</span><span class="o">=</span><span class="default_value">'area'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.ChemicalOrdering.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the featurizer</p>
<dl class="simple">
<dt>Args:</dt><dd><p>shells ([int]) - Which neighbor shells to evaluate
weight (str) - Attribute used to weigh neighbor contributions</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.ChemicalOrdering.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.ChemicalOrdering.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.ChemicalOrdering.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.ChemicalOrdering.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.ChemicalOrdering.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">strc</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.ChemicalOrdering.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, which has to be implemented
in any derived featurizer subclass.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).</p>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.ChemicalOrdering.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.ChemicalOrdering.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.structure.CoulombMatrix">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">CoulombMatrix</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">diag_elems</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">flatten</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.CoulombMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>The Coulomb matrix, a representation of nuclear coulombic interaction.</p>
<p>Generate the Coulomb matrix, M, of the input structure (or molecule). The
Coulomb matrix was put forward by Rupp et al. (Phys. Rev. Lett. 108, 058301,
2012) and is defined by off-diagonal elements M_ij = Z_i*Z_j/<a href="#id34"><span class="problematic" id="id35">|R_i-R_j|</span></a> and
diagonal elements 0.5*Z_i^2.4, where Z_i and R_i denote the nuclear charge
and the position of atom i, respectively.</p>
<p>Coulomb Matrix features are flattened (for ML-readiness) by default. Use
fit before featurizing to use flattened features. To return the matrix form,
set flatten=False.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>diag_elems (bool): flag indication whether (True, default) to use</dt><dd><p>the original definition of the diagonal elements; if set to False,
the diagonal elements are set to 0</p>
</dd>
<dt>flatten (bool): If True, returns a flattened vector based on eigenvalues</dt><dd><p>of the matrix form. Otherwise, returns a matrix object (single
feature), which will likely need to be processed further.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.structure.CoulombMatrix.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">diag_elems</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">flatten</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.CoulombMatrix.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.CoulombMatrix.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.CoulombMatrix.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.CoulombMatrix.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.CoulombMatrix.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.CoulombMatrix.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">s</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.CoulombMatrix.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get Coulomb matrix of input structure.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>s: input Structure (or Molecule) object.</p>
</dd>
<dt>Returns:</dt><dd><p>m: (Nsites x Nsites matrix) Coulomb matrix.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.CoulombMatrix.fit">
<code class="sig-name descname">fit</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">X</span></em>, <em class="sig-param"><span class="n">y</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.CoulombMatrix.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit the Coulomb Matrix to a list of structures.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>X ([Structure]): A list of pymatgen structures.
y : unused (added for consistency with overridden method signature)</p>
</dd>
<dt>Returns:</dt><dd><p>self</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.CoulombMatrix.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.CoulombMatrix.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.structure.DensityFeatures">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">DensityFeatures</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">desired_features</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.DensityFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Calculates density and density-like features</p>
<dl class="simple">
<dt>Features:</dt><dd><ul class="simple">
<li><p>density</p></li>
<li><p>volume per atom</p></li>
<li><p>(“vpa”), and packing fraction</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.structure.DensityFeatures.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">desired_features</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.DensityFeatures.__init__" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>desired_features: [str] - choose from “density”, “vpa”,</dt><dd><p>“packing fraction”</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.DensityFeatures.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.DensityFeatures.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.DensityFeatures.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.DensityFeatures.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.DensityFeatures.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">s</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.DensityFeatures.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, which has to be implemented
in any derived featurizer subclass.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).</p>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.DensityFeatures.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.DensityFeatures.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.DensityFeatures.precheck">
<code class="sig-name descname">precheck</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">s</span><span class="p">:</span> <span class="n">pymatgen.core.structure.Structure</span></em><span class="sig-paren">)</span> &#x2192; <a class="reference external" href="https://docs.python.org/3/library/functions.html#bool" title="(in Python v3.9)">bool</a><a class="headerlink" href="#matminer.featurizers.structure.DensityFeatures.precheck" title="Permalink to this definition">¶</a></dt>
<dd><p>Precheck a single entry. DensityFeatures does not work for disordered
structures. To precheck an entire dataframe (qnd automatically gather
the fraction of structures that will pass the precheck), please use
precheck_dataframe.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>s (pymatgen.Structure): The structure to precheck.</p>
</dd>
<dt>Returns:</dt><dd><p>(bool): If True, s passed the precheck; otherwise, it failed.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.structure.Dimensionality">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">Dimensionality</code><span class="sig-paren">(</span><em class="sig-param">nn_method=&lt;pymatgen.analysis.local_env.CrystalNN object&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.Dimensionality" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Returns dimensionality of structure: 1 means linear chains of atoms OR
isolated atoms/no bonds, 2 means layered, 3 means 3D connected
structure. This feature is sensitive to bond length tables that you use.</p>
<dl class="py method">
<dt id="matminer.featurizers.structure.Dimensionality.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param">nn_method=&lt;pymatgen.analysis.local_env.CrystalNN object&gt;</em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.Dimensionality.__init__" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt><a href="#id25"><span class="problematic" id="id26">**</span></a>nn_method: The nearest neighbor method used to determine atomic</dt><dd><p>connectivity.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.Dimensionality.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.Dimensionality.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.Dimensionality.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.Dimensionality.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.Dimensionality.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">s</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.Dimensionality.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, which has to be implemented
in any derived featurizer subclass.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).</p>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.Dimensionality.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.Dimensionality.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.structure.ElectronicRadialDistributionFunction">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">ElectronicRadialDistributionFunction</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">dr</span><span class="o">=</span><span class="default_value">0.05</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.ElectronicRadialDistributionFunction" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Calculate the inherent electronic radial distribution function (ReDF)</p>
<p>The ReDF is defined according to Willighagen et al., Acta Cryst., 2005, B61,
29-36.</p>
<p>The ReDF is a structure-integral RDF (i.e., summed over
all sites) in which the positions of neighboring sites
are weighted by electrostatic interactions inferred
from atomic partial charges. Atomic charges are obtained
from the ValenceIonicRadiusEvaluator class.</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>cutoff: (float) distance up to which the ReDF is to be</dt><dd><p>calculated (default: longest diagaonal in
primitive cell).</p>
</dd>
</dl>
<p>dr: (float) width of bins (“x”-axis) of ReDF (default: 0.05 A).</p>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.structure.ElectronicRadialDistributionFunction.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">dr</span><span class="o">=</span><span class="default_value">0.05</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.ElectronicRadialDistributionFunction.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.ElectronicRadialDistributionFunction.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.ElectronicRadialDistributionFunction.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.ElectronicRadialDistributionFunction.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.ElectronicRadialDistributionFunction.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.ElectronicRadialDistributionFunction.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">s</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.ElectronicRadialDistributionFunction.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get ReDF of input structure.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>s: input Structure object.</p>
</dd>
<dt>Returns: (dict) a copy of the electronic radial distribution</dt><dd><p>functions (ReDF) as a dictionary. The distance list
(“x”-axis values of ReDF) can be accessed via key
‘distances’; the ReDF itself is accessible via key
‘redf’.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.ElectronicRadialDistributionFunction.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.ElectronicRadialDistributionFunction.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.structure.EwaldEnergy">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">EwaldEnergy</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">accuracy</span><span class="o">=</span><span class="default_value">4</span></em>, <em class="sig-param"><span class="n">per_atom</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.EwaldEnergy" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Compute the energy from Coulombic interactions.</p>
<p>Note: The energy is computed using _charges already defined for the <a href="#id42"><span class="problematic" id="id43">structure_</span></a>.</p>
<dl class="simple">
<dt>Features:</dt><dd><p>ewald_energy - Coulomb interaction energy of the structure</p>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.structure.EwaldEnergy.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">accuracy</span><span class="o">=</span><span class="default_value">4</span></em>, <em class="sig-param"><span class="n">per_atom</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.EwaldEnergy.__init__" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Args:</dt><dd><p>accuracy (int): Accuracy of Ewald summation, number of decimal places</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.EwaldEnergy.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.EwaldEnergy.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.EwaldEnergy.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.EwaldEnergy.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.EwaldEnergy.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">strc</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.EwaldEnergy.featurize" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Args:</dt><dd><p>(Structure) - Structure being analyzed</p>
</dd>
<dt>Returns:</dt><dd><p>([float]) - Electrostatic energy of the structure</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.EwaldEnergy.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.EwaldEnergy.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.structure.GlobalInstabilityIndex">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">GlobalInstabilityIndex</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">r_cut</span><span class="o">=</span><span class="default_value">4.0</span></em>, <em class="sig-param"><span class="n">disordered_pymatgen</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.GlobalInstabilityIndex" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>The global instability index of a structure.</p>
<p>The default is to use IUCr 2016 bond valence parameters for computing
bond valence sums. If the structure has disordered site occupancies
or non-integer valences on sites, pymatgen’s bond valence sum method
can be used instead.</p>
<p>Note that pymatgen’s bond valence sum method is prone to error unless
the correct scale factor is supplied. A scale factor based on testing
with perovskites is used here.
TODO: Use scipy to optimize scale factor for minimizing GII</p>
<p>Based on the following publication:</p>
<dl>
<dt>‘Structural characterization of R2BaCuO5 (R = Y, Lu, Yb, Tm, Er, Ho,</dt><dd><p>Dy, Gd, Eu and Sm) oxides by X-ray and neutron diffraction’,
A.Salinas-Sanchez, J.L.Garcia-Muñoz, J.Rodriguez-Carvajal,
R.Saez-Puche, and J.L.Martinez, Journal of Solid State Chemistry,
100, 201-211 (1992),
<a class="reference external" href="https://doi.org/10.1016/0022-4596(92)90094-C">https://doi.org/10.1016/0022-4596(92)90094-C</a></p>
</dd>
<dt>Args:</dt><dd><p>r_cut: Float, how far to search for neighbors when computing bond valences
disordered_pymatgen: Boolean, whether to fall back on pymatgen’s bond</p>
<blockquote>
<div><p>valence sum method for disordered structures</p>
</div></blockquote>
</dd>
<dt>Features:</dt><dd><dl class="simple">
<dt>The global instability index is the square root of the sum of squared</dt><dd><p>differences of the bond valence sums from the formal valences
averaged over all atoms in the unit cell.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.structure.GlobalInstabilityIndex.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">r_cut</span><span class="o">=</span><span class="default_value">4.0</span></em>, <em class="sig-param"><span class="n">disordered_pymatgen</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.GlobalInstabilityIndex.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.GlobalInstabilityIndex.calc_bv_sum">
<code class="sig-name descname">calc_bv_sum</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">site_val</span></em>, <em class="sig-param"><span class="n">site_el</span></em>, <em class="sig-param"><span class="n">neighbor_list</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.GlobalInstabilityIndex.calc_bv_sum" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes bond valence sum for site.
Args:</p>
<blockquote>
<div><p>site_val (Integer): valence of site
site_el (String): element name
neighbor_list (List): List of neighboring sites and their distances</p>
</div></blockquote>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.GlobalInstabilityIndex.calc_gii_iucr">
<code class="sig-name descname">calc_gii_iucr</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">s</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.GlobalInstabilityIndex.calc_gii_iucr" title="Permalink to this definition">¶</a></dt>
<dd><p>Computes global instability index using tabulated bv params.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>s: Pymatgen Structure object</p>
</dd>
<dt>Returns:</dt><dd><p>gii: Float, the global instability index</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.GlobalInstabilityIndex.calc_gii_pymatgen">
<code class="sig-name descname">calc_gii_pymatgen</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">struct</span></em>, <em class="sig-param"><span class="n">scale_factor</span><span class="o">=</span><span class="default_value">0.965</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.GlobalInstabilityIndex.calc_gii_pymatgen" title="Permalink to this definition">¶</a></dt>
<dd><p>Calculates global instability index using Pymatgen’s bond valence sum
Args:</p>
<blockquote>
<div><p>struct: Pymatgen Structure object
scale: Float, tunable scale factor for bond valence</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>gii: Float, global instability index</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.GlobalInstabilityIndex.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.GlobalInstabilityIndex.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.GlobalInstabilityIndex.compute_bv">
<em class="property">static </em><code class="sig-name descname">compute_bv</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">params</span></em>, <em class="sig-param"><span class="n">dist</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.GlobalInstabilityIndex.compute_bv" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute bond valence from parameters.
Args:</p>
<blockquote>
<div><p>params: Dataframe with Ro and B parameters
dist: Float, distance to neighboring atom</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>bv: Float, bond valence</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.GlobalInstabilityIndex.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.GlobalInstabilityIndex.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.GlobalInstabilityIndex.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">struct</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.GlobalInstabilityIndex.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get global instability index.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>struct: Pymatgen Structure object</p>
</dd>
<dt>Returns:</dt><dd><p>[gii]: Length 1 list with float value</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.GlobalInstabilityIndex.get_bv_params">
<code class="sig-name descname">get_bv_params</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cation</span></em>, <em class="sig-param"><span class="n">anion</span></em>, <em class="sig-param"><span class="n">cat_val</span></em>, <em class="sig-param"><span class="n">an_val</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.GlobalInstabilityIndex.get_bv_params" title="Permalink to this definition">¶</a></dt>
<dd><p>Lookup bond valence parameters from IUPAC table.
Args:</p>
<blockquote>
<div><p>cation: String, cation element
anion: String, anion element
cat_val: Integer, cation formal valence
an_val: Integer, anion formal valence</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>bond_val_list: dataframe of bond valence parameters</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.GlobalInstabilityIndex.get_equiv_sites">
<code class="sig-name descname">get_equiv_sites</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">s</span></em>, <em class="sig-param"><span class="n">site</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.GlobalInstabilityIndex.get_equiv_sites" title="Permalink to this definition">¶</a></dt>
<dd><p>Find identical sites from analyzing space group symmetry.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.GlobalInstabilityIndex.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.GlobalInstabilityIndex.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.GlobalInstabilityIndex.precheck">
<code class="sig-name descname">precheck</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">struct</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.GlobalInstabilityIndex.precheck" title="Permalink to this definition">¶</a></dt>
<dd><p>Bond valence methods require atom pairs with oxidation states.</p>
<p>Additionally, check if at least the first and last site’s species
have a entry in the bond valence parameters.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>struct: Pymatgen Structure</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.structure.GlobalSymmetryFeatures">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">GlobalSymmetryFeatures</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">desired_features</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.GlobalSymmetryFeatures" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Determines symmetry features, e.g. spacegroup number and  crystal system</p>
<dl class="simple">
<dt>Features:</dt><dd><ul class="simple">
<li><p>Spacegroup number</p></li>
<li><p>Crystal system (1 of 7)</p></li>
<li><p>Centrosymmetry (has inversion symmetry)</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.structure.GlobalSymmetryFeatures.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">desired_features</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.GlobalSymmetryFeatures.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.GlobalSymmetryFeatures.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.GlobalSymmetryFeatures.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt id="matminer.featurizers.structure.GlobalSymmetryFeatures.crystal_idx">
<code class="sig-name descname">crystal_idx</code><em class="property"> = {'cubic': 1, 'hexagonal': 2, 'monoclinic': 6, 'orthorhombic': 5, 'tetragonal': 4, 'triclinic': 7, 'trigonal': 3}</em><a class="headerlink" href="#matminer.featurizers.structure.GlobalSymmetryFeatures.crystal_idx" title="Permalink to this definition">¶</a></dt>
<dd></dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.GlobalSymmetryFeatures.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.GlobalSymmetryFeatures.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.GlobalSymmetryFeatures.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">s</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.GlobalSymmetryFeatures.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, which has to be implemented
in any derived featurizer subclass.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).</p>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.GlobalSymmetryFeatures.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.GlobalSymmetryFeatures.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.structure.JarvisCFID">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">JarvisCFID</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">use_cell</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">use_chem</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">use_chg</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">use_rdf</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">use_adf</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">use_ddf</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">use_nn</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.JarvisCFID" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Classical Force-Field Inspired Descriptors (CFID) from Jarvis-ML.</p>
<p>Chemo-structural descriptors from five different sub-methods,cincluding
pairwise radial, nearest neighbor, bond-angle, dihedral-angle and
core-charge distributions. With all descriptors enabled, there are 1,557
features per structure.</p>
<p>Adapted from the nist/jarvis package hosted at:
<a class="reference external" href="https://github.com/usnistgov/jarvis">https://github.com/usnistgov/jarvis</a></p>
<dl>
<dt>Find details at: <a class="reference external" href="https://journals.aps.org/prmaterials/abstract/10.1103/">https://journals.aps.org/prmaterials/abstract/10.1103/</a></dt><dd><p>PhysRevMaterials.2.083801</p>
</dd>
<dt>Args/Features:</dt><dd><dl class="simple">
<dt>use_cell (bool): Use structure cell descriptors (4 features, based</dt><dd><p>on DensityFeatures and log volume per atom).</p>
</dd>
</dl>
<p>use_chem (bool): Use chemical composition descriptors (438 features)
use_chg (bool): Use core charge descriptors (378 features)
use_adf (bool): Use angular distribution function (179 features x 2, one</p>
<blockquote>
<div><p>set of features for each cutoff).</p>
</div></blockquote>
<p>use_rdf (bool): Use radial distribution function (100 features)
use_ddf (bool): Use dihedral angle distribution function (179 features)
use_nn (bool): Use nearest neighbors (100 descriptors)</p>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.structure.JarvisCFID.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">use_cell</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">use_chem</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">use_chg</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">use_rdf</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">use_adf</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">use_ddf</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">use_nn</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.JarvisCFID.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.JarvisCFID.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.JarvisCFID.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.JarvisCFID.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.JarvisCFID.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.JarvisCFID.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">s</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.JarvisCFID.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get chemo-structural CFID decriptors</p>
<dl class="simple">
<dt>Args:</dt><dd><p>s: Structure object</p>
</dd>
<dt>Returns:</dt><dd><p>(np.ndarray) Final descriptors</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.JarvisCFID.get_chem">
<code class="sig-name descname">get_chem</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">element</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.JarvisCFID.get_chem" title="Permalink to this definition">¶</a></dt>
<dd><p>Get chemical descriptors for an element</p>
<dl class="simple">
<dt>Args:</dt><dd><p>element: element name</p>
</dd>
<dt>Returns:</dt><dd><p>arr: descriptor array value</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.JarvisCFID.get_chg">
<code class="sig-name descname">get_chg</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">element</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.JarvisCFID.get_chg" title="Permalink to this definition">¶</a></dt>
<dd><p>Get charge descriptors for an element</p>
<dl class="simple">
<dt>Args:</dt><dd><p>element: element name</p>
</dd>
<dt>Returns:</dt><dd><p>arr: descriptor array values</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.JarvisCFID.get_distributions">
<code class="sig-name descname">get_distributions</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">structure</span></em>, <em class="sig-param"><span class="n">c_size</span><span class="o">=</span><span class="default_value">10.0</span></em>, <em class="sig-param"><span class="n">max_cut</span><span class="o">=</span><span class="default_value">5.0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.JarvisCFID.get_distributions" title="Permalink to this definition">¶</a></dt>
<dd><p>Get radial and angular distribution functions</p>
<dl class="simple">
<dt>Args:</dt><dd><p>structure: Structure object
c_size: max. cell size
max_cut: max. bond cut-off for angular distribution</p>
</dd>
<dt>Retruns:</dt><dd><p>adfa, adfb, ddf, rdf, bondo
Angular distribution upto first cut-off
Angular distribution upto second cut-off
Dihedral angle distribution upto first cut-off
Radial distribution funcion
Bond order distribution</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.JarvisCFID.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.JarvisCFID.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.structure.MaximumPackingEfficiency">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">MaximumPackingEfficiency</code><a class="headerlink" href="#matminer.featurizers.structure.MaximumPackingEfficiency" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Maximum possible packing efficiency of this structure</p>
<p>Uses a Voronoi tessellation to determine the largest radius each atom
can have before any atoms touches any one of their neighbors. Given the
maximum radius size, this class computes the maximum packing efficiency
of the structure as a feature.</p>
<dl class="simple">
<dt>Features:</dt><dd><p>max packing efficiency - Maximum possible packing efficiency</p>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.structure.MaximumPackingEfficiency.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.MaximumPackingEfficiency.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.MaximumPackingEfficiency.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.MaximumPackingEfficiency.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.MaximumPackingEfficiency.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">strc</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.MaximumPackingEfficiency.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, which has to be implemented
in any derived featurizer subclass.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).</p>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.MaximumPackingEfficiency.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.MaximumPackingEfficiency.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.structure.MinimumRelativeDistances">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">MinimumRelativeDistances</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">10.0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.MinimumRelativeDistances" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Determines the relative distance of each site to its closest neighbor.</p>
<p>We use the relative distance,
f_ij = r_ij / (r^atom_i + r^atom_j), as a measure rather than the
absolute distances, r_ij, to account for the fact that different
atoms/species have different sizes.  The function uses the
valence-ionic radius estimator implemented in Pymatgen.
Args:</p>
<blockquote>
<div><dl class="simple">
<dt>cutoff: (float) (absolute) distance up to which tentative</dt><dd><p>closest neighbors (on the basis of relative distances)
are to be determined.</p>
</dd>
</dl>
</div></blockquote>
<dl class="py method">
<dt id="matminer.featurizers.structure.MinimumRelativeDistances.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">10.0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.MinimumRelativeDistances.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.MinimumRelativeDistances.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.MinimumRelativeDistances.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.MinimumRelativeDistances.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.MinimumRelativeDistances.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.MinimumRelativeDistances.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">s</span></em>, <em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">10.0</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.MinimumRelativeDistances.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get minimum relative distances of all sites of the input structure.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>s: Pymatgen Structure object.</p>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>dists_relative_min: (list of floats) list of all minimum relative</dt><dd><p>distances (i.e., for all sites).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.MinimumRelativeDistances.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.MinimumRelativeDistances.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.structure.OrbitalFieldMatrix">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">OrbitalFieldMatrix</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">period_tag</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">flatten</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.OrbitalFieldMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Representation based on the valence shell electrons of neighboring atoms.</p>
<p>Each atom is described by a 32-element vector (or 39-element vector, see
period tag for details) uniquely representing the valence subshell.
A 32x32 (39x39) matrix is formed by multiplying two atomic vectors.
An OFM for an atomic environment is the sum of these matrices for each atom
the center atom coordinates with multiplied by a distance function
(In this case, 1/r times the weight of the coordinating atom in the Voronoi</p>
<blockquote>
<div><p>Polyhedra method). The OFM of a structure or molecule is the average of the
OFMs for all the sites in the structure.</p>
</div></blockquote>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>period_tag (bool): In the original OFM, an element is represented</dt><dd><p>by a vector of length 32, where each element is 1 or 0,
which represents the valence subshell of the element.
With period_tag=True, the vector size is increased
to 39, where the 7 extra elements represent the period
of the element. Note lanthanides are treated as period 6,
actinides as period 7. Default False as in the original paper.</p>
</dd>
<dt>flatten (bool): Flatten the avg OFM to a 1024-vector (if period_tag</dt><dd><p>False) or a 1521-vector (if period_tag=True).</p>
</dd>
</dl>
</dd>
<dt>…attribute:: size</dt><dd><p>Either 32 or 39, the size of the vectors used to describe elements.</p>
</dd>
<dt>Reference:</dt><dd><p><cite>Pham et al. _Sci Tech Adv Mat_. 2017 &lt;http://dx.doi.org/10.1080/14686996.2017.1378060&gt;_</cite></p>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.structure.OrbitalFieldMatrix.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">period_tag</span><span class="o">=</span><span class="default_value">False</span></em>, <em class="sig-param"><span class="n">flatten</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.OrbitalFieldMatrix.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the featurizer</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>period_tag (bool): In the original OFM, an element is represented</dt><dd><p>by a vector of length 32, where each element is 1 or 0,
which represents the valence subshell of the element.
With period_tag=True, the vector size is increased
to 39, where the 7 extra elements represent the period
of the element. Note lanthanides are treated as period 6,
actinides as period 7. Default False as in the original paper.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.OrbitalFieldMatrix.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.OrbitalFieldMatrix.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.OrbitalFieldMatrix.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.OrbitalFieldMatrix.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.OrbitalFieldMatrix.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">s</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.OrbitalFieldMatrix.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Makes a supercell for structure s (to protect sites
from coordinating with themselves), and then finds the mean
of the orbital field matrices of each site to characterize
a structure</p>
<dl class="simple">
<dt>Args:</dt><dd><p>s (Structure): structure to characterize</p>
</dd>
<dt>Returns:</dt><dd><dl class="simple">
<dt>mean_ofm (size X size matrix): orbital field matrix</dt><dd><p>characterizing s</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.OrbitalFieldMatrix.get_atom_ofms">
<code class="sig-name descname">get_atom_ofms</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">struct</span></em>, <em class="sig-param"><span class="n">symm</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.OrbitalFieldMatrix.get_atom_ofms" title="Permalink to this definition">¶</a></dt>
<dd><p>Calls get_single_ofm for every site in struct. If symm=True,
get_single_ofm is called for symmetrically distinct sites, and
counts is constructed such that ofms[i] occurs counts[i] times
in the structure</p>
<dl>
<dt>Args:</dt><dd><p>struct (Structure): structure for find ofms for
symm (bool): whether to calculate ofm for only symmetrically</p>
<blockquote>
<div><p>distinct sites</p>
</div></blockquote>
</dd>
<dt>Returns:</dt><dd><p>ofms ([size X size matrix] X len(struct)): ofms for struct
if symm:</p>
<blockquote>
<div><dl class="simple">
<dt>ofms ([size X size matrix] X number of symmetrically distinct sites):</dt><dd><p>ofms for struct</p>
</dd>
</dl>
<p>counts: number of identical sites for each ofm</p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.OrbitalFieldMatrix.get_mean_ofm">
<code class="sig-name descname">get_mean_ofm</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">ofms</span></em>, <em class="sig-param"><span class="n">counts</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.OrbitalFieldMatrix.get_mean_ofm" title="Permalink to this definition">¶</a></dt>
<dd><p>Averages a list of ofms, weights by counts</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.OrbitalFieldMatrix.get_ohv">
<code class="sig-name descname">get_ohv</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">sp</span></em>, <em class="sig-param"><span class="n">period_tag</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.OrbitalFieldMatrix.get_ohv" title="Permalink to this definition">¶</a></dt>
<dd><p>Get the “one-hot-vector” for pymatgen Element sp. This 32 or 39-length
vector represents the valence shell of the given element.
Args:</p>
<blockquote>
<div><p>sp (Element): element whose ohv should be returned
period_tag (bool): If true, the vector contains items</p>
<blockquote>
<div><p>corresponding to the period of the element</p>
</div></blockquote>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><p>my_ohv (numpy array length 39 if period_tag, else 32): ohv for sp</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.OrbitalFieldMatrix.get_single_ofm">
<code class="sig-name descname">get_single_ofm</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">site</span></em>, <em class="sig-param"><span class="n">site_dict</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.OrbitalFieldMatrix.get_single_ofm" title="Permalink to this definition">¶</a></dt>
<dd><p>Gets the orbital field matrix for a single chemical environment,
where site is the center atom whose environment is characterized and
site_dict is a dictionary of site : weight, where the weights are the
Voronoi Polyhedra weights of the corresponding coordinating sites.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>site (Site): center atom
site_dict (dict of Site:float): chemical environment</p>
</dd>
<dt>Returns:</dt><dd><p>atom_ofm (size X size numpy matrix): ofm for site</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.OrbitalFieldMatrix.get_structure_ofm">
<code class="sig-name descname">get_structure_ofm</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">struct</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.OrbitalFieldMatrix.get_structure_ofm" title="Permalink to this definition">¶</a></dt>
<dd><p>Calls get_mean_ofm on the results of get_atom_ofms
to give a size X size matrix characterizing a structure</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.OrbitalFieldMatrix.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.OrbitalFieldMatrix.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.structure.PartialRadialDistributionFunction">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">PartialRadialDistributionFunction</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">20.0</span></em>, <em class="sig-param"><span class="n">bin_size</span><span class="o">=</span><span class="default_value">0.1</span></em>, <em class="sig-param"><span class="n">include_elems</span><span class="o">=</span><span class="default_value">()</span></em>, <em class="sig-param"><span class="n">exclude_elems</span><span class="o">=</span><span class="default_value">()</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.PartialRadialDistributionFunction" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Compute the partial radial distribution function (PRDF) of an xtal structure</p>
<p>The PRDF of a crystal structure is the radial distibution function broken
down for each pair of atom types.  The PRDF was proposed as a structural
descriptor by [Schutt <em>et al.</em>]
(<a class="reference external" href="https://journals.aps.org/prb/abstract/10.1103/PhysRevB.89.205118">https://journals.aps.org/prb/abstract/10.1103/PhysRevB.89.205118</a>)</p>
<dl class="simple">
<dt>Args:</dt><dd><p>cutoff: (float) distance up to which to calculate the RDF.
bin_size: (float) size of each bin of the (discrete) RDF.
include_elems: (list of string), list of elements that must be included in PRDF
exclude_elems: (list of string), list of elmeents that should not be included in PRDF</p>
</dd>
<dt>Features:</dt><dd><dl class="simple">
<dt>Each feature corresponds to the density of number of bonds</dt><dd><p>for a certain pair of elements at a certain range of
distances. For example, “Al-Al PRDF r=1.00-1.50” corresponds
to the density of Al-Al bonds between 1 and 1.5 distance units
By default, this featurizer generates RDFs for each pair
of elements in the training set.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.structure.PartialRadialDistributionFunction.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">20.0</span></em>, <em class="sig-param"><span class="n">bin_size</span><span class="o">=</span><span class="default_value">0.1</span></em>, <em class="sig-param"><span class="n">include_elems</span><span class="o">=</span><span class="default_value">()</span></em>, <em class="sig-param"><span class="n">exclude_elems</span><span class="o">=</span><span class="default_value">()</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.PartialRadialDistributionFunction.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.PartialRadialDistributionFunction.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.PartialRadialDistributionFunction.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.PartialRadialDistributionFunction.compute_prdf">
<code class="sig-name descname">compute_prdf</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">s</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.PartialRadialDistributionFunction.compute_prdf" title="Permalink to this definition">¶</a></dt>
<dd><p>Compute the PRDF for a structure</p>
<dl>
<dt>Args:</dt><dd><p>s: (Structure), structure to be evaluated</p>
</dd>
<dt>Returns:</dt><dd><p>dist_bins - float, start of each of the bins
prdf - dict, where the keys is a pair of elements (strings),</p>
<blockquote>
<div><p>and the value is the radial distribution function for those paris of elements</p>
</div></blockquote>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.PartialRadialDistributionFunction.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.PartialRadialDistributionFunction.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.PartialRadialDistributionFunction.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">s</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.PartialRadialDistributionFunction.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get PRDF of the input structure.
Args:</p>
<blockquote>
<div><p>s: Pymatgen Structure object.</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>prdf, dist: (tuple of arrays) the first element is a</dt><dd><p>dictionary where keys are tuples of element
names and values are PRDFs.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.PartialRadialDistributionFunction.fit">
<code class="sig-name descname">fit</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">X</span></em>, <em class="sig-param"><span class="n">y</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.PartialRadialDistributionFunction.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Define the list of elements to be included in the PRDF. By default,
the PRDF will include all of the elements in <cite>X</cite></p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>X: (numpy array nx1) structures used in the training set. Each entry</dt><dd><p>must be Pymatgen Structure objects.</p>
</dd>
</dl>
<p>y: <em>Not used</em>
fit_kwargs: <em>not used</em></p>
</dd>
<dt>Returns:</dt><dd><p>self</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.PartialRadialDistributionFunction.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.PartialRadialDistributionFunction.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.structure.RadialDistributionFunction">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">RadialDistributionFunction</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">20.0</span></em>, <em class="sig-param"><span class="n">bin_size</span><span class="o">=</span><span class="default_value">0.1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.RadialDistributionFunction" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Calculate the radial distribution function (RDF) of a crystal structure.</p>
<dl class="simple">
<dt>Features:</dt><dd><ul class="simple">
<li><p>Radial distribution function</p></li>
</ul>
</dd>
<dt>Args:</dt><dd><p>cutoff: (float) distance up to which to calculate the RDF.
bin_size: (float) size of each bin of the (discrete) RDF.</p>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.structure.RadialDistributionFunction.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">cutoff</span><span class="o">=</span><span class="default_value">20.0</span></em>, <em class="sig-param"><span class="n">bin_size</span><span class="o">=</span><span class="default_value">0.1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.RadialDistributionFunction.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.RadialDistributionFunction.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.RadialDistributionFunction.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.RadialDistributionFunction.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.RadialDistributionFunction.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.RadialDistributionFunction.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">s</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.RadialDistributionFunction.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Get RDF of the input structure.
Args:</p>
<blockquote>
<div><p>s (Structure): Pymatgen Structure object.</p>
</div></blockquote>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>rdf, dist: (tuple of arrays) the first element is the</dt><dd><p>normalized RDF, whereas the second element is
the inner radius of the RDF bin.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.RadialDistributionFunction.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.RadialDistributionFunction.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.structure.SineCoulombMatrix">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">SineCoulombMatrix</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">diag_elems</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">flatten</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.SineCoulombMatrix" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>A variant of the Coulomb matrix developed for periodic crystals.</p>
<p>This function generates a variant of the Coulomb matrix developed
for periodic crystals by Faber et al. (Inter. J. Quantum Chem.
115, 16, 2015). It is identical to the Coulomb matrix, except
that the inverse distance function is replaced by the inverse of a
sin**2 function of the vector between the sites which is periodic
in the dimensions of the structure lattice. See paper for details.</p>
<p>Coulomb Matrix features are flattened (for ML-readiness) by default. Use
fit before featurizing to use flattened features. To return the matrix form,
set flatten=False.</p>
<dl class="simple">
<dt>Args:</dt><dd><dl class="simple">
<dt>diag_elems (bool): flag indication whether (True, default) to use</dt><dd><p>the original definition of the diagonal elements; if set to False,
the diagonal elements are set to 0</p>
</dd>
<dt>flatten (bool): If True, returns a flattened vector based on eigenvalues</dt><dd><p>of the matrix form. Otherwise, returns a matrix object (single
feature), which will likely need to be processed further.</p>
</dd>
</dl>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.structure.SineCoulombMatrix.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">diag_elems</span><span class="o">=</span><span class="default_value">True</span></em>, <em class="sig-param"><span class="n">flatten</span><span class="o">=</span><span class="default_value">True</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.SineCoulombMatrix.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.SineCoulombMatrix.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.SineCoulombMatrix.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.SineCoulombMatrix.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.SineCoulombMatrix.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.SineCoulombMatrix.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">s</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.SineCoulombMatrix.featurize" title="Permalink to this definition">¶</a></dt>
<dd><dl class="simple">
<dt>Args:</dt><dd><p>s (Structure or Molecule): input structure (or molecule)</p>
</dd>
<dt>Returns:</dt><dd><p>(Nsites x Nsites matrix) Sine matrix or</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.SineCoulombMatrix.fit">
<code class="sig-name descname">fit</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">X</span></em>, <em class="sig-param"><span class="n">y</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.SineCoulombMatrix.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Fit the Sine Coulomb Matrix to a list of structures.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>X ([Structure]): A list of pymatgen structures.
y : unused (added for consistency with overridden method signature)</p>
</dd>
<dt>Returns:</dt><dd><p>self</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.SineCoulombMatrix.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.SineCoulombMatrix.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.structure.SiteStatsFingerprint">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">SiteStatsFingerprint</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">site_featurizer</span></em>, <em class="sig-param"><span class="n">stats</span><span class="o">=</span><span class="default_value">'mean', 'std_dev'</span></em>, <em class="sig-param"><span class="n">min_oxi</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">max_oxi</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">covariance</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.SiteStatsFingerprint" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Computes statistics of properties across all sites in a structure.</p>
<p>This featurizer first uses a site featurizer class (see site.py for
options) to compute features of each site in a structure, and then computes
features of the entire structure by measuring statistics of each attribute.
Can optionally compute the the statistics of only sites with certain ranges
of oxidation states (e.g., only anions).</p>
<dl class="simple">
<dt>Features:</dt><dd><ul class="simple">
<li><p>Returns each statistic of each site feature</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.structure.SiteStatsFingerprint.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">site_featurizer</span></em>, <em class="sig-param"><span class="n">stats</span><span class="o">=</span><span class="default_value">'mean', 'std_dev'</span></em>, <em class="sig-param"><span class="n">min_oxi</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">max_oxi</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="n">covariance</span><span class="o">=</span><span class="default_value">False</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.SiteStatsFingerprint.__init__" title="Permalink to this definition">¶</a></dt>
<dd><dl>
<dt>Args:</dt><dd><p>site_featurizer (BaseFeaturizer): a site-based featurizer
stats ([str]): list of weighted statistics to compute for each feature.</p>
<blockquote>
<div><p>If stats is None, a list is returned for each features
that contains the calculated feature for each site in the
structure.
<a href="#id27"><span class="problematic" id="id28">*</span></a>Note for nth mode, stat must be ‘n*_mode’; e.g. stat=’2nd_mode’</p>
</div></blockquote>
<dl class="simple">
<dt>min_oxi (int): minimum site oxidation state for inclusion (e.g.,</dt><dd><p>zero means metals/cations only)</p>
</dd>
</dl>
<p>max_oxi (int): maximum site oxidation state for inclusion
covariance (bool): Whether to compute the covariance of site features</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.SiteStatsFingerprint.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.SiteStatsFingerprint.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.SiteStatsFingerprint.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.SiteStatsFingerprint.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.SiteStatsFingerprint.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">s</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.SiteStatsFingerprint.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, which has to be implemented
in any derived featurizer subclass.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).</p>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.SiteStatsFingerprint.from_preset">
<em class="property">static </em><code class="sig-name descname">from_preset</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">preset</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.SiteStatsFingerprint.from_preset" title="Permalink to this definition">¶</a></dt>
<dd><p>Create a SiteStatsFingerprint class according to a preset</p>
<dl class="simple">
<dt>Args:</dt><dd><p>preset (str) - Name of preset
kwargs - Options for SiteStatsFingerprint</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.SiteStatsFingerprint.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.SiteStatsFingerprint.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.structure.StructuralComplexity">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">StructuralComplexity</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">symprec</span><span class="o">=</span><span class="default_value">0.1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.StructuralComplexity" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Shannon information entropy of a structure.</p>
<p>This descriptor treat a structure as a message
to evaluate structural complexity (<img class="math" src="_images/math/b988975be41fd13b4d091c10202ba19374643586.png" alt="S"/>)
using the following equation:</p>
<p><img class="math" src="_images/math/ba11027b67a5958f16fb1495fba3b5fb7eadcaee.png" alt="S = - v \sum_{i=1}^{k} p_i \log_2 p_i"/></p>
<p><img class="math" src="_images/math/0ee4480fa3fe2856c28f74b0aa03fde5e83cfb0a.png" alt="p_i = m_i / v"/></p>
<p>where <img class="math" src="_images/math/02d98909b5d6acd6a7ff927d4d42790bdd407d58.png" alt="v"/> is the total number of atoms in the unit cell,
<img class="math" src="_images/math/5247131b3b1b9cc457d29a9b08a19c2062ed3d16.png" alt="p_i"/> is the probability mass function,
<img class="math" src="_images/math/9630132210b904754c9ab272b61cb527d12263ca.png" alt="k"/> is the number of symmetrically inequivalent sites, and
<img class="math" src="_images/math/e1aaac0305e1b00c3c08173dd3e8402b6320c110.png" alt="m_i"/> is the number of sites classified in <img class="math" src="_images/math/5aa339d4daf45a810dda332e3c80a0698e526e04.png" alt="i"/> th
symmetrically inequivalent site.</p>
<dl class="simple">
<dt>Features:</dt><dd><ul class="simple">
<li><p>information entropy (bits/atom)</p></li>
<li><p>information entropy (bits/unit cell)</p></li>
</ul>
</dd>
<dt>Args:</dt><dd><p>symprec: precision for symmetrizing a structure</p>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.structure.StructuralComplexity.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">symprec</span><span class="o">=</span><span class="default_value">0.1</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.StructuralComplexity.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.StructuralComplexity.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.StructuralComplexity.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.StructuralComplexity.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.StructuralComplexity.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.StructuralComplexity.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">struct</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.StructuralComplexity.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, which has to be implemented
in any derived featurizer subclass.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).</p>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.StructuralComplexity.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.StructuralComplexity.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.structure.StructuralHeterogeneity">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">StructuralHeterogeneity</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">weight</span><span class="o">=</span><span class="default_value">'area'</span></em>, <em class="sig-param"><span class="n">stats</span><span class="o">=</span><span class="default_value">'minimum', 'maximum', 'range', 'mean', 'avg_dev'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.StructuralHeterogeneity" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Variance in the bond lengths and atomic volumes in a structure</p>
<p>These features are based on several statistics derived from the Voronoi
tessellation of a structure. The first set of features relate to the
variance in the average bond length across all atoms in the structure.
The second relate to the variance of bond lengths between each neighbor
of each atom. The final feature is the variance in Voronoi cell sizes
across the structure.</p>
<p>We define the ‘average bond length’ of a site as the weighted average of
the bond lengths for all neighbors. By default, the weight is the
area of the face between the sites.</p>
<p>The ‘neighbor distance variation’ is defined as the weighted mean absolute
deviation in both length for all neighbors of a particular site. As before,
the weight is according to face area by default. For this statistic, we
divide the mean absolute deviation by the mean neighbor distance for that
site.</p>
<dl class="simple">
<dt>Features:</dt><dd><dl class="simple">
<dt>mean absolute deviation in relative bond length - Mean absolute deviation</dt><dd><p>in the average bond lengths for all sites, divided by the
mean average bond length</p>
</dd>
<dt>max relative bond length - Maximum average bond length, divided by the</dt><dd><p>mean average bond length</p>
</dd>
<dt>min relative bond length - Minimum average bond length, divided by the</dt><dd><p>mean average bond length</p>
</dd>
<dt>[stat] neighbor distance variation - Statistic (e.g., mean) of the</dt><dd><p>neighbor distance variation</p>
</dd>
<dt>mean absolute deviation in relative cell size - Mean absolute deviation</dt><dd><p>in the Voronoi cell volume across all sites in the structure.
Divided by the mean Voronoi cell volume.</p>
</dd>
</dl>
</dd>
<dt>References:</dt><dd><p><a class="reference external" href="http://link.aps.org/doi/10.1103/PhysRevB.96.024104">Ward et al. _PRB_ 2017</a></p>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.structure.StructuralHeterogeneity.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">weight</span><span class="o">=</span><span class="default_value">'area'</span></em>, <em class="sig-param"><span class="n">stats</span><span class="o">=</span><span class="default_value">'minimum', 'maximum', 'range', 'mean', 'avg_dev'</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.StructuralHeterogeneity.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize self.  See help(type(self)) for accurate signature.</p>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.StructuralHeterogeneity.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.StructuralHeterogeneity.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.StructuralHeterogeneity.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.StructuralHeterogeneity.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.StructuralHeterogeneity.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">strc</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.StructuralHeterogeneity.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, which has to be implemented
in any derived featurizer subclass.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).</p>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.StructuralHeterogeneity.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.StructuralHeterogeneity.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.structure.StructureComposition">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">StructureComposition</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">featurizer</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.StructureComposition" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>Features related to the composition of a structure</p>
<p>This class is just a wrapper that calls a composition-based featurizer
on the composition of a Structure</p>
<dl class="simple">
<dt>Features:</dt><dd><ul class="simple">
<li><p>Depends on the featurizer</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt id="matminer.featurizers.structure.StructureComposition.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">featurizer</span><span class="o">=</span><span class="default_value">None</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.StructureComposition.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the featurizer</p>
<dl class="simple">
<dt>Args:</dt><dd><p>featurizer (BaseFeaturizer) - Composition-based featurizer</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.StructureComposition.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.StructureComposition.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.StructureComposition.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.StructureComposition.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.StructureComposition.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">strc</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.StructureComposition.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, which has to be implemented
in any derived featurizer subclass.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).</p>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.StructureComposition.fit">
<code class="sig-name descname">fit</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">X</span></em>, <em class="sig-param"><span class="n">y</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">fit_kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.StructureComposition.fit" title="Permalink to this definition">¶</a></dt>
<dd><p>Update the parameters of this featurizer based on available data</p>
<dl class="simple">
<dt>Args:</dt><dd><p>X - [list of tuples], training data</p>
</dd>
<dt>Returns:</dt><dd><p>self</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.StructureComposition.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.StructureComposition.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class">
<dt id="matminer.featurizers.structure.XRDPowderPattern">
<em class="property">class </em><code class="sig-prename descclassname">matminer.featurizers.structure.</code><code class="sig-name descname">XRDPowderPattern</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">two_theta_range</span><span class="o">=</span><span class="default_value">0, 127</span></em>, <em class="sig-param"><span class="n">bw_method</span><span class="o">=</span><span class="default_value">0.05</span></em>, <em class="sig-param"><span class="n">pattern_length</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.XRDPowderPattern" title="Permalink to this definition">¶</a></dt>
<dd><p>Bases: <a class="reference internal" href="#matminer.featurizers.base.BaseFeaturizer" title="matminer.featurizers.base.BaseFeaturizer"><code class="xref py py-class docutils literal notranslate"><span class="pre">matminer.featurizers.base.BaseFeaturizer</span></code></a></p>
<p>1D array representing powder diffraction of a structure as calculated by
pymatgen. The powder is smeared / normalized according to gaussian_kde.</p>
<dl class="py method">
<dt id="matminer.featurizers.structure.XRDPowderPattern.__init__">
<code class="sig-name descname">__init__</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">two_theta_range</span><span class="o">=</span><span class="default_value">0, 127</span></em>, <em class="sig-param"><span class="n">bw_method</span><span class="o">=</span><span class="default_value">0.05</span></em>, <em class="sig-param"><span class="n">pattern_length</span><span class="o">=</span><span class="default_value">None</span></em>, <em class="sig-param"><span class="o">**</span><span class="n">kwargs</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.XRDPowderPattern.__init__" title="Permalink to this definition">¶</a></dt>
<dd><p>Initialize the featurizer.</p>
<dl>
<dt>Args:</dt><dd><dl class="simple">
<dt>two_theta_range ([float of length 2]): Tuple for range of</dt><dd><p>two_thetas to calculate in degrees. Defaults to (0, 90). Set to
None if you want all diffracted beams within the limiting
sphere of radius 2 / wavelength.</p>
</dd>
</dl>
<p>bw_method (float): how much to smear the XRD pattern
pattern_length (float): length of final array; defaults to one value</p>
<blockquote>
<div><p>per degree (i.e. two_theta_range + 1)</p>
</div></blockquote>
<dl class="simple">
<dt><a href="#id30"><span class="problematic" id="id31">**</span></a>kwargs: any other arguments to pass into pymatgen’s XRDCalculator,</dt><dd><p>such as the type of radiation.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.XRDPowderPattern.citations">
<code class="sig-name descname">citations</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.XRDPowderPattern.citations" title="Permalink to this definition">¶</a></dt>
<dd><p>Citation(s) and reference(s) for this feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should be a string citation,</dt><dd><p>ideally in BibTeX format.</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.XRDPowderPattern.feature_labels">
<code class="sig-name descname">feature_labels</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.XRDPowderPattern.feature_labels" title="Permalink to this definition">¶</a></dt>
<dd><p>Generate attribute names.</p>
<dl class="simple">
<dt>Returns:</dt><dd><p>([str]) attribute labels.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.XRDPowderPattern.featurize">
<code class="sig-name descname">featurize</code><span class="sig-paren">(</span><em class="sig-param"><span class="n">strc</span></em><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.XRDPowderPattern.featurize" title="Permalink to this definition">¶</a></dt>
<dd><p>Main featurizer function, which has to be implemented
in any derived featurizer subclass.</p>
<dl class="simple">
<dt>Args:</dt><dd><p>x: input data to featurize (type depends on featurizer).</p>
</dd>
<dt>Returns:</dt><dd><p>(list) one or more features.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt id="matminer.featurizers.structure.XRDPowderPattern.implementors">
<code class="sig-name descname">implementors</code><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="headerlink" href="#matminer.featurizers.structure.XRDPowderPattern.implementors" title="Permalink to this definition">¶</a></dt>
<dd><p>List of implementors of the feature.</p>
<dl class="simple">
<dt>Returns:</dt><dd><dl class="simple">
<dt>(list) each element should either be a string with author name (e.g.,</dt><dd><p>“Anubhav Jain”) or a dictionary  with required key “name” and other
keys like “email” or “institution” (e.g., {“name”: “Anubhav
Jain”, “email”: “<a class="reference external" href="mailto:ajain&#37;&#52;&#48;lbl&#46;gov">ajain<span>&#64;</span>lbl<span>&#46;</span>gov</a>”, “institution”: “LBNL”}).</p>
</dd>
</dl>
</dd>
</dl>
</dd></dl>

</dd></dl>

</div>
<div class="section" id="module-matminer.featurizers">
<span id="module-contents"></span><h2>Module contents<a class="headerlink" href="#module-matminer.featurizers" title="Permalink to this headline">¶</a></h2>
</div>
</div>


            <div class="clearer"></div>
          </div>
        </div>
      </div>
      <div class="sphinxsidebar" role="navigation" aria-label="main navigation">
        <div class="sphinxsidebarwrapper">
  <h3><a href="index.html">Table of Contents</a></h3>
  <ul>
<li><a class="reference internal" href="#">matminer.featurizers package</a><ul>
<li><a class="reference internal" href="#subpackages">Subpackages</a></li>
<li><a class="reference internal" href="#submodules">Submodules</a></li>
<li><a class="reference internal" href="#module-matminer.featurizers.bandstructure">matminer.featurizers.bandstructure module</a></li>
<li><a class="reference internal" href="#module-matminer.featurizers.base">matminer.featurizers.base module</a></li>
<li><a class="reference internal" href="#module-matminer.featurizers.composition">matminer.featurizers.composition module</a></li>
<li><a class="reference internal" href="#module-matminer.featurizers.conversions">matminer.featurizers.conversions module</a></li>
<li><a class="reference internal" href="#module-matminer.featurizers.deprecated">matminer.featurizers.deprecated module</a></li>
<li><a class="reference internal" href="#module-matminer.featurizers.dos">matminer.featurizers.dos module</a></li>
<li><a class="reference internal" href="#module-matminer.featurizers.function">matminer.featurizers.function module</a></li>
<li><a class="reference internal" href="#module-matminer.featurizers.site">matminer.featurizers.site module</a></li>
<li><a class="reference internal" href="#module-matminer.featurizers.structure">matminer.featurizers.structure module</a></li>
<li><a class="reference internal" href="#module-matminer.featurizers">Module contents</a></li>
</ul>
</li>
</ul>

  <div role="note" aria-label="source link">
    <h3>This Page</h3>
    <ul class="this-page-menu">
      <li><a href="_sources/matminer.featurizers.rst.txt"
            rel="nofollow">Show Source</a></li>
    </ul>
   </div>
<div id="searchbox" style="display: none" role="search">
  <h3 id="searchlabel">Quick search</h3>
    <div class="searchformwrapper">
    <form class="search" action="search.html" method="get">
      <input type="text" name="q" aria-labelledby="searchlabel" />
      <input type="submit" value="Go" />
    </form>
    </div>
</div>
<script>$('#searchbox').show(0);</script>
        </div>
      </div>
      <div class="clearer"></div>
    </div>
    <div class="related" role="navigation" aria-label="related navigation">
      <h3>Navigation</h3>
      <ul>
        <li class="right" style="margin-right: 10px">
          <a href="genindex.html" title="General Index"
             >index</a></li>
        <li class="right" >
          <a href="py-modindex.html" title="Python Module Index"
             >modules</a> |</li>
        <li class="nav-item nav-item-0"><a href="index.html">matminer 0.6.5 documentation</a> &#187;</li>
        <li class="nav-item nav-item-this"><a href="">matminer.featurizers package</a></li> 
      </ul>
    </div>

    <div class="footer" role="contentinfo">
        &#169; Copyright 2015, Anubhav Jain.
      Created using <a href="https://www.sphinx-doc.org/">Sphinx</a> 3.2.1.
    </div>

  </body>
</html>